// @mastra/core/dist/background-tasks/index.cjs
'use strict';

var chunkDRYYU2XF_cjs = require('../chunk-DRYYU2XF.cjs');



Object.defineProperty(exports, "BACKGROUND_TASK_WORKFLOW_ID", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.BACKGROUND_TASK_WORKFLOW_ID; }
});
Object.defineProperty(exports, "BackgroundTaskManager", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.BackgroundTaskManager; }
});
Object.defineProperty(exports, "backgroundOverrideJsonSchema", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.backgroundOverrideJsonSchema; }
});
Object.defineProperty(exports, "backgroundOverrideZodSchema", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.backgroundOverrideZodSchema; }
});
Object.defineProperty(exports, "createBackgroundTask", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.createBackgroundTask; }
});
Object.defineProperty(exports, "generateBackgroundTaskSystemPrompt", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.generateBackgroundTaskSystemPrompt; }
});
Object.defineProperty(exports, "resolveBackgroundConfig", {
  enumerable: true,
  get: function () { return chunkDRYYU2XF_cjs.resolveBackgroundConfig; }
});
