// @mastra/server/dist/server/handlers/agent-controller.cjs
'use strict';

var chunkFPJGLJUZ_cjs = require('../../chunk-FPJGLJUZ.cjs');



Object.defineProperty(exports, "ABORT_AGENT_CONTROLLER_SESSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.ABORT_AGENT_CONTROLLER_SESSION_ROUTE; }
});
Object.defineProperty(exports, "AGENT_CONTROLLER_TOOL_APPROVAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.AGENT_CONTROLLER_TOOL_APPROVAL_ROUTE; }
});
Object.defineProperty(exports, "AGENT_CONTROLLER_TOOL_SUSPENSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.AGENT_CONTROLLER_TOOL_SUSPENSION_ROUTE; }
});
Object.defineProperty(exports, "CLEAR_AGENT_CONTROLLER_GOAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.CLEAR_AGENT_CONTROLLER_GOAL_ROUTE; }
});
Object.defineProperty(exports, "CLONE_AGENT_CONTROLLER_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.CLONE_AGENT_CONTROLLER_THREAD_ROUTE; }
});
Object.defineProperty(exports, "CREATE_AGENT_CONTROLLER_SESSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.CREATE_AGENT_CONTROLLER_SESSION_ROUTE; }
});
Object.defineProperty(exports, "CREATE_AGENT_CONTROLLER_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.CREATE_AGENT_CONTROLLER_THREAD_ROUTE; }
});
Object.defineProperty(exports, "DELETE_AGENT_CONTROLLER_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.DELETE_AGENT_CONTROLLER_THREAD_ROUTE; }
});
Object.defineProperty(exports, "FOLLOW_UP_AGENT_CONTROLLER_SESSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.FOLLOW_UP_AGENT_CONTROLLER_SESSION_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CONTROLLER_GOAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_GOAL_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CONTROLLER_OM_RECORD_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_OM_RECORD_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CONTROLLER_PERMISSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_PERMISSIONS_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CONTROLLER_RESOURCE_IDS_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_RESOURCE_IDS_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CONTROLLER_SESSION_STATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_SESSION_STATE_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_CONTROLLER_WORKSPACE_STATUS_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_WORKSPACE_STATUS_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_CONTROLLERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_CONTROLLER_MODELS_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_MODELS_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_CONTROLLER_MODES_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_MODES_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_CONTROLLER_THREADS_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_THREADS_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_CONTROLLER_THREAD_MESSAGES_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_THREAD_MESSAGES_ROUTE; }
});
Object.defineProperty(exports, "RENAME_AGENT_CONTROLLER_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.RENAME_AGENT_CONTROLLER_THREAD_ROUTE; }
});
Object.defineProperty(exports, "SEND_AGENT_CONTROLLER_MESSAGE_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SEND_AGENT_CONTROLLER_MESSAGE_ROUTE; }
});
Object.defineProperty(exports, "SEND_AGENT_CONTROLLER_NOTIFICATION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SEND_AGENT_CONTROLLER_NOTIFICATION_ROUTE; }
});
Object.defineProperty(exports, "SET_AGENT_CONTROLLER_CATEGORY_PERMISSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_CATEGORY_PERMISSION_ROUTE; }
});
Object.defineProperty(exports, "SET_AGENT_CONTROLLER_GOAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_GOAL_ROUTE; }
});
Object.defineProperty(exports, "SET_AGENT_CONTROLLER_RESOURCE_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_RESOURCE_ID_ROUTE; }
});
Object.defineProperty(exports, "SET_AGENT_CONTROLLER_SESSION_STATE_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_SESSION_STATE_ROUTE; }
});
Object.defineProperty(exports, "SET_AGENT_CONTROLLER_TOOL_PERMISSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_TOOL_PERMISSION_ROUTE; }
});
Object.defineProperty(exports, "STEER_AGENT_CONTROLLER_SESSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.STEER_AGENT_CONTROLLER_SESSION_ROUTE; }
});
Object.defineProperty(exports, "STREAM_AGENT_CONTROLLER_SESSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.STREAM_AGENT_CONTROLLER_SESSION_ROUTE; }
});
Object.defineProperty(exports, "SWITCH_AGENT_CONTROLLER_MODEL_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SWITCH_AGENT_CONTROLLER_MODEL_ROUTE; }
});
Object.defineProperty(exports, "SWITCH_AGENT_CONTROLLER_MODE_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SWITCH_AGENT_CONTROLLER_MODE_ROUTE; }
});
Object.defineProperty(exports, "SWITCH_AGENT_CONTROLLER_THREAD_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.SWITCH_AGENT_CONTROLLER_THREAD_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_AGENT_CONTROLLER_GOAL_ROUTE", {
  enumerable: true,
  get: function () { return chunkFPJGLJUZ_cjs.UPDATE_AGENT_CONTROLLER_GOAL_ROUTE; }
});
