// @mastra/server/dist/server/handlers/agent-versions.cjs
'use strict';

var chunkAR5PBVYY_cjs = require('../../chunk-AR5PBVYY.cjs');



Object.defineProperty(exports, "ACTIVATE_AGENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.ACTIVATE_AGENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "COMPARE_AGENT_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.COMPARE_AGENT_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "CREATE_AGENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.CREATE_AGENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "DELETE_AGENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.DELETE_AGENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "GET_AGENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.GET_AGENT_VERSION_ROUTE; }
});
Object.defineProperty(exports, "LIST_AGENT_VERSIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.LIST_AGENT_VERSIONS_ROUTE; }
});
Object.defineProperty(exports, "RESTORE_AGENT_VERSION_ROUTE", {
  enumerable: true,
  get: function () { return chunkAR5PBVYY_cjs.RESTORE_AGENT_VERSION_ROUTE; }
});
