// @mastra/server/dist/server/handlers/authorship.cjs
'use strict';

var chunkHDIYJI2Z_cjs = require('../../chunk-HDIYJI2Z.cjs');



Object.defineProperty(exports, "assertExecuteAccess", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.assertExecuteAccess; }
});
Object.defineProperty(exports, "assertOwnership", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.assertOwnership; }
});
Object.defineProperty(exports, "assertReadAccess", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.assertReadAccess; }
});
Object.defineProperty(exports, "assertShareAccess", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.assertShareAccess; }
});
Object.defineProperty(exports, "assertWriteAccess", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.assertWriteAccess; }
});
Object.defineProperty(exports, "getCallerAuthorId", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.getCallerAuthorId; }
});
Object.defineProperty(exports, "getCallerPermissions", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.getCallerPermissions; }
});
Object.defineProperty(exports, "hasAdminBypass", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.hasAdminBypass; }
});
Object.defineProperty(exports, "hasScopedPermission", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.hasScopedPermission; }
});
Object.defineProperty(exports, "matchesAuthorFilter", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.matchesAuthorFilter; }
});
Object.defineProperty(exports, "resolveAuthorFilter", {
  enumerable: true,
  get: function () { return chunkHDIYJI2Z_cjs.resolveAuthorFilter; }
});
