// @mastra/server/dist/server/handlers/gateway-memory-client.cjs
'use strict';

var chunk3QVZXWIE_cjs = require('../../chunk-3QVZXWIE.cjs');



Object.defineProperty(exports, "GatewayMemoryClient", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.GatewayMemoryClient; }
});
Object.defineProperty(exports, "getGatewayClient", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.getGatewayClient; }
});
Object.defineProperty(exports, "isGatewayAgent", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.isGatewayAgent; }
});
Object.defineProperty(exports, "isGatewayAgentAsync", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.isGatewayAgentAsync; }
});
Object.defineProperty(exports, "toLocalMessage", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.toLocalMessage; }
});
Object.defineProperty(exports, "toLocalOMRecord", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.toLocalOMRecord; }
});
Object.defineProperty(exports, "toLocalThread", {
  enumerable: true,
  get: function () { return chunk3QVZXWIE_cjs.toLocalThread; }
});
