// @mastra/server/dist/server/handlers/heartbeats.cjs
'use strict';

var chunk55BEULMJ_cjs = require('../../chunk-55BEULMJ.cjs');



Object.defineProperty(exports, "CREATE_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.CREATE_HEARTBEAT_ROUTE; }
});
Object.defineProperty(exports, "DELETE_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.DELETE_HEARTBEAT_ROUTE; }
});
Object.defineProperty(exports, "GET_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.GET_HEARTBEAT_ROUTE; }
});
Object.defineProperty(exports, "LIST_HEARTBEATS_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.LIST_HEARTBEATS_ROUTE; }
});
Object.defineProperty(exports, "PAUSE_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.PAUSE_HEARTBEAT_ROUTE; }
});
Object.defineProperty(exports, "RESUME_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.RESUME_HEARTBEAT_ROUTE; }
});
Object.defineProperty(exports, "RUN_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.RUN_HEARTBEAT_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_HEARTBEAT_ROUTE", {
  enumerable: true,
  get: function () { return chunk55BEULMJ_cjs.UPDATE_HEARTBEAT_ROUTE; }
});
