// @mastra/server/dist/server/handlers/mcp.cjs
'use strict';

var chunkUXYPH6HP_cjs = require('../../chunk-UXYPH6HP.cjs');



Object.defineProperty(exports, "EXECUTE_MCP_SERVER_TOOL_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.EXECUTE_MCP_SERVER_TOOL_ROUTE; }
});
Object.defineProperty(exports, "GET_MCP_SERVER_DETAIL_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.GET_MCP_SERVER_DETAIL_ROUTE; }
});
Object.defineProperty(exports, "GET_MCP_SERVER_TOOL_DETAIL_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.GET_MCP_SERVER_TOOL_DETAIL_ROUTE; }
});
Object.defineProperty(exports, "LIST_MCP_SERVERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.LIST_MCP_SERVERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_MCP_SERVER_RESOURCES_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.LIST_MCP_SERVER_RESOURCES_ROUTE; }
});
Object.defineProperty(exports, "LIST_MCP_SERVER_TOOLS_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.LIST_MCP_SERVER_TOOLS_ROUTE; }
});
Object.defineProperty(exports, "MCP_HTTP_TRANSPORT_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.MCP_HTTP_TRANSPORT_ROUTE; }
});
Object.defineProperty(exports, "MCP_SSE_MESSAGES_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.MCP_SSE_MESSAGES_ROUTE; }
});
Object.defineProperty(exports, "MCP_SSE_TRANSPORT_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.MCP_SSE_TRANSPORT_ROUTE; }
});
Object.defineProperty(exports, "READ_MCP_SERVER_RESOURCE_ROUTE", {
  enumerable: true,
  get: function () { return chunkUXYPH6HP_cjs.READ_MCP_SERVER_RESOURCE_ROUTE; }
});
