// @mastra/server/dist/server/handlers/scores.cjs
'use strict';

var chunkI4X4UHQW_cjs = require('../../chunk-I4X4UHQW.cjs');



Object.defineProperty(exports, "GET_SCORER_ROUTE", {
  enumerable: true,
  get: function () { return chunkI4X4UHQW_cjs.GET_SCORER_ROUTE; }
});
Object.defineProperty(exports, "LIST_SCORERS_ROUTE", {
  enumerable: true,
  get: function () { return chunkI4X4UHQW_cjs.LIST_SCORERS_ROUTE; }
});
Object.defineProperty(exports, "LIST_SCORES_BY_ENTITY_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkI4X4UHQW_cjs.LIST_SCORES_BY_ENTITY_ID_ROUTE; }
});
Object.defineProperty(exports, "LIST_SCORES_BY_RUN_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkI4X4UHQW_cjs.LIST_SCORES_BY_RUN_ID_ROUTE; }
});
Object.defineProperty(exports, "LIST_SCORES_BY_SCORER_ID_ROUTE", {
  enumerable: true,
  get: function () { return chunkI4X4UHQW_cjs.LIST_SCORES_BY_SCORER_ID_ROUTE; }
});
Object.defineProperty(exports, "SAVE_SCORE_ROUTE", {
  enumerable: true,
  get: function () { return chunkI4X4UHQW_cjs.SAVE_SCORE_ROUTE; }
});
