// @mastra/server/dist/server/handlers/stored-agents.cjs
'use strict';

var chunk6QQ4I6IY_cjs = require('../../chunk-6QQ4I6IY.cjs');



Object.defineProperty(exports, "CREATE_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.CREATE_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "DELETE_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.DELETE_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "EXPORT_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.EXPORT_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "GET_STORED_AGENT_DEPENDENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.GET_STORED_AGENT_DEPENDENTS_ROUTE; }
});
Object.defineProperty(exports, "GET_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.GET_STORED_AGENT_ROUTE; }
});
Object.defineProperty(exports, "LIST_STORED_AGENTS_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.LIST_STORED_AGENTS_ROUTE; }
});
Object.defineProperty(exports, "OPEN_STORED_AGENT_CHANGE_REQUEST_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.OPEN_STORED_AGENT_CHANGE_REQUEST_ROUTE; }
});
Object.defineProperty(exports, "PREVIEW_INSTRUCTIONS_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.PREVIEW_INSTRUCTIONS_ROUTE; }
});
Object.defineProperty(exports, "UPDATE_STORED_AGENT_ROUTE", {
  enumerable: true,
  get: function () { return chunk6QQ4I6IY_cjs.UPDATE_STORED_AGENT_ROUTE; }
});
