// @mastra/deployer/dist/chunk-TDNEF2M5.js
import { extractMastraOption } from './chunk-MPEFRVWJ.js';
import { getWorkspaceInformation, analyzeBundle, aliasHono } from './chunk-FQC6UUHS.js';
import { getInputOptions, nodeModulesExtensionResolver } from './chunk-2WEUOK6I.js';
import { tsConfigPaths } from './chunk-ORSNDVAB.js';
import { slash, getPackageName } from './chunk-V6F2NMS2.js';
import { dirname, posix, join } from 'path';
import { noopLogger } from '@mastra/core/logger';
import * as pkg from 'empathic/package';
import { watch } from 'rollup';
import { readdir, mkdir, writeFile, lstat, readFile, cp } from 'fs/promises';
import matter from 'gray-matter';

async function getInputOptions2(entryFile, platform, env, {
  sourcemap = false,
  bundlerOptions = {
    enableSourcemap: false,
    enableEsmShim: true,
    externals: true
  }
} = {}) {
  const closestPkgJson = pkg.up({ cwd: dirname(entryFile) });
  const projectRoot = closestPkgJson ? dirname(slash(closestPkgJson)) : slash(process.cwd());
  const { workspaceMap, workspaceRoot } = await getWorkspaceInformation({ mastraEntryFile: entryFile });
  const analyzeEntryResult = await analyzeBundle(
    [entryFile],
    entryFile,
    {
      outputDir: posix.join(process.cwd(), ".mastra", ".build"),
      projectRoot: workspaceRoot || process.cwd(),
      platform,
      isDev: true,
      bundlerOptions
    },
    noopLogger
  );
  const deps = /* @__PURE__ */ new Map();
  for (const [dep, metadata] of analyzeEntryResult.dependencies.entries()) {
    const pkgName = getPackageName(dep);
    if (pkgName && workspaceMap.has(pkgName)) {
      deps.set(dep, metadata);
    }
  }
  const inputOptions = await getInputOptions(
    entryFile,
    {
      dependencies: deps,
      externalDependencies: /* @__PURE__ */ new Map(),
      workspaceMap
    },
    platform,
    env,
    { sourcemap, isDev: true, workspaceRoot, projectRoot, externalsPreset: bundlerOptions?.externals === true }
  );
  if (Array.isArray(inputOptions.plugins)) {
    const plugins = [];
    inputOptions.plugins.forEach((plugin) => {
      if (plugin?.name === "node-resolve") {
        return;
      }
      if (plugin?.name === "tsconfig-paths") {
        plugins.push(
          tsConfigPaths({
            localResolve: true
          })
        );
        return;
      }
      plugins.push(plugin);
    });
    inputOptions.plugins = plugins;
    inputOptions.plugins.push(aliasHono());
    inputOptions.plugins.push(nodeModulesExtensionResolver());
  }
  return inputOptions;
}
async function createWatcher(inputOptions, outputOptions) {
  const watcher = await watch({
    ...inputOptions,
    output: {
      ...outputOptions,
      format: "esm",
      entryFileNames: "[name].mjs",
      chunkFileNames: "[name].mjs"
    }
  });
  return watcher;
}

// src/build/serverOptions.ts
async function getServerOptions(entryFile, outputDir, logger) {
  const result = await extractMastraOption("server", entryFile, outputDir, logger);
  if (!result) {
    return null;
  }
  return result.getConfig();
}
var CONFIG_BASENAMES = ["config.ts", "config.js"];
var WORKSPACE_BASENAMES = ["workspace.ts", "workspace.js"];
var MEMORY_BASENAMES = ["memory.ts", "memory.js"];
var INSTRUCTIONS_BASENAME = "instructions.md";
var TOOL_EXTENSIONS = [".ts", ".js"];
var SKILL_MODULE_EXTENSIONS = [".ts", ".js"];
var SKILL_MD_BASENAME = "SKILL.md";
async function exists(path) {
  try {
    return !(await lstat(path)).isSymbolicLink();
  } catch {
    return false;
  }
}
async function realDirectory(path) {
  try {
    const info = await lstat(path);
    if (info.isDirectory() && !info.isSymbolicLink()) {
      return slash(path);
    }
  } catch {
  }
  return void 0;
}
async function directoryExists(path) {
  return realDirectory(path);
}
async function firstExisting(dir, basenames) {
  for (const basename of basenames) {
    const candidate = join(dir, basename);
    if (await exists(candidate)) {
      return slash(candidate);
    }
  }
  return void 0;
}
function isTestFile(basename) {
  return /\.(test|spec)\.(ts|js)$/.test(basename);
}
function toolKey(basename) {
  return basename.replace(/\.(ts|js)$/, "");
}
async function discoverTools(toolsDir) {
  if (!await exists(toolsDir)) {
    return [];
  }
  let entries;
  try {
    entries = await readdir(toolsDir);
  } catch {
    return [];
  }
  const tools = [];
  for (const basename of entries.sort()) {
    if (isTestFile(basename)) {
      continue;
    }
    if (!TOOL_EXTENSIONS.some((ext) => basename.endsWith(ext))) {
      continue;
    }
    const path = join(toolsDir, basename);
    const stats = await lstat(path);
    if (stats.isSymbolicLink() || stats.isDirectory()) {
      continue;
    }
    tools.push({ key: toolKey(basename), path: slash(path) });
  }
  return tools;
}
async function readReferences(referencesDir) {
  if (!await exists(referencesDir)) {
    return {};
  }
  const references = {};
  let entries;
  try {
    entries = await readdir(referencesDir);
  } catch {
    return {};
  }
  for (const basename of entries.sort()) {
    const path = join(referencesDir, basename);
    const stats = await lstat(path);
    if (stats.isSymbolicLink() || stats.isDirectory()) {
      continue;
    }
    references[basename] = await readFile(path, "utf-8");
  }
  return references;
}
async function parsePackagedSkill(skillMdPath, fallbackName, references = {}) {
  const raw = await readFile(skillMdPath, "utf-8");
  const parsed = matter(raw);
  const frontmatter = parsed.data;
  const name = frontmatter.name ?? fallbackName;
  const description = frontmatter.description ?? "";
  const instructions = parsed.content.trim();
  return { kind: "packaged", name, description, instructions, references };
}
function skillModuleName(basename) {
  return basename.replace(/\.(ts|js)$/, "");
}
async function discoverSkills(skillsDir) {
  if (!await exists(skillsDir)) {
    return [];
  }
  let entries;
  try {
    entries = await readdir(skillsDir);
  } catch {
    return [];
  }
  const skills = [];
  for (const basename of entries.sort()) {
    if (isTestFile(basename)) {
      continue;
    }
    const path = join(skillsDir, basename);
    const stats = await lstat(path);
    if (stats.isSymbolicLink()) {
      continue;
    }
    const isDir = stats.isDirectory();
    if (isDir) {
      const skillMd = join(path, SKILL_MD_BASENAME);
      if (await exists(skillMd)) {
        const references = await readReferences(join(path, "references"));
        skills.push(await parsePackagedSkill(skillMd, skillModuleName(basename), references));
      }
      continue;
    }
    if (SKILL_MODULE_EXTENSIONS.some((ext) => basename.endsWith(ext))) {
      skills.push({ kind: "module", path: slash(path) });
      continue;
    }
    if (basename.endsWith(".md")) {
      const skill = await parsePackagedSkill(path, basename.replace(/\.md$/, ""));
      skills.push(skill);
    }
  }
  return skills;
}
async function discoverAgentDir(dir, name, allowSubagents, onWarn) {
  const configPath = await firstExisting(dir, CONFIG_BASENAMES);
  const instructionsPath = await exists(join(dir, INSTRUCTIONS_BASENAME)) ? slash(join(dir, INSTRUCTIONS_BASENAME)) : void 0;
  if (!configPath && !instructionsPath) {
    return void 0;
  }
  const workspacePath = await firstExisting(dir, WORKSPACE_BASENAMES);
  const memoryPath = await firstExisting(dir, MEMORY_BASENAMES);
  const workspaceSeedDir = await directoryExists(join(dir, "workspace"));
  const tools = await discoverTools(join(dir, "tools"));
  const skills = await discoverSkills(join(dir, "skills"));
  const subagents = await discoverSubagents(dir, allowSubagents, onWarn);
  return {
    name,
    dir: slash(dir),
    configPath,
    instructionsPath,
    workspacePath,
    memoryPath,
    workspaceSeedDir,
    tools,
    skills,
    subagents
  };
}
async function discoverSubagents(parentDir, allowSubagents, onWarn) {
  const subagentsDir = join(parentDir, "subagents");
  if (!await exists(subagentsDir)) {
    return [];
  }
  if (!allowSubagents) {
    onWarn?.(
      `Ignoring nested subagents in "${slash(subagentsDir)}": subagents are one level deep only, so a subagent cannot declare its own subagents.`
    );
    return [];
  }
  let entries;
  try {
    entries = await readdir(subagentsDir);
  } catch {
    return [];
  }
  const subagents = [];
  for (const name of entries.sort()) {
    const dir = join(subagentsDir, name);
    if (!await realDirectory(dir)) {
      continue;
    }
    const child = await discoverAgentDir(dir, name, false, onWarn);
    if (child) {
      subagents.push(child);
    }
  }
  return subagents;
}
async function discoverFsAgents(mastraDir, onWarn) {
  const agentsDir = join(mastraDir, "agents");
  if (!await exists(agentsDir)) {
    return [];
  }
  let entries;
  try {
    entries = await readdir(agentsDir);
  } catch {
    return [];
  }
  const discovered = [];
  for (const name of entries.sort()) {
    const dir = join(agentsDir, name);
    if (!await realDirectory(dir)) {
      continue;
    }
    const agent = await discoverAgentDir(dir, name, true, onWarn);
    if (agent) {
      discovered.push(agent);
    }
  }
  return discovered;
}
function sanitizeIdentifier(name, prefix, index) {
  const cleaned = name.replace(/[^a-zA-Z0-9_$]/g, "_");
  return `${prefix}_${index}_${cleaned}`;
}
async function emitAgentEntry(agent, idPath, workspaceName, allowSubagents, lines) {
  const configIdent = sanitizeIdentifier(agent.name, "config", idPath);
  const toolIdents = [];
  if (agent.configPath) {
    lines.push(`import ${configIdent} from ${JSON.stringify(agent.configPath)};`);
  }
  let workspaceIdent;
  if (agent.workspacePath) {
    workspaceIdent = sanitizeIdentifier(`${agent.name}_workspace`, "workspace", idPath);
    lines.push(`import ${workspaceIdent} from ${JSON.stringify(agent.workspacePath)};`);
  }
  let memoryIdent;
  if (agent.memoryPath) {
    memoryIdent = sanitizeIdentifier(`${agent.name}_memory`, "memory", idPath);
    lines.push(`import ${memoryIdent} from ${JSON.stringify(agent.memoryPath)};`);
  }
  for (let t = 0; t < agent.tools.length; t++) {
    const tool = agent.tools[t];
    const ident = sanitizeIdentifier(`${agent.name}_${tool.key}`, "tool", `${idPath}_${t}`);
    lines.push(`import ${ident} from ${JSON.stringify(tool.path)};`);
    toolIdents.push({ key: tool.key, ident });
  }
  const skillExprs = [];
  const agentSkills = agent.skills ?? [];
  for (let s = 0; s < agentSkills.length; s++) {
    const skill = agentSkills[s];
    if (skill.kind === "module") {
      const ident = sanitizeIdentifier(`${agent.name}_skill`, "skill", `${idPath}_${s}`);
      lines.push(`import ${ident} from ${JSON.stringify(skill.path)};`);
      skillExprs.push(ident);
    } else {
      const referenceFields = Object.entries(skill.references).map(
        ([key, value]) => `${JSON.stringify(key)}: ${JSON.stringify(value)}`
      );
      const skillFields = [
        `name: ${JSON.stringify(skill.name)}`,
        `description: ${JSON.stringify(skill.description)}`,
        `instructions: ${JSON.stringify(skill.instructions)}`
      ];
      if (referenceFields.length > 0) {
        skillFields.push(`references: { ${referenceFields.join(", ")} }`);
      }
      skillExprs.push(`__createSkill({ ${skillFields.join(", ")} })`);
    }
  }
  let instructionsMd;
  if (agent.instructionsPath) {
    instructionsMd = await readFile(agent.instructionsPath, "utf-8");
  }
  const subagentExprs = [];
  if (allowSubagents) {
    for (let c = 0; c < agent.subagents.length; c++) {
      const child = agent.subagents[c];
      const childExpr = await emitAgentEntry(child, `${idPath}_${c}`, `${workspaceName}/${child.name}`, false, lines);
      subagentExprs.push(childExpr);
    }
  }
  const entryFields = [`name: ${JSON.stringify(agent.name)}`];
  if (agent.configPath) {
    entryFields.push(`config: ${configIdent}`);
  }
  if (instructionsMd !== void 0) {
    entryFields.push(`instructionsMd: ${JSON.stringify(instructionsMd)}`);
  }
  if (toolIdents.length > 0) {
    const toolEntries = toolIdents.map(({ key, ident }) => `{ key: ${JSON.stringify(key)}, tool: ${ident} }`);
    entryFields.push(`tools: [${toolEntries.join(", ")}]`);
  }
  if (skillExprs.length > 0) {
    entryFields.push(`skills: [${skillExprs.join(", ")}]`);
  }
  if (subagentExprs.length > 0) {
    entryFields.push(`subagents: [${subagentExprs.join(", ")}]`);
  }
  if (workspaceIdent) {
    entryFields.push(`workspace: ${workspaceIdent}`);
  }
  if (memoryIdent) {
    entryFields.push(`memory: ${memoryIdent}`);
  }
  entryFields.push(`defaultWorkspaceBasePath: __workspaceBasePath(${JSON.stringify(workspaceName)})`);
  return `{ ${entryFields.join(", ")} }`;
}
async function generateFsAgentsModule(userEntry, agents) {
  const lines = [];
  const hasInlineSkills = agents.some(
    (a) => [a, ...a.subagents ?? []].some((x) => (x.skills ?? []).some((s) => s.kind === "packaged"))
  );
  lines.push(`import { assembleAgentFromFsEntry } from '@mastra/core/agent';`);
  if (hasInlineSkills) {
    lines.push(`import { createSkill as __createSkill } from '@mastra/core/skills';`);
  }
  lines.push(`import { fileURLToPath as __fileURLToPath } from 'node:url';`);
  lines.push(`import { dirname as __dirname, join as __join } from 'node:path';`);
  lines.push(`import * as __userEntry from ${JSON.stringify(userEntry)};`);
  lines.push(`export * from ${JSON.stringify(userEntry)};`);
  lines.push(``);
  lines.push(`const __bundleDir = __dirname(__fileURLToPath(import.meta.url));`);
  lines.push(`const __workspaceBasePath = name => __join(__bundleDir, 'workspace', ...name.split('/'));`);
  lines.push(``);
  const entryExprs = [];
  for (let i = 0; i < agents.length; i++) {
    const agent = agents[i];
    const expr = await emitAgentEntry(agent, `${i}`, agent.name, true, lines);
    entryExprs.push(expr);
  }
  lines.push(``);
  lines.push(`const __fsAgentEntries = [`);
  for (const expr of entryExprs) {
    lines.push(`  ${expr},`);
  }
  lines.push(`];`);
  lines.push(``);
  lines.push(`const __fsAgents = Object.create(null);`);
  lines.push(`for (const __entry of __fsAgentEntries) {`);
  lines.push(`  __fsAgents[__entry.name] = assembleAgentFromFsEntry(__entry, {`);
  lines.push(`    onWarn: msg => __userEntry.mastra?.getLogger?.()?.warn?.(msg) ?? console.warn(msg),`);
  lines.push(`  });`);
  lines.push(`}`);
  lines.push(``);
  lines.push(`if (__userEntry.mastra && typeof __userEntry.mastra.__registerFsAgents === 'function') {`);
  lines.push(`  __userEntry.mastra.__registerFsAgents(__fsAgents);`);
  lines.push(`}`);
  lines.push(``);
  lines.push(`export const mastra = __userEntry.mastra;`);
  return lines.join("\n");
}
async function prepareFsAgentsEntry(mastraDir, entryFile, outputDirectory) {
  const agents = await discoverFsAgents(mastraDir);
  if (agents.length === 0) {
    return { entryFile, toolPaths: [], agentCount: 0 };
  }
  const moduleSource = await generateFsAgentsModule(slash(entryFile), agents);
  const generatedEntry = join(outputDirectory, ".mastra-fs-agents-entry.mjs");
  const normalizedMastraDir = slash(mastraDir);
  const toolPaths = [
    posix.join(normalizedMastraDir, "agents/*/tools/**/*.{js,ts}"),
    `!${posix.join(normalizedMastraDir, "agents/*/tools/**/*.{test,spec}.{js,ts}")}`,
    `!${posix.join(normalizedMastraDir, "agents/*/tools/**/__tests__/**")}`
  ];
  return { entryFile: generatedEntry, toolPaths, agentCount: agents.length, moduleSource };
}
async function writeFsAgentsEntry(result) {
  if (!result.moduleSource) {
    return;
  }
  await mkdir(dirname(result.entryFile), { recursive: true });
  await writeFile(result.entryFile, result.moduleSource, "utf-8");
}
async function rejectSymlinks(source) {
  const stats = await lstat(source);
  return !stats.isSymbolicLink();
}
async function mirrorAgentSeeds(agent, workspaceName, bundleDir, mirrored) {
  if (agent.workspaceSeedDir) {
    const destination = join(bundleDir, "workspace", ...workspaceName.split("/"));
    await mkdir(destination, { recursive: true });
    await cp(agent.workspaceSeedDir, destination, { recursive: true, filter: rejectSymlinks });
    mirrored.push(workspaceName);
  }
  for (const child of agent.subagents ?? []) {
    await mirrorAgentSeeds(child, `${workspaceName}/${child.name}`, bundleDir, mirrored);
  }
}
async function mirrorFsAgentWorkspaces(mastraDir, bundleDir) {
  const agents = await discoverFsAgents(mastraDir);
  const mirrored = [];
  for (const agent of agents) {
    await mirrorAgentSeeds(agent, agent.name, bundleDir, mirrored);
  }
  return mirrored;
}

export { createWatcher, discoverFsAgents, generateFsAgentsModule, getInputOptions2 as getInputOptions, getServerOptions, mirrorFsAgentWorkspaces, prepareFsAgentsEntry, writeFsAgentsEntry };
