// @mastra/deployer/dist/index.js
import { Bundler } from './chunk-T6LS4OKA.js';
import { extractMastraOption } from './chunk-MPEFRVWJ.js';
import { DepsService, FileService } from './chunk-5GX66TF5.js';
export { Deps, FileService, createChildProcessLogger, createPinoStream } from './chunk-5GX66TF5.js';

// src/deploy/base.ts
var Deployer = class extends Bundler {
  deps = new DepsService();
  constructor(args) {
    super(args.name, "DEPLOYER");
    this.deps.__setLogger(this.logger);
  }
  getEnvFiles() {
    const possibleFiles = [".env.production", ".env.local", ".env"];
    try {
      const fileService = new FileService();
      const envFile = fileService.getFirstExistingFile(possibleFiles);
      return Promise.resolve([envFile]);
    } catch {
    }
    return Promise.resolve([]);
  }
};

// src/build/deployer.ts
async function getDeployer(entryFile, outputDir, logger) {
  const result = await extractMastraOption("deployer", entryFile, outputDir, logger);
  if (!result) {
    return null;
  }
  return result.getConfig();
}

export { Deployer, getDeployer };
