// @mastra/server/dist/chunk-4WZMR4IW.cjs
'use strict';

var zod = require('zod');

// src/server/schemas/schedules.ts
var scheduleStatusSchema = zod.z.enum(["active", "paused"]);
var workflowScheduleTargetSchema = zod.z.object({
  type: zod.z.literal("workflow"),
  workflowId: zod.z.string(),
  inputData: zod.z.unknown().optional(),
  initialState: zod.z.unknown().optional(),
  requestContext: zod.z.record(zod.z.string(), zod.z.unknown()).optional()
});
var signalAttributesSchema = zod.z.record(
  zod.z.string(),
  zod.z.union([zod.z.string(), zod.z.number(), zod.z.boolean(), zod.z.null()]).optional()
);
var ifActiveSchema = zod.z.object({
  behavior: zod.z.enum(["deliver", "persist", "discard"]).optional(),
  attributes: signalAttributesSchema.optional()
});
var ifIdleSchema = zod.z.object({
  behavior: zod.z.enum(["wake", "persist", "discard"]).optional(),
  attributes: signalAttributesSchema.optional(),
  streamOptions: zod.z.object({
    requestContext: zod.z.record(zod.z.string(), zod.z.unknown()).optional()
  }).optional()
});
var heartbeatScheduleTargetSchema = zod.z.object({
  type: zod.z.literal("heartbeat"),
  agentId: zod.z.string(),
  prompt: zod.z.string(),
  threadId: zod.z.string().optional(),
  resourceId: zod.z.string().optional(),
  signalType: zod.z.string().optional(),
  tagName: zod.z.string().optional(),
  attributes: signalAttributesSchema.optional(),
  ifActive: ifActiveSchema.optional(),
  ifIdle: ifIdleSchema.optional(),
  providerOptions: zod.z.record(zod.z.string(), zod.z.unknown()).optional(),
  requestContext: zod.z.record(zod.z.string(), zod.z.unknown()).optional()
});
var scheduleTargetSchema = zod.z.discriminatedUnion("type", [
  workflowScheduleTargetSchema,
  heartbeatScheduleTargetSchema
]);
var workflowRunStatusSchema = zod.z.enum([
  "running",
  "success",
  "failed",
  "tripwire",
  "suspended",
  "waiting",
  "pending",
  "canceled",
  "bailed",
  "paused",
  "skipped"
]);
var scheduleRunSummarySchema = zod.z.object({
  status: workflowRunStatusSchema,
  startedAt: zod.z.number().optional(),
  completedAt: zod.z.number().optional(),
  durationMs: zod.z.number().optional(),
  error: zod.z.string().optional()
});
var scheduleResponseSchema = zod.z.object({
  id: zod.z.string(),
  target: scheduleTargetSchema,
  cron: zod.z.string(),
  timezone: zod.z.string().optional(),
  status: scheduleStatusSchema,
  nextFireAt: zod.z.number(),
  lastFireAt: zod.z.number().optional(),
  lastRunId: zod.z.string().optional(),
  lastRun: scheduleRunSummarySchema.optional(),
  metadata: zod.z.record(zod.z.string(), zod.z.unknown()).optional(),
  ownerType: zod.z.string().optional(),
  ownerId: zod.z.string().optional(),
  createdAt: zod.z.number(),
  updatedAt: zod.z.number()
});
var scheduleTriggerOutcomeSchema = zod.z.enum([
  "published",
  "succeeded",
  "delivered",
  "persisted",
  "discarded",
  "skipped",
  "aborted",
  "failed",
  // Legacy queue/notification outcomes — no longer written, but trigger rows
  // persisted by older builds may still carry them. Kept readable so the
  // response validator does not reject historical rows. Mirrors the core
  // ScheduleTriggerOutcome union.
  "acked",
  "alerted",
  "deferred",
  "appended-from-queue",
  "dropped-stale",
  "dropped-superseded",
  "dropped-busy"
]);
var scheduleTriggerKindSchema = zod.z.enum(["schedule-fire", "queue-drain", "manual"]);
var scheduleTriggerResponseSchema = zod.z.object({
  id: zod.z.string().optional(),
  scheduleId: zod.z.string(),
  runId: zod.z.string().nullable(),
  scheduledFireAt: zod.z.number(),
  actualFireAt: zod.z.number(),
  outcome: scheduleTriggerOutcomeSchema,
  error: zod.z.string().optional(),
  triggerKind: scheduleTriggerKindSchema.optional(),
  parentTriggerId: zod.z.string().optional(),
  metadata: zod.z.record(zod.z.string(), zod.z.unknown()).optional(),
  run: scheduleRunSummarySchema.optional()
});
var listSchedulesQuerySchema = zod.z.object({
  workflowId: zod.z.string().optional(),
  status: scheduleStatusSchema.optional(),
  ownerType: zod.z.string().optional(),
  ownerId: zod.z.string().optional()
}).refine((data) => data.ownerId === void 0 || data.ownerType !== void 0, {
  message: "ownerId can only be used together with ownerType",
  path: ["ownerId"]
});
var listSchedulesResponseSchema = zod.z.object({
  schedules: zod.z.array(scheduleResponseSchema)
});
var scheduleIdPathParams = zod.z.object({
  scheduleId: zod.z.string()
});
var listScheduleTriggersQuerySchema = zod.z.object({
  limit: zod.z.coerce.number().int().positive().optional(),
  fromActualFireAt: zod.z.coerce.number().int().nonnegative().optional(),
  toActualFireAt: zod.z.coerce.number().int().nonnegative().optional()
});
var listScheduleTriggersResponseSchema = zod.z.object({
  triggers: zod.z.array(scheduleTriggerResponseSchema)
});

exports.listScheduleTriggersQuerySchema = listScheduleTriggersQuerySchema;
exports.listScheduleTriggersResponseSchema = listScheduleTriggersResponseSchema;
exports.listSchedulesQuerySchema = listSchedulesQuerySchema;
exports.listSchedulesResponseSchema = listSchedulesResponseSchema;
exports.scheduleIdPathParams = scheduleIdPathParams;
exports.scheduleResponseSchema = scheduleResponseSchema;
exports.scheduleRunSummarySchema = scheduleRunSummarySchema;
