// @mastra/server/dist/chunk-FKM644Z6.cjs
'use strict';

var chunk3EZOV6A3_cjs = require('./chunk-3EZOV6A3.cjs');
var chunkSORUUFZK_cjs = require('./chunk-SORUUFZK.cjs');
var chunkEG6DDVIJ_cjs = require('./chunk-EG6DDVIJ.cjs');
var chunkH7NWUAHC_cjs = require('./chunk-H7NWUAHC.cjs');
var chunkYDUA7L5F_cjs = require('./chunk-YDUA7L5F.cjs');
var chunkZGAUP346_cjs = require('./chunk-ZGAUP346.cjs');
var chunkLRRO5PBP_cjs = require('./chunk-LRRO5PBP.cjs');
var chunk6QQ4I6IY_cjs = require('./chunk-6QQ4I6IY.cjs');
var chunkIQZE4X6U_cjs = require('./chunk-IQZE4X6U.cjs');
var chunkEMQZIPXG_cjs = require('./chunk-EMQZIPXG.cjs');
var chunkKE32UGIC_cjs = require('./chunk-KE32UGIC.cjs');
var chunkJYWPYMQS_cjs = require('./chunk-JYWPYMQS.cjs');
var chunkXPJZ72ZH_cjs = require('./chunk-XPJZ72ZH.cjs');
var chunkQ5CGBHAI_cjs = require('./chunk-Q5CGBHAI.cjs');
var chunkJ6XQWEEV_cjs = require('./chunk-J6XQWEEV.cjs');
var chunkFOANUZQ3_cjs = require('./chunk-FOANUZQ3.cjs');
var chunkMTWCBGFH_cjs = require('./chunk-MTWCBGFH.cjs');
var chunkMJH4EEUV_cjs = require('./chunk-MJH4EEUV.cjs');
var chunk55BEULMJ_cjs = require('./chunk-55BEULMJ.cjs');
var chunkAR5PBVYY_cjs = require('./chunk-AR5PBVYY.cjs');
var chunk6QZLGOUC_cjs = require('./chunk-6QZLGOUC.cjs');
var chunkLAY7G3RT_cjs = require('./chunk-LAY7G3RT.cjs');
var chunkSXMPMMTX_cjs = require('./chunk-SXMPMMTX.cjs');
var chunkQPICFOIK_cjs = require('./chunk-QPICFOIK.cjs');
var chunkPMGFP7II_cjs = require('./chunk-PMGFP7II.cjs');
var chunkDR2XLMGB_cjs = require('./chunk-DR2XLMGB.cjs');
var chunkYRZYCQGE_cjs = require('./chunk-YRZYCQGE.cjs');
var chunkI4X4UHQW_cjs = require('./chunk-I4X4UHQW.cjs');
var chunkOJB6MLRD_cjs = require('./chunk-OJB6MLRD.cjs');
var chunkUXYPH6HP_cjs = require('./chunk-UXYPH6HP.cjs');
var chunkLVGS5ADS_cjs = require('./chunk-LVGS5ADS.cjs');
var chunkY32XM46K_cjs = require('./chunk-Y32XM46K.cjs');
var chunk5GQWTNAY_cjs = require('./chunk-5GQWTNAY.cjs');
var chunkKN22ZEKU_cjs = require('./chunk-KN22ZEKU.cjs');
var chunk5O3XY7Z3_cjs = require('./chunk-5O3XY7Z3.cjs');
var chunkTGFTKXT4_cjs = require('./chunk-TGFTKXT4.cjs');
var chunkSCY53JIH_cjs = require('./chunk-SCY53JIH.cjs');
var chunkAJ4VAZFT_cjs = require('./chunk-AJ4VAZFT.cjs');
var chunkI4COPLOY_cjs = require('./chunk-I4COPLOY.cjs');
var chunkOE5GVCW4_cjs = require('./chunk-OE5GVCW4.cjs');
var chunkFPJGLJUZ_cjs = require('./chunk-FPJGLJUZ.cjs');

// src/server/server-adapter/routes/permissions.ts
var METHOD_TO_ACTION = {
  GET: "read",
  POST: "write",
  // Default for POST, may be overridden to 'execute'
  PUT: "write",
  PATCH: "write",
  DELETE: "delete"
};
var EXECUTE_PATTERNS = [
  "/generate",
  "/stream",
  "/execute",
  "/start",
  "/resume",
  "/restart",
  "/cancel",
  "/approve",
  "/decline",
  "/speak",
  "/listen",
  "/query",
  "/search",
  "/observe",
  "/time-travel",
  "/enhance",
  "/clone"
];
var PUBLISH_PATTERNS = ["/publish", "/activate", "/restore"];
var STORED_RESOURCE_SEGMENTS = {
  agents: "stored-agents",
  "mcp-clients": "stored-mcp-clients",
  "prompt-blocks": "stored-prompt-blocks",
  scorers: "stored-scorers",
  skills: "stored-skills",
  workspaces: "stored-workspaces"
};
function extractResource(path) {
  const segments = path.replace(/^\//, "").split("/");
  if (segments.length === 0) {
    return null;
  }
  const firstSegment = segments[0];
  if (firstSegment === "stored" && segments[1]) {
    return STORED_RESOURCE_SEGMENTS[segments[1]] ?? null;
  }
  if (firstSegment === ".well-known") {
    return "a2a";
  }
  return firstSegment || null;
}
function deriveAction(method, path) {
  const upperMethod = method.toUpperCase();
  if (upperMethod === "POST") {
    if (path.startsWith("/stored/")) {
      const isPublishOperation = PUBLISH_PATTERNS.some((pattern) => path.endsWith(pattern));
      if (isPublishOperation) {
        return "publish";
      }
    }
    const isExecuteOperation = EXECUTE_PATTERNS.some((pattern) => path.includes(pattern));
    return isExecuteOperation ? "execute" : "write";
  }
  return METHOD_TO_ACTION[upperMethod] || "read";
}
function derivePermission(route) {
  if (route.method === "ALL") {
    return null;
  }
  const resource = extractResource(route.path);
  if (!resource) {
    return null;
  }
  const action = deriveAction(route.method, route.path);
  return `${resource}:${action}`;
}
function getEffectivePermission(route) {
  if (route.requiresAuth === false) {
    return null;
  }
  if (route.requiresPermission) {
    return route.requiresPermission;
  }
  return derivePermission(route);
}

// src/server/server-adapter/routes/a2a.ts
var A2A_ROUTES = [chunkTGFTKXT4_cjs.GET_AGENT_CARD_ROUTE, chunkTGFTKXT4_cjs.AGENT_EXECUTION_ROUTE];

// src/server/server-adapter/routes/agent-builder.ts
var AGENT_BUILDER_ROUTES = [
  chunkI4COPLOY_cjs.LIST_AGENT_BUILDER_ACTIONS_ROUTE,
  chunkI4COPLOY_cjs.GET_AGENT_BUILDER_ACTION_BY_ID_ROUTE,
  chunkI4COPLOY_cjs.LIST_AGENT_BUILDER_ACTION_RUNS_ROUTE,
  chunkI4COPLOY_cjs.GET_AGENT_BUILDER_ACTION_RUN_BY_ID_ROUTE,
  chunkI4COPLOY_cjs.CREATE_AGENT_BUILDER_ACTION_RUN_ROUTE,
  chunkI4COPLOY_cjs.STREAM_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.START_ASYNC_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.START_AGENT_BUILDER_ACTION_RUN_ROUTE,
  chunkI4COPLOY_cjs.OBSERVE_STREAM_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.OBSERVE_STREAM_LEGACY_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.RESUME_ASYNC_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.RESUME_NO_WAIT_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.RESUME_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.RESUME_STREAM_AGENT_BUILDER_ACTION_ROUTE,
  chunkI4COPLOY_cjs.CANCEL_AGENT_BUILDER_ACTION_RUN_ROUTE
];

// src/server/server-adapter/routes/agent-controller.ts
var AGENT_CONTROLLER_ROUTES = [
  chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLERS_ROUTE,
  chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_MODES_ROUTE,
  chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_MODELS_ROUTE,
  chunkFPJGLJUZ_cjs.CREATE_AGENT_CONTROLLER_SESSION_ROUTE,
  chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_SESSION_STATE_ROUTE,
  chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_THREADS_ROUTE,
  chunkFPJGLJUZ_cjs.CREATE_AGENT_CONTROLLER_THREAD_ROUTE,
  chunkFPJGLJUZ_cjs.DELETE_AGENT_CONTROLLER_THREAD_ROUTE,
  chunkFPJGLJUZ_cjs.RENAME_AGENT_CONTROLLER_THREAD_ROUTE,
  chunkFPJGLJUZ_cjs.CLONE_AGENT_CONTROLLER_THREAD_ROUTE,
  chunkFPJGLJUZ_cjs.LIST_AGENT_CONTROLLER_THREAD_MESSAGES_ROUTE,
  chunkFPJGLJUZ_cjs.STREAM_AGENT_CONTROLLER_SESSION_ROUTE,
  chunkFPJGLJUZ_cjs.SEND_AGENT_CONTROLLER_MESSAGE_ROUTE,
  chunkFPJGLJUZ_cjs.STEER_AGENT_CONTROLLER_SESSION_ROUTE,
  chunkFPJGLJUZ_cjs.FOLLOW_UP_AGENT_CONTROLLER_SESSION_ROUTE,
  chunkFPJGLJUZ_cjs.ABORT_AGENT_CONTROLLER_SESSION_ROUTE,
  chunkFPJGLJUZ_cjs.AGENT_CONTROLLER_TOOL_APPROVAL_ROUTE,
  chunkFPJGLJUZ_cjs.AGENT_CONTROLLER_TOOL_SUSPENSION_ROUTE,
  chunkFPJGLJUZ_cjs.SWITCH_AGENT_CONTROLLER_MODE_ROUTE,
  chunkFPJGLJUZ_cjs.SWITCH_AGENT_CONTROLLER_MODEL_ROUTE,
  chunkFPJGLJUZ_cjs.SWITCH_AGENT_CONTROLLER_THREAD_ROUTE,
  chunkFPJGLJUZ_cjs.SEND_AGENT_CONTROLLER_NOTIFICATION_ROUTE,
  chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_WORKSPACE_STATUS_ROUTE,
  chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_OM_RECORD_ROUTE,
  chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_RESOURCE_ID_ROUTE,
  chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_RESOURCE_IDS_ROUTE,
  chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_GOAL_ROUTE,
  chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_GOAL_ROUTE,
  chunkFPJGLJUZ_cjs.UPDATE_AGENT_CONTROLLER_GOAL_ROUTE,
  chunkFPJGLJUZ_cjs.CLEAR_AGENT_CONTROLLER_GOAL_ROUTE,
  chunkFPJGLJUZ_cjs.GET_AGENT_CONTROLLER_PERMISSIONS_ROUTE,
  chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_CATEGORY_PERMISSION_ROUTE,
  chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_TOOL_PERMISSION_ROUTE,
  chunkFPJGLJUZ_cjs.SET_AGENT_CONTROLLER_SESSION_STATE_ROUTE
];

// src/server/server-adapter/routes/agents.ts
var AGENTS_ROUTES = [
  // ============================================================================
  // Agent Core Routes
  // ============================================================================
  chunkSCY53JIH_cjs.LIST_AGENTS_ROUTE,
  chunkSCY53JIH_cjs.GET_PROVIDERS_ROUTE,
  chunkSCY53JIH_cjs.GET_AGENT_BY_ID_ROUTE,
  chunkSCY53JIH_cjs.CLONE_AGENT_ROUTE,
  // ============================================================================
  // Voice Routes
  // ============================================================================
  chunkPMGFP7II_cjs.GET_SPEAKERS_ROUTE,
  chunkPMGFP7II_cjs.GET_SPEAKERS_DEPRECATED_ROUTE,
  // ============================================================================
  // Agent Execution Routes
  // ============================================================================
  chunkSCY53JIH_cjs.GENERATE_AGENT_ROUTE,
  chunkSCY53JIH_cjs.GENERATE_AGENT_VNEXT_ROUTE,
  chunkSCY53JIH_cjs.STREAM_GENERATE_ROUTE,
  chunkSCY53JIH_cjs.STREAM_UNTIL_IDLE_GENERATE_ROUTE,
  chunkSCY53JIH_cjs.STREAM_GENERATE_VNEXT_DEPRECATED_ROUTE,
  // ============================================================================
  // Resumable Stream Routes
  // ============================================================================
  chunkSCY53JIH_cjs.OBSERVE_AGENT_STREAM_ROUTE,
  chunkSCY53JIH_cjs.SEND_AGENT_MESSAGE_ROUTE,
  chunkSCY53JIH_cjs.QUEUE_AGENT_MESSAGE_ROUTE,
  chunkSCY53JIH_cjs.SEND_AGENT_SIGNAL_ROUTE,
  chunkSCY53JIH_cjs.ABORT_AGENT_THREAD_ROUTE,
  chunkSCY53JIH_cjs.SUBSCRIBE_AGENT_THREAD_ROUTE,
  // ============================================================================
  // Tool Routes
  // ============================================================================
  chunkYRZYCQGE_cjs.EXECUTE_AGENT_TOOL_ROUTE,
  chunkSCY53JIH_cjs.APPROVE_TOOL_CALL_ROUTE,
  chunkSCY53JIH_cjs.SEND_TOOL_APPROVAL_ROUTE,
  chunkSCY53JIH_cjs.LIST_SUSPENDED_RUNS_ROUTE,
  chunkSCY53JIH_cjs.DECLINE_TOOL_CALL_ROUTE,
  chunkSCY53JIH_cjs.RESUME_STREAM_ROUTE,
  chunkSCY53JIH_cjs.APPROVE_TOOL_CALL_GENERATE_ROUTE,
  chunkSCY53JIH_cjs.DECLINE_TOOL_CALL_GENERATE_ROUTE,
  chunkSCY53JIH_cjs.APPROVE_NETWORK_TOOL_CALL_ROUTE,
  chunkSCY53JIH_cjs.DECLINE_NETWORK_TOOL_CALL_ROUTE,
  chunkSCY53JIH_cjs.RESUME_STREAM_UNTIL_IDLE_ROUTE,
  // ============================================================================
  // Network Routes
  // ============================================================================
  chunkSCY53JIH_cjs.STREAM_NETWORK_ROUTE,
  // ============================================================================
  // Model Management Routes
  // ============================================================================
  chunkSCY53JIH_cjs.UPDATE_AGENT_MODEL_ROUTE,
  chunkSCY53JIH_cjs.RESET_AGENT_MODEL_ROUTE,
  chunkSCY53JIH_cjs.REORDER_AGENT_MODEL_LIST_ROUTE,
  chunkSCY53JIH_cjs.UPDATE_AGENT_MODEL_IN_MODEL_LIST_ROUTE,
  // ============================================================================
  // Instruction Enhancement Routes
  // ============================================================================
  chunkSCY53JIH_cjs.ENHANCE_INSTRUCTIONS_ROUTE,
  // ============================================================================
  // Agent Tool Routes
  // ============================================================================
  chunkYRZYCQGE_cjs.GET_AGENT_TOOL_ROUTE,
  // ============================================================================
  // Agent Skill Routes
  // ============================================================================
  chunkSCY53JIH_cjs.GET_AGENT_SKILL_ROUTE,
  // ============================================================================
  // Voice/Speech Routes
  // ============================================================================
  chunkPMGFP7II_cjs.GENERATE_SPEECH_ROUTE,
  chunkPMGFP7II_cjs.GENERATE_SPEECH_DEPRECATED_ROUTE,
  chunkPMGFP7II_cjs.TRANSCRIBE_SPEECH_ROUTE,
  chunkPMGFP7II_cjs.TRANSCRIBE_SPEECH_DEPRECATED_ROUTE,
  chunkPMGFP7II_cjs.GET_LISTENER_ROUTE,
  // ============================================================================
  // Deprecated Routes
  // ============================================================================
  chunkSCY53JIH_cjs.STREAM_VNEXT_DEPRECATED_ROUTE,
  chunkSCY53JIH_cjs.STREAM_UI_MESSAGE_VNEXT_DEPRECATED_ROUTE,
  chunkSCY53JIH_cjs.STREAM_UI_MESSAGE_DEPRECATED_ROUTE
];

// src/server/server-adapter/routes/background-tasks.ts
var BACKGROUND_TASK_ROUTES = [
  chunk6QZLGOUC_cjs.BACKGROUND_TASK_STREAM_ROUTE,
  chunk6QZLGOUC_cjs.LIST_BACKGROUND_TASKS_ROUTE,
  chunk6QZLGOUC_cjs.GET_BACKGROUND_TASK_ROUTE
];

// src/server/server-adapter/routes/channels.ts
var CHANNELS_ROUTES = [
  chunkSXMPMMTX_cjs.LIST_CHANNEL_PLATFORMS_ROUTE,
  chunkSXMPMMTX_cjs.LIST_CHANNEL_INSTALLATIONS_ROUTE,
  chunkSXMPMMTX_cjs.CONNECT_CHANNEL_ROUTE,
  chunkSXMPMMTX_cjs.DISCONNECT_CHANNEL_ROUTE
];

// src/server/server-adapter/routes/conversations.ts
var CONVERSATIONS_ROUTES = [
  chunkKN22ZEKU_cjs.CREATE_CONVERSATION_ROUTE,
  chunkKN22ZEKU_cjs.GET_CONVERSATION_ROUTE,
  chunkKN22ZEKU_cjs.GET_CONVERSATION_ITEMS_ROUTE,
  chunkKN22ZEKU_cjs.DELETE_CONVERSATION_ROUTE
];

// src/server/server-adapter/routes/datasets.ts
var DATASETS_ROUTES = [
  // Dataset CRUD
  chunkMTWCBGFH_cjs.LIST_DATASETS_ROUTE,
  chunkMTWCBGFH_cjs.CREATE_DATASET_ROUTE,
  chunkMTWCBGFH_cjs.GET_DATASET_ROUTE,
  chunkMTWCBGFH_cjs.UPDATE_DATASET_ROUTE,
  chunkMTWCBGFH_cjs.DELETE_DATASET_ROUTE,
  // Item list and add
  chunkMTWCBGFH_cjs.LIST_ITEMS_ROUTE,
  chunkMTWCBGFH_cjs.ADD_ITEM_ROUTE,
  // Batch operations - MUST come before item-specific routes to avoid /items/batch matching /items/:itemId
  chunkMTWCBGFH_cjs.BATCH_INSERT_ITEMS_ROUTE,
  chunkMTWCBGFH_cjs.BATCH_DELETE_ITEMS_ROUTE,
  // Item-specific CRUD (uses :itemId param)
  chunkMTWCBGFH_cjs.GET_ITEM_ROUTE,
  chunkMTWCBGFH_cjs.UPDATE_ITEM_ROUTE,
  chunkMTWCBGFH_cjs.DELETE_ITEM_ROUTE,
  // Version operations
  chunkMTWCBGFH_cjs.LIST_DATASET_VERSIONS_ROUTE,
  chunkMTWCBGFH_cjs.LIST_ITEM_VERSIONS_ROUTE,
  chunkMTWCBGFH_cjs.GET_ITEM_VERSION_ROUTE,
  // All experiments (cross-dataset) - MUST come before dataset-scoped experiments
  chunkMTWCBGFH_cjs.LIST_ALL_EXPERIMENTS_ROUTE,
  // Experiment review summary (cross-dataset) - MUST come before dataset-scoped experiments
  chunkMTWCBGFH_cjs.EXPERIMENT_REVIEW_SUMMARY_ROUTE,
  // Experiment operations (dataset-scoped)
  chunkMTWCBGFH_cjs.LIST_EXPERIMENTS_ROUTE,
  chunkMTWCBGFH_cjs.TRIGGER_EXPERIMENT_ROUTE,
  chunkMTWCBGFH_cjs.GET_EXPERIMENT_ROUTE,
  chunkMTWCBGFH_cjs.LIST_EXPERIMENT_RESULTS_ROUTE,
  chunkMTWCBGFH_cjs.UPDATE_EXPERIMENT_RESULT_ROUTE,
  // Analytics
  chunkMTWCBGFH_cjs.COMPARE_EXPERIMENTS_ROUTE,
  // AI generation
  chunkMTWCBGFH_cjs.GENERATE_ITEMS_ROUTE,
  // Failure analysis
  chunkMTWCBGFH_cjs.CLUSTER_FAILURES_ROUTE
];

// src/server/server-adapter/routes/editor-builder.ts
var EDITOR_BUILDER_ROUTES = [
  chunkMJH4EEUV_cjs.GET_EDITOR_BUILDER_SETTINGS_ROUTE,
  chunkMJH4EEUV_cjs.GET_EDITOR_BUILDER_AVAILABLE_MODELS_ROUTE,
  chunkMJH4EEUV_cjs.GET_INFRASTRUCTURE_STATUS_ROUTE,
  chunkLAY7G3RT_cjs.LIST_BUILDER_REGISTRIES_ROUTE,
  chunkLAY7G3RT_cjs.BUILDER_REGISTRY_SEARCH_ROUTE,
  chunkLAY7G3RT_cjs.BUILDER_REGISTRY_POPULAR_ROUTE,
  chunkLAY7G3RT_cjs.BUILDER_REGISTRY_PREVIEW_ROUTE,
  chunkLAY7G3RT_cjs.BUILDER_REGISTRY_INSTALL_ROUTE
];

// src/server/server-adapter/routes/heartbeats.ts
var HEARTBEATS_ROUTES = [
  chunk55BEULMJ_cjs.LIST_HEARTBEATS_ROUTE,
  chunk55BEULMJ_cjs.GET_HEARTBEAT_ROUTE,
  chunk55BEULMJ_cjs.CREATE_HEARTBEAT_ROUTE,
  chunk55BEULMJ_cjs.UPDATE_HEARTBEAT_ROUTE,
  chunk55BEULMJ_cjs.DELETE_HEARTBEAT_ROUTE,
  chunk55BEULMJ_cjs.PAUSE_HEARTBEAT_ROUTE,
  chunk55BEULMJ_cjs.RESUME_HEARTBEAT_ROUTE,
  chunk55BEULMJ_cjs.RUN_HEARTBEAT_ROUTE
];

// src/server/server-adapter/routes/legacy.ts
var LEGACY_ROUTES = [
  // ============================================================================
  // Legacy Agent Routes
  // ============================================================================
  chunkSCY53JIH_cjs.GENERATE_LEGACY_ROUTE,
  chunkSCY53JIH_cjs.STREAM_GENERATE_LEGACY_ROUTE,
  // ============================================================================
  // Legacy Workflow Routes
  // ============================================================================
  chunkOE5GVCW4_cjs.STREAM_LEGACY_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.OBSERVE_STREAM_LEGACY_WORKFLOW_ROUTE
];

// src/server/server-adapter/routes/logs.ts
var LOGS_ROUTES = [chunk5O3XY7Z3_cjs.LIST_LOG_TRANSPORTS_ROUTE, chunk5O3XY7Z3_cjs.LIST_LOGS_ROUTE, chunk5O3XY7Z3_cjs.LIST_LOGS_BY_RUN_ID_ROUTE];

// src/server/server-adapter/routes/mcp.ts
var MCP_ROUTES = [
  // ============================================================================
  // MCP Server Registry Routes
  // ============================================================================
  chunkUXYPH6HP_cjs.LIST_MCP_SERVERS_ROUTE,
  chunkUXYPH6HP_cjs.GET_MCP_SERVER_DETAIL_ROUTE,
  // ============================================================================
  // MCP Server Tool Routes
  // ============================================================================
  chunkUXYPH6HP_cjs.LIST_MCP_SERVER_TOOLS_ROUTE,
  chunkUXYPH6HP_cjs.GET_MCP_SERVER_TOOL_DETAIL_ROUTE,
  chunkUXYPH6HP_cjs.EXECUTE_MCP_SERVER_TOOL_ROUTE,
  // ============================================================================
  // MCP Server Resource Routes
  // ============================================================================
  chunkUXYPH6HP_cjs.LIST_MCP_SERVER_RESOURCES_ROUTE,
  chunkUXYPH6HP_cjs.READ_MCP_SERVER_RESOURCE_ROUTE,
  // ============================================================================
  // MCP Transport Routes (handled by adapters)
  // ============================================================================
  chunkUXYPH6HP_cjs.MCP_HTTP_TRANSPORT_ROUTE,
  chunkUXYPH6HP_cjs.MCP_SSE_TRANSPORT_ROUTE,
  chunkUXYPH6HP_cjs.MCP_SSE_MESSAGES_ROUTE
];

// src/server/server-adapter/routes/memory.ts
var MEMORY_ROUTES = [
  chunkLVGS5ADS_cjs.GET_MEMORY_STATUS_ROUTE,
  chunkLVGS5ADS_cjs.GET_MEMORY_CONFIG_ROUTE,
  chunkLVGS5ADS_cjs.GET_OBSERVATIONAL_MEMORY_ROUTE,
  chunkLVGS5ADS_cjs.AWAIT_BUFFER_STATUS_ROUTE,
  chunkLVGS5ADS_cjs.LIST_THREADS_ROUTE,
  chunkLVGS5ADS_cjs.GET_THREAD_BY_ID_ROUTE,
  chunkLVGS5ADS_cjs.LIST_MESSAGES_ROUTE,
  chunkLVGS5ADS_cjs.GET_WORKING_MEMORY_ROUTE,
  chunkLVGS5ADS_cjs.SAVE_MESSAGES_ROUTE,
  chunkLVGS5ADS_cjs.CREATE_THREAD_ROUTE,
  chunkLVGS5ADS_cjs.UPDATE_THREAD_ROUTE,
  chunkLVGS5ADS_cjs.DELETE_THREAD_ROUTE,
  chunkLVGS5ADS_cjs.CLONE_THREAD_ROUTE,
  chunkLVGS5ADS_cjs.UPDATE_WORKING_MEMORY_ROUTE,
  chunkLVGS5ADS_cjs.DELETE_MESSAGES_ROUTE,
  chunkLVGS5ADS_cjs.SEARCH_MEMORY_ROUTE,
  chunkLVGS5ADS_cjs.GET_MEMORY_STATUS_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.LIST_THREADS_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.GET_THREAD_BY_ID_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.LIST_MESSAGES_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.SAVE_MESSAGES_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.CREATE_THREAD_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.UPDATE_THREAD_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.DELETE_THREAD_NETWORK_ROUTE,
  chunkLVGS5ADS_cjs.DELETE_MESSAGES_NETWORK_ROUTE
];

// src/server/server-adapter/routes/observability.ts
var OBSERVABILITY_ROUTES = [
  // Legacy
  chunkY32XM46K_cjs.LIST_TRACES_ROUTE,
  chunkY32XM46K_cjs.LIST_TRACES_LIGHT_ROUTE,
  chunkY32XM46K_cjs.LIST_BRANCHES_ROUTE,
  chunkY32XM46K_cjs.GET_BRANCH_ROUTE,
  chunkY32XM46K_cjs.GET_TRACE_ROUTE,
  chunkY32XM46K_cjs.GET_TRACE_LIGHT_ROUTE,
  chunkY32XM46K_cjs.GET_SPAN_ROUTE,
  chunkY32XM46K_cjs.GET_TRACE_TRAJECTORY_ROUTE,
  chunkY32XM46K_cjs.SCORE_TRACES_ROUTE,
  chunkY32XM46K_cjs.LIST_SCORES_BY_SPAN_ROUTE,
  // New observability routes
  chunk5GQWTNAY_cjs.LIST_METRICS,
  chunk5GQWTNAY_cjs.LIST_LOGS,
  chunk5GQWTNAY_cjs.LIST_SCORES,
  chunk5GQWTNAY_cjs.CREATE_SCORE,
  chunk5GQWTNAY_cjs.GET_SCORE,
  chunk5GQWTNAY_cjs.GET_SCORE_AGGREGATE,
  chunk5GQWTNAY_cjs.GET_SCORE_BREAKDOWN,
  chunk5GQWTNAY_cjs.GET_SCORE_TIME_SERIES,
  chunk5GQWTNAY_cjs.GET_SCORE_PERCENTILES,
  chunk5GQWTNAY_cjs.LIST_FEEDBACK,
  chunk5GQWTNAY_cjs.CREATE_FEEDBACK,
  chunk5GQWTNAY_cjs.GET_FEEDBACK_AGGREGATE,
  chunk5GQWTNAY_cjs.GET_FEEDBACK_BREAKDOWN,
  chunk5GQWTNAY_cjs.GET_FEEDBACK_TIME_SERIES,
  chunk5GQWTNAY_cjs.GET_FEEDBACK_PERCENTILES,
  chunk5GQWTNAY_cjs.GET_METRIC_AGGREGATE,
  chunk5GQWTNAY_cjs.GET_METRIC_BREAKDOWN,
  chunk5GQWTNAY_cjs.GET_METRIC_TIME_SERIES,
  chunk5GQWTNAY_cjs.GET_METRIC_PERCENTILES,
  chunk5GQWTNAY_cjs.GET_METRIC_NAMES,
  chunk5GQWTNAY_cjs.GET_METRIC_LABEL_KEYS,
  chunk5GQWTNAY_cjs.GET_METRIC_LABEL_VALUES,
  chunk5GQWTNAY_cjs.GET_ENTITY_TYPES,
  chunk5GQWTNAY_cjs.GET_ENTITY_NAMES,
  chunk5GQWTNAY_cjs.GET_SERVICE_NAMES,
  chunk5GQWTNAY_cjs.GET_ENVIRONMENTS,
  chunk5GQWTNAY_cjs.GET_TAGS
];

// src/server/server-adapter/routes/processor-providers.ts
var PROCESSOR_PROVIDER_ROUTES = [chunkJYWPYMQS_cjs.LIST_PROCESSOR_PROVIDERS_ROUTE, chunkJYWPYMQS_cjs.GET_PROCESSOR_PROVIDER_ROUTE];

// src/server/server-adapter/routes/processors.ts
var PROCESSORS_ROUTES = [chunkXPJZ72ZH_cjs.LIST_PROCESSORS_ROUTE, chunkXPJZ72ZH_cjs.GET_PROCESSOR_BY_ID_ROUTE, chunkXPJZ72ZH_cjs.EXECUTE_PROCESSOR_ROUTE];

// src/server/server-adapter/routes/responses.ts
var RESPONSES_ROUTES = [chunkOJB6MLRD_cjs.CREATE_RESPONSE_ROUTE, chunkOJB6MLRD_cjs.GET_RESPONSE_ROUTE, chunkOJB6MLRD_cjs.DELETE_RESPONSE_ROUTE];

// src/server/server-adapter/routes/schedules.ts
var SCHEDULES_ROUTES = [
  chunkJ6XQWEEV_cjs.LIST_SCHEDULES_ROUTE,
  chunkJ6XQWEEV_cjs.GET_SCHEDULE_ROUTE,
  chunkJ6XQWEEV_cjs.LIST_SCHEDULE_TRIGGERS_ROUTE,
  chunkJ6XQWEEV_cjs.PAUSE_SCHEDULE_ROUTE,
  chunkJ6XQWEEV_cjs.RESUME_SCHEDULE_ROUTE
];

// src/server/server-adapter/routes/scorers.ts
var SCORES_ROUTES = [
  chunkI4X4UHQW_cjs.LIST_SCORERS_ROUTE,
  chunkI4X4UHQW_cjs.GET_SCORER_ROUTE,
  chunkI4X4UHQW_cjs.LIST_SCORES_BY_RUN_ID_ROUTE,
  chunkI4X4UHQW_cjs.LIST_SCORES_BY_SCORER_ID_ROUTE,
  chunkI4X4UHQW_cjs.LIST_SCORES_BY_ENTITY_ID_ROUTE,
  chunkI4X4UHQW_cjs.SAVE_SCORE_ROUTE
];

// src/server/server-adapter/routes/stored-agents.ts
var STORED_AGENTS_ROUTES = [
  // ============================================================================
  // Stored Agents CRUD Routes
  // IMPORTANT: Routes with literal paths (e.g., /preview-instructions) must come
  // BEFORE routes with path parameters (e.g., /:storedAgentId) to ensure correct matching.
  // ============================================================================
  chunk6QQ4I6IY_cjs.LIST_STORED_AGENTS_ROUTE,
  chunk6QQ4I6IY_cjs.PREVIEW_INSTRUCTIONS_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE
  chunk6QQ4I6IY_cjs.GET_STORED_AGENT_DEPENDENTS_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE (longer literal)
  chunk6QQ4I6IY_cjs.EXPORT_STORED_AGENT_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE
  chunk6QQ4I6IY_cjs.OPEN_STORED_AGENT_CHANGE_REQUEST_ROUTE,
  // Must be before GET_STORED_AGENT_ROUTE
  chunk6QQ4I6IY_cjs.GET_STORED_AGENT_ROUTE,
  chunk6QQ4I6IY_cjs.CREATE_STORED_AGENT_ROUTE,
  chunk6QQ4I6IY_cjs.UPDATE_STORED_AGENT_ROUTE,
  chunk6QQ4I6IY_cjs.DELETE_STORED_AGENT_ROUTE,
  // ============================================================================
  // Agent Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  // ============================================================================
  chunkAR5PBVYY_cjs.LIST_AGENT_VERSIONS_ROUTE,
  chunkAR5PBVYY_cjs.CREATE_AGENT_VERSION_ROUTE,
  chunkAR5PBVYY_cjs.COMPARE_AGENT_VERSIONS_ROUTE,
  // Must be before GET_AGENT_VERSION_ROUTE
  chunkAR5PBVYY_cjs.GET_AGENT_VERSION_ROUTE,
  chunkAR5PBVYY_cjs.ACTIVATE_AGENT_VERSION_ROUTE,
  chunkAR5PBVYY_cjs.RESTORE_AGENT_VERSION_ROUTE,
  chunkAR5PBVYY_cjs.DELETE_AGENT_VERSION_ROUTE,
  // ============================================================================
  // Favorites (EE)
  // ============================================================================
  chunkLRRO5PBP_cjs.FAVORITE_STORED_AGENT_ROUTE,
  chunkLRRO5PBP_cjs.UNFAVORITE_STORED_AGENT_ROUTE
];

// src/server/server-adapter/routes/stored-mcp-clients.ts
var STORED_MCP_CLIENTS_ROUTES = [
  // Stored MCP Clients CRUD Routes
  chunkIQZE4X6U_cjs.LIST_STORED_MCP_CLIENTS_ROUTE,
  chunkIQZE4X6U_cjs.GET_STORED_MCP_CLIENT_ROUTE,
  chunkIQZE4X6U_cjs.CREATE_STORED_MCP_CLIENT_ROUTE,
  chunkIQZE4X6U_cjs.UPDATE_STORED_MCP_CLIENT_ROUTE,
  chunkIQZE4X6U_cjs.DELETE_STORED_MCP_CLIENT_ROUTE,
  // MCP Client Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  chunkFOANUZQ3_cjs.LIST_MCP_CLIENT_VERSIONS_ROUTE,
  chunkFOANUZQ3_cjs.CREATE_MCP_CLIENT_VERSION_ROUTE,
  chunkFOANUZQ3_cjs.COMPARE_MCP_CLIENT_VERSIONS_ROUTE,
  chunkFOANUZQ3_cjs.GET_MCP_CLIENT_VERSION_ROUTE,
  chunkFOANUZQ3_cjs.ACTIVATE_MCP_CLIENT_VERSION_ROUTE,
  chunkFOANUZQ3_cjs.RESTORE_MCP_CLIENT_VERSION_ROUTE,
  chunkFOANUZQ3_cjs.DELETE_MCP_CLIENT_VERSION_ROUTE
];

// src/server/server-adapter/routes/stored-prompt-blocks.ts
var STORED_PROMPT_BLOCKS_ROUTES = [
  // Stored Prompt Blocks CRUD Routes
  chunkEMQZIPXG_cjs.LIST_STORED_PROMPT_BLOCKS_ROUTE,
  chunkEMQZIPXG_cjs.GET_STORED_PROMPT_BLOCK_ROUTE,
  chunkEMQZIPXG_cjs.CREATE_STORED_PROMPT_BLOCK_ROUTE,
  chunkEMQZIPXG_cjs.UPDATE_STORED_PROMPT_BLOCK_ROUTE,
  chunkEMQZIPXG_cjs.DELETE_STORED_PROMPT_BLOCK_ROUTE,
  // Prompt Block Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  chunkQ5CGBHAI_cjs.LIST_PROMPT_BLOCK_VERSIONS_ROUTE,
  chunkQ5CGBHAI_cjs.CREATE_PROMPT_BLOCK_VERSION_ROUTE,
  chunkQ5CGBHAI_cjs.COMPARE_PROMPT_BLOCK_VERSIONS_ROUTE,
  chunkQ5CGBHAI_cjs.GET_PROMPT_BLOCK_VERSION_ROUTE,
  chunkQ5CGBHAI_cjs.ACTIVATE_PROMPT_BLOCK_VERSION_ROUTE,
  chunkQ5CGBHAI_cjs.RESTORE_PROMPT_BLOCK_VERSION_ROUTE,
  chunkQ5CGBHAI_cjs.DELETE_PROMPT_BLOCK_VERSION_ROUTE
];

// src/server/server-adapter/routes/stored-scorers.ts
var STORED_SCORERS_ROUTES = [
  // Stored Scorers CRUD Routes
  chunkKE32UGIC_cjs.LIST_STORED_SCORERS_ROUTE,
  chunkKE32UGIC_cjs.GET_STORED_SCORER_ROUTE,
  chunkKE32UGIC_cjs.CREATE_STORED_SCORER_ROUTE,
  chunkKE32UGIC_cjs.UPDATE_STORED_SCORER_ROUTE,
  chunkKE32UGIC_cjs.DELETE_STORED_SCORER_ROUTE,
  // Scorer Versions Routes
  // IMPORTANT: Routes with literal paths (e.g., /compare) must come BEFORE
  // routes with path parameters (e.g., /:versionId) to ensure correct matching.
  chunkZGAUP346_cjs.LIST_SCORER_VERSIONS_ROUTE,
  chunkZGAUP346_cjs.CREATE_SCORER_VERSION_ROUTE,
  chunkZGAUP346_cjs.COMPARE_SCORER_VERSIONS_ROUTE,
  chunkZGAUP346_cjs.GET_SCORER_VERSION_ROUTE,
  chunkZGAUP346_cjs.ACTIVATE_SCORER_VERSION_ROUTE,
  chunkZGAUP346_cjs.RESTORE_SCORER_VERSION_ROUTE,
  chunkZGAUP346_cjs.DELETE_SCORER_VERSION_ROUTE
];

// src/server/server-adapter/routes/stored-skills.ts
var STORED_SKILLS_ROUTES = [
  // Stored Skills CRUD Routes
  chunkSORUUFZK_cjs.LIST_STORED_SKILLS_ROUTE,
  chunkSORUUFZK_cjs.GET_STORED_SKILL_ROUTE,
  chunkSORUUFZK_cjs.CREATE_STORED_SKILL_ROUTE,
  chunkSORUUFZK_cjs.UPDATE_STORED_SKILL_ROUTE,
  chunkSORUUFZK_cjs.DELETE_STORED_SKILL_ROUTE,
  // Publish
  chunkSORUUFZK_cjs.PUBLISH_STORED_SKILL_ROUTE,
  // Favorites (EE)
  chunk3EZOV6A3_cjs.FAVORITE_STORED_SKILL_ROUTE,
  chunk3EZOV6A3_cjs.UNFAVORITE_STORED_SKILL_ROUTE
];

// src/server/server-adapter/routes/stored-workspaces.ts
var STORED_WORKSPACES_ROUTES = [
  // Stored Workspaces CRUD Routes
  chunkEG6DDVIJ_cjs.LIST_STORED_WORKSPACES_ROUTE,
  chunkEG6DDVIJ_cjs.GET_STORED_WORKSPACE_ROUTE,
  chunkEG6DDVIJ_cjs.CREATE_STORED_WORKSPACE_ROUTE,
  chunkEG6DDVIJ_cjs.UPDATE_STORED_WORKSPACE_ROUTE,
  chunkEG6DDVIJ_cjs.DELETE_STORED_WORKSPACE_ROUTE
];

// src/server/server-adapter/routes/system.ts
var SYSTEM_ROUTES = [chunkH7NWUAHC_cjs.GET_SYSTEM_PACKAGES_ROUTE, chunkH7NWUAHC_cjs.GET_API_SCHEMA_ROUTE];

// src/server/server-adapter/routes/tool-providers.ts
var TOOL_PROVIDER_ROUTES = [
  chunkYDUA7L5F_cjs.LIST_TOOL_PROVIDERS_ROUTE,
  chunkYDUA7L5F_cjs.LIST_TOOL_PROVIDER_TOOLKITS_ROUTE,
  chunkYDUA7L5F_cjs.LIST_TOOL_PROVIDER_TOOLS_ROUTE,
  chunkYDUA7L5F_cjs.GET_TOOL_PROVIDER_TOOL_SCHEMA_ROUTE,
  chunkYDUA7L5F_cjs.AUTHORIZE_TOOL_PROVIDER_ROUTE,
  chunkYDUA7L5F_cjs.GET_TOOL_PROVIDER_AUTH_STATUS_ROUTE,
  chunkYDUA7L5F_cjs.TOOL_PROVIDER_CONNECTION_STATUS_ROUTE,
  chunkYDUA7L5F_cjs.LIST_TOOL_PROVIDER_CONNECTIONS_ROUTE,
  chunkYDUA7L5F_cjs.LIST_TOOL_PROVIDER_CONNECTION_FIELDS_ROUTE,
  chunkYDUA7L5F_cjs.DISCONNECT_TOOL_PROVIDER_CONNECTION_ROUTE,
  chunkYDUA7L5F_cjs.UPDATE_TOOL_PROVIDER_CONNECTION_ROUTE,
  chunkYDUA7L5F_cjs.GET_TOOL_PROVIDER_CONNECTION_USAGE_ROUTE,
  chunkYDUA7L5F_cjs.GET_TOOL_PROVIDER_HEALTH_ROUTE
];

// src/server/server-adapter/routes/tools.ts
var TOOLS_ROUTES = [chunkYRZYCQGE_cjs.LIST_TOOLS_ROUTE, chunkYRZYCQGE_cjs.GET_TOOL_BY_ID_ROUTE, chunkYRZYCQGE_cjs.EXECUTE_TOOL_ROUTE];

// src/server/server-adapter/routes/vectors.ts
var VECTORS_ROUTES = [
  chunkQPICFOIK_cjs.UPSERT_VECTORS_ROUTE,
  chunkQPICFOIK_cjs.CREATE_INDEX_ROUTE,
  chunkQPICFOIK_cjs.QUERY_VECTORS_ROUTE,
  chunkQPICFOIK_cjs.LIST_INDEXES_ROUTE,
  chunkQPICFOIK_cjs.DESCRIBE_INDEX_ROUTE,
  chunkQPICFOIK_cjs.DELETE_INDEX_ROUTE,
  chunkQPICFOIK_cjs.LIST_VECTORS_ROUTE,
  chunkQPICFOIK_cjs.LIST_EMBEDDERS_ROUTE
];

// src/server/server-adapter/routes/workflows.ts
var WORKFLOWS_ROUTES = [
  chunkOE5GVCW4_cjs.LIST_WORKFLOWS_ROUTE,
  chunkOE5GVCW4_cjs.GET_WORKFLOW_BY_ID_ROUTE,
  chunkOE5GVCW4_cjs.LIST_WORKFLOW_RUNS_ROUTE,
  chunkOE5GVCW4_cjs.GET_WORKFLOW_RUN_BY_ID_ROUTE,
  chunkOE5GVCW4_cjs.DELETE_WORKFLOW_RUN_BY_ID_ROUTE,
  chunkOE5GVCW4_cjs.CREATE_WORKFLOW_RUN_ROUTE,
  chunkOE5GVCW4_cjs.STREAM_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESUME_STREAM_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.START_ASYNC_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.START_WORKFLOW_RUN_ROUTE,
  chunkOE5GVCW4_cjs.OBSERVE_STREAM_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESUME_ASYNC_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESUME_NO_WAIT_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESUME_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.CANCEL_WORKFLOW_RUN_ROUTE,
  chunkOE5GVCW4_cjs.TIME_TRAVEL_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.TIME_TRAVEL_ASYNC_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.TIME_TRAVEL_STREAM_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESTART_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESTART_ASYNC_WORKFLOW_ROUTE,
  chunkOE5GVCW4_cjs.RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ROUTE,
  chunkOE5GVCW4_cjs.RESTART_ALL_ACTIVE_WORKFLOW_RUNS_ASYNC_ROUTE,
  chunkOE5GVCW4_cjs.EXECUTE_WORKFLOW_STEP_ROUTE,
  chunkOE5GVCW4_cjs.RECEIVE_WORKFLOW_EVENT_ROUTE
];

// src/server/server-adapter/routes/workspace.ts
var WORKSPACE_ROUTES = [
  // List all workspaces route (at /api/workspaces)
  chunkDR2XLMGB_cjs.LIST_WORKSPACES_ROUTE,
  // Get workspace route (at /api/workspaces/:workspaceId)
  chunkDR2XLMGB_cjs.GET_WORKSPACE_ROUTE,
  // Filesystem routes (at /api/workspaces/:workspaceId/fs/*)
  ...chunkDR2XLMGB_cjs.WORKSPACE_FS_ROUTES,
  // Search routes (at /api/workspaces/:workspaceId/search, /api/workspaces/:workspaceId/index)
  ...chunkDR2XLMGB_cjs.WORKSPACE_SEARCH_ROUTES,
  // Skills routes (search must come before parameterized routes)
  ...chunkDR2XLMGB_cjs.WORKSPACE_SKILLS_ROUTES,
  // skills.sh proxy routes (at /api/workspaces/:workspaceId/skills-sh/*)
  ...chunkDR2XLMGB_cjs.WORKSPACE_SKILLS_SH_ROUTES
];

// src/server/server-adapter/routes/index.ts
var SERVER_ROUTES = [
  ...AGENTS_ROUTES,
  ...chunkAJ4VAZFT_cjs.AUTH_ROUTES,
  ...WORKFLOWS_ROUTES,
  ...TOOLS_ROUTES,
  ...PROCESSORS_ROUTES,
  ...RESPONSES_ROUTES,
  ...CONVERSATIONS_ROUTES,
  ...MEMORY_ROUTES,
  ...SCORES_ROUTES,
  ...OBSERVABILITY_ROUTES,
  ...LOGS_ROUTES,
  ...VECTORS_ROUTES,
  ...A2A_ROUTES,
  ...WORKSPACE_ROUTES,
  ...LEGACY_ROUTES,
  ...MCP_ROUTES,
  ...STORED_AGENTS_ROUTES,
  ...STORED_MCP_CLIENTS_ROUTES,
  ...STORED_PROMPT_BLOCKS_ROUTES,
  ...STORED_SCORERS_ROUTES,
  ...STORED_WORKSPACES_ROUTES,
  ...STORED_SKILLS_ROUTES,
  ...TOOL_PROVIDER_ROUTES,
  ...PROCESSOR_PROVIDER_ROUTES,
  ...SYSTEM_ROUTES,
  ...DATASETS_ROUTES,
  ...BACKGROUND_TASK_ROUTES,
  ...EDITOR_BUILDER_ROUTES,
  ...AGENT_BUILDER_ROUTES,
  ...SCHEDULES_ROUTES,
  ...HEARTBEATS_ROUTES,
  ...CHANNELS_ROUTES,
  ...AGENT_CONTROLLER_ROUTES
];

exports.SERVER_ROUTES = SERVER_ROUTES;
exports.deriveAction = deriveAction;
exports.derivePermission = derivePermission;
exports.extractResource = extractResource;
exports.getEffectivePermission = getEffectivePermission;
