// @mastra/server/dist/chunk-4HFIGLPA.js
import { z } from 'zod';

// src/server/schemas/schedules.ts
var scheduleStatusSchema = z.enum(["active", "paused"]);
var workflowScheduleTargetSchema = z.object({
  type: z.literal("workflow"),
  workflowId: z.string(),
  inputData: z.unknown().optional(),
  initialState: z.unknown().optional(),
  requestContext: z.record(z.string(), z.unknown()).optional()
});
var signalAttributesSchema = z.record(
  z.string(),
  z.union([z.string(), z.number(), z.boolean(), z.null()]).optional()
);
var ifActiveSchema = z.object({
  behavior: z.enum(["deliver", "persist", "discard"]).optional(),
  attributes: signalAttributesSchema.optional()
});
var ifIdleSchema = z.object({
  behavior: z.enum(["wake", "persist", "discard"]).optional(),
  attributes: signalAttributesSchema.optional(),
  streamOptions: z.object({
    requestContext: z.record(z.string(), z.unknown()).optional()
  }).optional()
});
var heartbeatScheduleTargetSchema = z.object({
  type: z.literal("heartbeat"),
  agentId: z.string(),
  prompt: z.string(),
  threadId: z.string().optional(),
  resourceId: z.string().optional(),
  signalType: z.string().optional(),
  tagName: z.string().optional(),
  attributes: signalAttributesSchema.optional(),
  ifActive: ifActiveSchema.optional(),
  ifIdle: ifIdleSchema.optional(),
  providerOptions: z.record(z.string(), z.unknown()).optional(),
  requestContext: z.record(z.string(), z.unknown()).optional()
});
var scheduleTargetSchema = z.discriminatedUnion("type", [
  workflowScheduleTargetSchema,
  heartbeatScheduleTargetSchema
]);
var workflowRunStatusSchema = z.enum([
  "running",
  "success",
  "failed",
  "tripwire",
  "suspended",
  "waiting",
  "pending",
  "canceled",
  "bailed",
  "paused",
  "skipped"
]);
var scheduleRunSummarySchema = z.object({
  status: workflowRunStatusSchema,
  startedAt: z.number().optional(),
  completedAt: z.number().optional(),
  durationMs: z.number().optional(),
  error: z.string().optional()
});
var scheduleResponseSchema = z.object({
  id: z.string(),
  target: scheduleTargetSchema,
  cron: z.string(),
  timezone: z.string().optional(),
  status: scheduleStatusSchema,
  nextFireAt: z.number(),
  lastFireAt: z.number().optional(),
  lastRunId: z.string().optional(),
  lastRun: scheduleRunSummarySchema.optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
  ownerType: z.string().optional(),
  ownerId: z.string().optional(),
  createdAt: z.number(),
  updatedAt: z.number()
});
var scheduleTriggerOutcomeSchema = z.enum([
  "published",
  "succeeded",
  "delivered",
  "persisted",
  "discarded",
  "skipped",
  "aborted",
  "failed",
  // Legacy queue/notification outcomes — no longer written, but trigger rows
  // persisted by older builds may still carry them. Kept readable so the
  // response validator does not reject historical rows. Mirrors the core
  // ScheduleTriggerOutcome union.
  "acked",
  "alerted",
  "deferred",
  "appended-from-queue",
  "dropped-stale",
  "dropped-superseded",
  "dropped-busy"
]);
var scheduleTriggerKindSchema = z.enum(["schedule-fire", "queue-drain", "manual"]);
var scheduleTriggerResponseSchema = z.object({
  id: z.string().optional(),
  scheduleId: z.string(),
  runId: z.string().nullable(),
  scheduledFireAt: z.number(),
  actualFireAt: z.number(),
  outcome: scheduleTriggerOutcomeSchema,
  error: z.string().optional(),
  triggerKind: scheduleTriggerKindSchema.optional(),
  parentTriggerId: z.string().optional(),
  metadata: z.record(z.string(), z.unknown()).optional(),
  run: scheduleRunSummarySchema.optional()
});
var listSchedulesQuerySchema = z.object({
  workflowId: z.string().optional(),
  status: scheduleStatusSchema.optional(),
  ownerType: z.string().optional(),
  ownerId: z.string().optional()
}).refine((data) => data.ownerId === void 0 || data.ownerType !== void 0, {
  message: "ownerId can only be used together with ownerType",
  path: ["ownerId"]
});
var listSchedulesResponseSchema = z.object({
  schedules: z.array(scheduleResponseSchema)
});
var scheduleIdPathParams = z.object({
  scheduleId: z.string()
});
var listScheduleTriggersQuerySchema = z.object({
  limit: z.coerce.number().int().positive().optional(),
  fromActualFireAt: z.coerce.number().int().nonnegative().optional(),
  toActualFireAt: z.coerce.number().int().nonnegative().optional()
});
var listScheduleTriggersResponseSchema = z.object({
  triggers: z.array(scheduleTriggerResponseSchema)
});

export { listScheduleTriggersQuerySchema, listScheduleTriggersResponseSchema, listSchedulesQuerySchema, listSchedulesResponseSchema, scheduleIdPathParams, scheduleResponseSchema, scheduleRunSummarySchema };
