// @mastra/server/dist/chunk-S6TNTR3A.cjs
'use strict';

var chunkPTPMBE6M_cjs = require('./chunk-PTPMBE6M.cjs');
var v4 = require('zod/v4');

var actionIdPathParams = v4.z.object({
  actionId: v4.z.string().describe("Unique identifier for the agent-builder action")
});
var actionRunPathParams = v4.z.object({
  actionId: v4.z.string().describe("Unique identifier for the agent-builder action"),
  runId: v4.z.string().describe("Unique identifier for the action run")
});
var streamAgentBuilderBodySchema = chunkPTPMBE6M_cjs.streamWorkflowBodySchema;
var streamLegacyAgentBuilderBodySchema = chunkPTPMBE6M_cjs.streamLegacyWorkflowBodySchema;
var resumeAgentBuilderBodySchema = chunkPTPMBE6M_cjs.resumeBodySchema;
var startAsyncAgentBuilderBodySchema = chunkPTPMBE6M_cjs.startAsyncWorkflowBodySchema;

exports.actionIdPathParams = actionIdPathParams;
exports.actionRunPathParams = actionRunPathParams;
exports.resumeAgentBuilderBodySchema = resumeAgentBuilderBodySchema;
exports.startAsyncAgentBuilderBodySchema = startAsyncAgentBuilderBodySchema;
exports.streamAgentBuilderBodySchema = streamAgentBuilderBodySchema;
exports.streamLegacyAgentBuilderBodySchema = streamLegacyAgentBuilderBodySchema;
