// @mastra/server/dist/observational-memory-PBLNZABH-MQEKJOVA.cjs
'use strict';

var chunk2WUMO4RN_cjs = require('./chunk-2WUMO4RN.cjs');
var chunk3D55EABW_cjs = require('./chunk-3D55EABW.cjs');



Object.defineProperty(exports, "Extractor", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.Extractor; }
});
Object.defineProperty(exports, "ModelByInputTokens", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.ModelByInputTokens; }
});
Object.defineProperty(exports, "OBSERVER_SYSTEM_PROMPT", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.OBSERVER_SYSTEM_PROMPT; }
});
Object.defineProperty(exports, "ObservationalMemory", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.ObservationalMemory; }
});
Object.defineProperty(exports, "ObservationalMemoryProcessor", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.ObservationalMemoryProcessor; }
});
Object.defineProperty(exports, "TokenCounter", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.TokenCounter; }
});
Object.defineProperty(exports, "WorkingMemoryExtractor", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.WorkingMemoryExtractor; }
});
Object.defineProperty(exports, "buildObserverPrompt", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.buildObserverPrompt; }
});
Object.defineProperty(exports, "buildObserverSystemPrompt", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.buildObserverSystemPrompt; }
});
Object.defineProperty(exports, "combineObservationGroupRanges", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.combineObservationGroupRanges; }
});
Object.defineProperty(exports, "deriveObservationGroupProvenance", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.deriveObservationGroupProvenance; }
});
Object.defineProperty(exports, "extractCurrentTask", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.extractCurrentTask; }
});
Object.defineProperty(exports, "formatMessagesForObserver", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.formatMessagesForObserver; }
});
Object.defineProperty(exports, "getObservationsAsOf", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.getObservationsAsOf; }
});
Object.defineProperty(exports, "hasCurrentTaskSection", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.hasCurrentTaskSection; }
});
Object.defineProperty(exports, "injectAnchorIds", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.injectAnchorIds; }
});
Object.defineProperty(exports, "optimizeObservationsForContext", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.optimizeObservationsForContext; }
});
Object.defineProperty(exports, "parseAnchorId", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.parseAnchorId; }
});
Object.defineProperty(exports, "parseObservationGroups", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.parseObservationGroups; }
});
Object.defineProperty(exports, "parseObserverOutput", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.parseObserverOutput; }
});
Object.defineProperty(exports, "reconcileObservationGroupsFromReflection", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.reconcileObservationGroupsFromReflection; }
});
Object.defineProperty(exports, "renderObservationGroupsForReflection", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.renderObservationGroupsForReflection; }
});
Object.defineProperty(exports, "stripEphemeralAnchorIds", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.stripEphemeralAnchorIds; }
});
Object.defineProperty(exports, "stripObservationGroups", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.stripObservationGroups; }
});
Object.defineProperty(exports, "wrapInObservationGroup", {
  enumerable: true,
  get: function () { return chunk2WUMO4RN_cjs.wrapInObservationGroup; }
});
Object.defineProperty(exports, "OBSERVATIONAL_MEMORY_DEFAULTS", {
  enumerable: true,
  get: function () { return chunk3D55EABW_cjs.OBSERVATIONAL_MEMORY_DEFAULTS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_INSTRUCTIONS", {
  enumerable: true,
  get: function () { return chunk3D55EABW_cjs.OBSERVATION_CONTEXT_INSTRUCTIONS; }
});
Object.defineProperty(exports, "OBSERVATION_CONTEXT_PROMPT", {
  enumerable: true,
  get: function () { return chunk3D55EABW_cjs.OBSERVATION_CONTEXT_PROMPT; }
});
Object.defineProperty(exports, "OBSERVATION_CONTINUATION_HINT", {
  enumerable: true,
  get: function () { return chunk3D55EABW_cjs.OBSERVATION_CONTINUATION_HINT; }
});
