// @mastra/core/dist/chunk-V7JI3TXI.js
import { PubSub, isLeaseProvider } from './chunk-2S76KRB5.js';

// src/events/caching-pubsub.ts
var CachingPubSub = class extends PubSub {
  constructor(inner, cache, options = {}) {
    super();
    this.inner = inner;
    this.cache = cache;
    this.keyPrefix = options.keyPrefix ?? "pubsub:";
    this.logger = options.logger;
  }
  inner;
  cache;
  keyPrefix;
  logger;
  /** Maps original callbacks to their wrapped versions for proper unsubscribe */
  callbackMap = /* @__PURE__ */ new Map();
  get supportsNativeBatching() {
    return this.inner.supportsNativeBatching;
  }
  /**
   * Log an error message using the configured logger or console.error.
   */
  logError(message, error) {
    if (this.logger) {
      this.logger.error(message, error);
    } else {
      console.error(message, error);
    }
  }
  /**
   * Stable key used to deduplicate an event across the cache-replay and
   * live-delivery paths.
   *
   * We cannot dedup on `event.id`: `CachingPubSub.publish` assigns the id and
   * caches the event with it, but inner PubSub implementations
   * (EventEmitterPubSub, UnixSocketPubSub, …) regenerate `id` inside their own
   * `publish`, so the cached copy and the live copy of the SAME publish carry
   * different ids. The sequential `index` is assigned here and is preserved by
   * every inner implementation, so it matches across both paths. Events without
   * an index are never cached (see `publish`), so they can't be replay/live
   * duplicated — falling back to `id` for them is safe.
   */
  dedupKey(event) {
    return event.index !== void 0 ? `i:${event.index}` : `id:${event.id}`;
  }
  /**
   * Get the cache key for a topic's event list
   */
  getCacheKey(topic) {
    return `${this.keyPrefix}${topic}`;
  }
  /**
   * Get the cache key for a topic's index counter
   */
  getCounterKey(topic) {
    return `${this.keyPrefix}${topic}:counter`;
  }
  /**
   * Publish an event to a topic.
   * The event is cached with a sequential index before being published to the inner PubSub.
   *
   * Uses atomic increment for index assignment to prevent race conditions
   * when multiple events are published concurrently.
   */
  async publish(topic, event, options) {
    const cacheKey = this.getCacheKey(topic);
    const counterKey = this.getCounterKey(topic);
    let index;
    let indexFailed = false;
    try {
      index = await this.cache.increment(counterKey) - 1;
    } catch (error) {
      this.logError(`[CachingPubSub] Failed to increment counter for ${topic}`, error);
      indexFailed = true;
    }
    const fullEvent = {
      ...event,
      id: crypto.randomUUID(),
      createdAt: /* @__PURE__ */ new Date(),
      ...index !== void 0 ? { index } : {}
    };
    if (!indexFailed) {
      try {
        await this.cache.listPush(cacheKey, fullEvent);
      } catch (error) {
        this.logError(`[CachingPubSub] Failed to cache event for ${topic}`, error);
      }
    }
    await this.inner.publish(topic, fullEvent, options);
  }
  /**
   * Subscribe to live events on a topic (no replay).
   */
  async subscribe(topic, cb, options) {
    await this.inner.subscribe(topic, cb, options);
  }
  /**
   * Subscribe to a topic with automatic replay of cached events.
   *
   * Order of operations:
   * 1. Subscribe to live events FIRST (to avoid missing events during replay)
   * 2. Fetch and replay cached history
   * 3. Deduplicate events at the boundary using event IDs
   *
   * Each subscriber gets its own deduplication set to ensure
   * multiple subscribers can independently receive all events.
   */
  async subscribeWithReplay(topic, cb) {
    let seen = /* @__PURE__ */ new Set();
    const wrappedCb = (event, ack) => {
      if (seen) {
        const key = this.dedupKey(event);
        if (!seen.has(key)) {
          seen.add(key);
          cb(event, ack);
        }
      } else {
        cb(event, ack);
      }
    };
    this.callbackMap.set(cb, wrappedCb);
    await this.inner.subscribe(topic, wrappedCb);
    const history = await this.getHistory(topic);
    for (const event of history) {
      const key = this.dedupKey(event);
      if (!seen.has(key)) {
        seen.add(key);
        cb(event);
      }
    }
    seen = null;
  }
  /**
   * Subscribe to a topic with replay starting from a specific index.
   * More efficient than full replay when the client knows their last position.
   *
   * @param topic - The topic to subscribe to
   * @param offset - Start replaying from this index (0-based)
   * @param cb - Callback invoked for each event
   */
  async subscribeFromOffset(topic, offset, cb) {
    let seen = /* @__PURE__ */ new Set();
    const wrappedCb = (event, ack) => {
      if (seen) {
        const key = this.dedupKey(event);
        if (!seen.has(key)) {
          seen.add(key);
          cb(event, ack);
        }
      } else {
        cb(event, ack);
      }
    };
    this.callbackMap.set(cb, wrappedCb);
    await this.inner.subscribe(topic, wrappedCb);
    const history = await this.getHistory(topic, offset);
    for (const event of history) {
      const key = this.dedupKey(event);
      if (!seen.has(key)) {
        seen.add(key);
        cb(event);
      }
    }
    seen = null;
  }
  /**
   * Unsubscribe from a topic.
   */
  async unsubscribe(topic, cb) {
    const wrappedCb = this.callbackMap.get(cb) ?? cb;
    this.callbackMap.delete(cb);
    await this.inner.unsubscribe(topic, wrappedCb);
  }
  /**
   * Get historical events for a topic from cache.
   */
  async getHistory(topic, offset = 0) {
    const cacheKey = this.getCacheKey(topic);
    const events = await this.cache.listFromTo(cacheKey, offset);
    return events;
  }
  /**
   * Flush any pending operations on the inner PubSub.
   */
  async flush() {
    await this.inner.flush();
  }
  /**
   * Expose the inner's {@link LeaseProvider} when it has one, otherwise
   * `undefined`. Leasing is a capability of the underlying backend
   * (e.g. Redis), not of the caching decorator itself — so rather than
   * unconditionally declaring lease methods (which would make
   * {@link isLeaseProvider} report `true` even when the inner can't
   * coordinate a lock), we surface the inner's capability directly. The
   * signals runtime unwraps this so wrapping with caching preserves real
   * distributed lease semantics without faking them.
   */
  getLeaseProvider() {
    return isLeaseProvider(this.inner) ? this.inner : void 0;
  }
  /**
   * Clear cached events for a specific topic.
   * Call this when a stream completes to free memory.
   * Also clears the index counter.
   */
  async clearTopic(topic) {
    const cacheKey = this.getCacheKey(topic);
    const counterKey = this.getCounterKey(topic);
    await Promise.all([this.cache.delete(cacheKey), this.cache.delete(counterKey)]);
  }
  /**
   * Get the inner PubSub instance.
   * Useful for accessing implementation-specific methods like close().
   */
  getInner() {
    return this.inner;
  }
};
function withCaching(pubsub, cache, options) {
  return new CachingPubSub(pubsub, cache, options);
}

export { CachingPubSub, withCaching };
