// @mastra/core/dist/chunk-OYARTFJQ.cjs
'use strict';

var chunk5FCWY4Z3_cjs = require('./chunk-5FCWY4Z3.cjs');
var chunkWFEGLWHX_cjs = require('./chunk-WFEGLWHX.cjs');
var chunkVIRV756L_cjs = require('./chunk-VIRV756L.cjs');
var chunkA45A2T7R_cjs = require('./chunk-A45A2T7R.cjs');
var chunk6QR4RQID_cjs = require('./chunk-6QR4RQID.cjs');
var chunkXSOONORA_cjs = require('./chunk-XSOONORA.cjs');
var v4 = require('zod/v4');
var WebSocket = require('ws');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var WebSocket__default = /*#__PURE__*/_interopDefault(WebSocket);

var openaiErrorDataSchema = v4.z.object({
  error: v4.z.object({
    message: v4.z.string(),
    // The additional information below is handled loosely to support
    // OpenAI-compatible providers that have slightly different error
    // responses:
    type: v4.z.string().nullish(),
    param: v4.z.any().nullish(),
    code: v4.z.union([v4.z.string(), v4.z.number()]).nullish()
  })
});
var openaiFailedResponseHandler = chunkA45A2T7R_cjs.createJsonErrorResponseHandler({
  errorSchema: openaiErrorDataSchema,
  errorToMessage: (data) => data.error.message
});
function getOpenAILanguageModelCapabilities(modelId) {
  const supportsFlexProcessing = modelId.startsWith("o3") || modelId.startsWith("o4-mini") || modelId.startsWith("gpt-5") && !modelId.startsWith("gpt-5-chat");
  const supportsPriorityProcessing = modelId.startsWith("gpt-4") || modelId.startsWith("gpt-5") && !modelId.startsWith("gpt-5-nano") && !modelId.startsWith("gpt-5-chat") && !modelId.startsWith("gpt-5.4-nano") || modelId.startsWith("o3") || modelId.startsWith("o4-mini");
  const isReasoningModel = !(modelId.startsWith("gpt-3") || modelId.startsWith("gpt-4") || modelId.startsWith("chatgpt-4o") || modelId.startsWith("gpt-5-chat"));
  const supportsNonReasoningParameters = modelId.startsWith("gpt-5.1") || modelId.startsWith("gpt-5.2") || modelId.startsWith("gpt-5.3") || modelId.startsWith("gpt-5.4");
  const systemMessageMode = isReasoningModel ? "developer" : "system";
  return {
    supportsFlexProcessing,
    supportsPriorityProcessing,
    isReasoningModel,
    systemMessageMode,
    supportsNonReasoningParameters
  };
}
function convertToOpenAIChatMessages({
  prompt,
  systemMessageMode = "system"
}) {
  const messages = [];
  const warnings = [];
  for (const { role, content } of prompt) {
    switch (role) {
      case "system": {
        switch (systemMessageMode) {
          case "system": {
            messages.push({ role: "system", content });
            break;
          }
          case "developer": {
            messages.push({ role: "developer", content });
            break;
          }
          case "remove": {
            warnings.push({
              type: "other",
              message: "system messages are removed for this model"
            });
            break;
          }
          default: {
            const _exhaustiveCheck = systemMessageMode;
            throw new Error(
              `Unsupported system message mode: ${_exhaustiveCheck}`
            );
          }
        }
        break;
      }
      case "user": {
        if (content.length === 1 && content[0].type === "text") {
          messages.push({ role: "user", content: content[0].text });
          break;
        }
        messages.push({
          role: "user",
          content: content.map((part, index) => {
            var _a, _b, _c;
            switch (part.type) {
              case "text": {
                return { type: "text", text: part.text };
              }
              case "file": {
                if (part.mediaType.startsWith("image/")) {
                  const mediaType = part.mediaType === "image/*" ? "image/jpeg" : part.mediaType;
                  return {
                    type: "image_url",
                    image_url: {
                      url: part.data instanceof URL ? part.data.toString() : `data:${mediaType};base64,${chunkA45A2T7R_cjs.convertToBase64(part.data)}`,
                      // OpenAI specific extension: image detail
                      detail: (_b = (_a = part.providerOptions) == null ? void 0 : _a.openai) == null ? void 0 : _b.imageDetail
                    }
                  };
                } else if (part.mediaType.startsWith("audio/")) {
                  if (part.data instanceof URL) {
                    throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
                      functionality: "audio file parts with URLs"
                    });
                  }
                  switch (part.mediaType) {
                    case "audio/wav": {
                      return {
                        type: "input_audio",
                        input_audio: {
                          data: chunkA45A2T7R_cjs.convertToBase64(part.data),
                          format: "wav"
                        }
                      };
                    }
                    case "audio/mp3":
                    case "audio/mpeg": {
                      return {
                        type: "input_audio",
                        input_audio: {
                          data: chunkA45A2T7R_cjs.convertToBase64(part.data),
                          format: "mp3"
                        }
                      };
                    }
                    default: {
                      throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
                        functionality: `audio content parts with media type ${part.mediaType}`
                      });
                    }
                  }
                } else if (part.mediaType === "application/pdf") {
                  if (part.data instanceof URL) {
                    throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
                      functionality: "PDF file parts with URLs"
                    });
                  }
                  return {
                    type: "file",
                    file: typeof part.data === "string" && part.data.startsWith("file-") ? { file_id: part.data } : {
                      filename: (_c = part.filename) != null ? _c : `part-${index}.pdf`,
                      file_data: `data:application/pdf;base64,${chunkA45A2T7R_cjs.convertToBase64(part.data)}`
                    }
                  };
                } else {
                  throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
                    functionality: `file part media type ${part.mediaType}`
                  });
                }
              }
            }
          })
        });
        break;
      }
      case "assistant": {
        let text = "";
        const toolCalls = [];
        for (const part of content) {
          switch (part.type) {
            case "text": {
              text += part.text;
              break;
            }
            case "tool-call": {
              toolCalls.push({
                id: part.toolCallId,
                type: "function",
                function: {
                  name: part.toolName,
                  arguments: JSON.stringify(part.input)
                }
              });
              break;
            }
          }
        }
        messages.push({
          role: "assistant",
          content: text,
          tool_calls: toolCalls.length > 0 ? toolCalls : void 0
        });
        break;
      }
      case "tool": {
        for (const toolResponse of content) {
          const output = toolResponse.output;
          let contentValue;
          switch (output.type) {
            case "text":
            case "error-text":
              contentValue = output.value;
              break;
            case "content":
            case "json":
            case "error-json":
              contentValue = JSON.stringify(output.value);
              break;
          }
          messages.push({
            role: "tool",
            tool_call_id: toolResponse.toolCallId,
            content: contentValue
          });
        }
        break;
      }
      default: {
        const _exhaustiveCheck = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }
  return { messages, warnings };
}
function getResponseMetadata({
  id,
  model,
  created
}) {
  return {
    id: id != null ? id : void 0,
    modelId: model != null ? model : void 0,
    timestamp: created ? new Date(created * 1e3) : void 0
  };
}
function mapOpenAIFinishReason(finishReason) {
  switch (finishReason) {
    case "stop":
      return "stop";
    case "length":
      return "length";
    case "content_filter":
      return "content-filter";
    case "function_call":
    case "tool_calls":
      return "tool-calls";
    default:
      return "unknown";
  }
}
var openaiChatResponseSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      id: v4.z.string().nullish(),
      created: v4.z.number().nullish(),
      model: v4.z.string().nullish(),
      choices: v4.z.array(
        v4.z.object({
          message: v4.z.object({
            role: v4.z.literal("assistant").nullish(),
            content: v4.z.string().nullish(),
            tool_calls: v4.z.array(
              v4.z.object({
                id: v4.z.string().nullish(),
                type: v4.z.literal("function"),
                function: v4.z.object({
                  name: v4.z.string(),
                  arguments: v4.z.string()
                })
              })
            ).nullish(),
            annotations: v4.z.array(
              v4.z.object({
                type: v4.z.literal("url_citation"),
                url_citation: v4.z.object({
                  start_index: v4.z.number(),
                  end_index: v4.z.number(),
                  url: v4.z.string(),
                  title: v4.z.string()
                })
              })
            ).nullish()
          }),
          index: v4.z.number(),
          logprobs: v4.z.object({
            content: v4.z.array(
              v4.z.object({
                token: v4.z.string(),
                logprob: v4.z.number(),
                top_logprobs: v4.z.array(
                  v4.z.object({
                    token: v4.z.string(),
                    logprob: v4.z.number()
                  })
                )
              })
            ).nullish()
          }).nullish(),
          finish_reason: v4.z.string().nullish()
        })
      ),
      usage: v4.z.object({
        prompt_tokens: v4.z.number().nullish(),
        completion_tokens: v4.z.number().nullish(),
        total_tokens: v4.z.number().nullish(),
        prompt_tokens_details: v4.z.object({
          cached_tokens: v4.z.number().nullish()
        }).nullish(),
        completion_tokens_details: v4.z.object({
          reasoning_tokens: v4.z.number().nullish(),
          accepted_prediction_tokens: v4.z.number().nullish(),
          rejected_prediction_tokens: v4.z.number().nullish()
        }).nullish()
      }).nullish()
    })
  )
);
var openaiChatChunkSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.union([
      v4.z.object({
        id: v4.z.string().nullish(),
        created: v4.z.number().nullish(),
        model: v4.z.string().nullish(),
        choices: v4.z.array(
          v4.z.object({
            delta: v4.z.object({
              role: v4.z.enum(["assistant"]).nullish(),
              content: v4.z.string().nullish(),
              tool_calls: v4.z.array(
                v4.z.object({
                  index: v4.z.number(),
                  id: v4.z.string().nullish(),
                  type: v4.z.literal("function").nullish(),
                  function: v4.z.object({
                    name: v4.z.string().nullish(),
                    arguments: v4.z.string().nullish()
                  })
                })
              ).nullish(),
              annotations: v4.z.array(
                v4.z.object({
                  type: v4.z.literal("url_citation"),
                  url_citation: v4.z.object({
                    start_index: v4.z.number(),
                    end_index: v4.z.number(),
                    url: v4.z.string(),
                    title: v4.z.string()
                  })
                })
              ).nullish()
            }).nullish(),
            logprobs: v4.z.object({
              content: v4.z.array(
                v4.z.object({
                  token: v4.z.string(),
                  logprob: v4.z.number(),
                  top_logprobs: v4.z.array(
                    v4.z.object({
                      token: v4.z.string(),
                      logprob: v4.z.number()
                    })
                  )
                })
              ).nullish()
            }).nullish(),
            finish_reason: v4.z.string().nullish(),
            index: v4.z.number()
          })
        ),
        usage: v4.z.object({
          prompt_tokens: v4.z.number().nullish(),
          completion_tokens: v4.z.number().nullish(),
          total_tokens: v4.z.number().nullish(),
          prompt_tokens_details: v4.z.object({
            cached_tokens: v4.z.number().nullish()
          }).nullish(),
          completion_tokens_details: v4.z.object({
            reasoning_tokens: v4.z.number().nullish(),
            accepted_prediction_tokens: v4.z.number().nullish(),
            rejected_prediction_tokens: v4.z.number().nullish()
          }).nullish()
        }).nullish()
      }),
      openaiErrorDataSchema
    ])
  )
);
var openaiChatLanguageModelOptions = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      /**
       * Modify the likelihood of specified tokens appearing in the completion.
       *
       * Accepts a JSON object that maps tokens (specified by their token ID in
       * the GPT tokenizer) to an associated bias value from -100 to 100.
       */
      logitBias: v4.z.record(v4.z.coerce.number(), v4.z.number()).optional(),
      /**
       * Return the log probabilities of the tokens.
       *
       * Setting to true will return the log probabilities of the tokens that
       * were generated.
       *
       * Setting to a number will return the log probabilities of the top n
       * tokens that were generated.
       */
      logprobs: v4.z.union([v4.z.boolean(), v4.z.number()]).optional(),
      /**
       * Whether to enable parallel function calling during tool use. Default to true.
       */
      parallelToolCalls: v4.z.boolean().optional(),
      /**
       * A unique identifier representing your end-user, which can help OpenAI to
       * monitor and detect abuse.
       */
      user: v4.z.string().optional(),
      /**
       * Reasoning effort for reasoning models. Defaults to `medium`.
       */
      reasoningEffort: v4.z.enum(["none", "minimal", "low", "medium", "high", "xhigh"]).optional(),
      /**
       * Maximum number of completion tokens to generate. Useful for reasoning models.
       */
      maxCompletionTokens: v4.z.number().optional(),
      /**
       * Whether to enable persistence in responses API.
       */
      store: v4.z.boolean().optional(),
      /**
       * Metadata to associate with the request.
       */
      metadata: v4.z.record(v4.z.string().max(64), v4.z.string().max(512)).optional(),
      /**
       * Parameters for prediction mode.
       */
      prediction: v4.z.record(v4.z.string(), v4.z.any()).optional(),
      /**
       * Whether to use structured outputs.
       *
       * @default true
       */
      structuredOutputs: v4.z.boolean().optional(),
      /**
       * Service tier for the request.
       * - 'auto': Default service tier. The request will be processed with the service tier configured in the
       *           Project settings. Unless otherwise configured, the Project will use 'default'.
       * - 'flex': 50% cheaper processing at the cost of increased latency. Only available for o3 and o4-mini models.
       * - 'priority': Higher-speed processing with predictably low latency at premium cost. Available for Enterprise customers.
       * - 'default': The request will be processed with the standard pricing and performance for the selected model.
       *
       * @default 'auto'
       */
      serviceTier: v4.z.enum(["auto", "flex", "priority", "default"]).optional(),
      /**
       * Whether to use strict JSON schema validation.
       *
       * @default false
       */
      strictJsonSchema: v4.z.boolean().optional(),
      /**
       * Controls the verbosity of the model's responses.
       * Lower values will result in more concise responses, while higher values will result in more verbose responses.
       */
      textVerbosity: v4.z.enum(["low", "medium", "high"]).optional(),
      /**
       * A cache key for prompt caching. Allows manual control over prompt caching behavior.
       * Useful for improving cache hit rates and working around automatic caching issues.
       */
      promptCacheKey: v4.z.string().optional(),
      /**
       * The retention policy for the prompt cache.
       * - 'in_memory': Default. Standard prompt caching behavior.
       * - '24h': Extended prompt caching that keeps cached prefixes active for up to 24 hours.
       *          Currently only available for 5.1 series models.
       *
       * @default 'in_memory'
       */
      promptCacheRetention: v4.z.enum(["in_memory", "24h"]).optional(),
      /**
       * A stable identifier used to help detect users of your application
       * that may be violating OpenAI's usage policies. The IDs should be a
       * string that uniquely identifies each user. We recommend hashing their
       * username or email address, in order to avoid sending us any identifying
       * information.
       */
      safetyIdentifier: v4.z.string().optional()
    })
  )
);
function prepareChatTools({
  tools,
  toolChoice,
  structuredOutputs,
  strictJsonSchema
}) {
  tools = (tools == null ? void 0 : tools.length) ? tools : void 0;
  const toolWarnings = [];
  if (tools == null) {
    return { tools: void 0, toolChoice: void 0, toolWarnings };
  }
  const openaiTools = [];
  for (const tool of tools) {
    switch (tool.type) {
      case "function":
        openaiTools.push({
          type: "function",
          function: {
            name: tool.name,
            description: tool.description,
            parameters: tool.inputSchema,
            strict: structuredOutputs ? strictJsonSchema : void 0
          }
        });
        break;
      default:
        toolWarnings.push({ type: "unsupported-tool", tool });
        break;
    }
  }
  if (toolChoice == null) {
    return { tools: openaiTools, toolChoice: void 0, toolWarnings };
  }
  const type = toolChoice.type;
  switch (type) {
    case "auto":
    case "none":
    case "required":
      return { tools: openaiTools, toolChoice: type, toolWarnings };
    case "tool":
      return {
        tools: openaiTools,
        toolChoice: {
          type: "function",
          function: {
            name: toolChoice.toolName
          }
        },
        toolWarnings
      };
    default: {
      const _exhaustiveCheck = type;
      throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
        functionality: `tool choice type: ${_exhaustiveCheck}`
      });
    }
  }
}
var OpenAIChatLanguageModel = class {
  constructor(modelId, config) {
    this.specificationVersion = "v2";
    this.supportedUrls = {
      "image/*": [/^https?:\/\/.*$/]
    };
    this.modelId = modelId;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  async getArgs({
    prompt,
    maxOutputTokens,
    temperature,
    topP,
    topK,
    frequencyPenalty,
    presencePenalty,
    stopSequences,
    responseFormat,
    seed,
    tools,
    toolChoice,
    providerOptions
  }) {
    var _a, _b, _c, _d;
    const warnings = [];
    const openaiOptions = (_a = await chunkA45A2T7R_cjs.parseProviderOptions({
      provider: "openai",
      providerOptions,
      schema: openaiChatLanguageModelOptions
    })) != null ? _a : {};
    const structuredOutputs = (_b = openaiOptions.structuredOutputs) != null ? _b : true;
    const modelCapabilities = getOpenAILanguageModelCapabilities(this.modelId);
    if (topK != null) {
      warnings.push({
        type: "unsupported-setting",
        setting: "topK"
      });
    }
    if ((responseFormat == null ? void 0 : responseFormat.type) === "json" && responseFormat.schema != null && !structuredOutputs) {
      warnings.push({
        type: "unsupported-setting",
        setting: "responseFormat",
        details: "JSON response format schema is only supported with structuredOutputs"
      });
    }
    const { messages, warnings: messageWarnings } = convertToOpenAIChatMessages(
      {
        prompt,
        systemMessageMode: modelCapabilities.systemMessageMode
      }
    );
    warnings.push(...messageWarnings);
    const strictJsonSchema = (_c = openaiOptions.strictJsonSchema) != null ? _c : false;
    const baseArgs = {
      // model id:
      model: this.modelId,
      // model specific settings:
      logit_bias: openaiOptions.logitBias,
      logprobs: openaiOptions.logprobs === true || typeof openaiOptions.logprobs === "number" ? true : void 0,
      top_logprobs: typeof openaiOptions.logprobs === "number" ? openaiOptions.logprobs : typeof openaiOptions.logprobs === "boolean" ? openaiOptions.logprobs ? 0 : void 0 : void 0,
      user: openaiOptions.user,
      parallel_tool_calls: openaiOptions.parallelToolCalls,
      // standardized settings:
      max_tokens: maxOutputTokens,
      temperature,
      top_p: topP,
      frequency_penalty: frequencyPenalty,
      presence_penalty: presencePenalty,
      response_format: (responseFormat == null ? void 0 : responseFormat.type) === "json" ? structuredOutputs && responseFormat.schema != null ? {
        type: "json_schema",
        json_schema: {
          schema: responseFormat.schema,
          strict: strictJsonSchema,
          name: (_d = responseFormat.name) != null ? _d : "response",
          description: responseFormat.description
        }
      } : { type: "json_object" } : void 0,
      stop: stopSequences,
      seed,
      verbosity: openaiOptions.textVerbosity,
      // openai specific settings:
      // TODO AI SDK 6: remove, we auto-map maxOutputTokens now
      max_completion_tokens: openaiOptions.maxCompletionTokens,
      store: openaiOptions.store,
      metadata: openaiOptions.metadata,
      prediction: openaiOptions.prediction,
      reasoning_effort: openaiOptions.reasoningEffort,
      service_tier: openaiOptions.serviceTier,
      prompt_cache_key: openaiOptions.promptCacheKey,
      prompt_cache_retention: openaiOptions.promptCacheRetention,
      safety_identifier: openaiOptions.safetyIdentifier,
      // messages:
      messages
    };
    if (modelCapabilities.isReasoningModel) {
      if (openaiOptions.reasoningEffort !== "none" || !modelCapabilities.supportsNonReasoningParameters) {
        if (baseArgs.temperature != null) {
          baseArgs.temperature = void 0;
          warnings.push({
            type: "unsupported-setting",
            setting: "temperature",
            details: "temperature is not supported for reasoning models"
          });
        }
        if (baseArgs.top_p != null) {
          baseArgs.top_p = void 0;
          warnings.push({
            type: "unsupported-setting",
            setting: "topP",
            details: "topP is not supported for reasoning models"
          });
        }
        if (baseArgs.logprobs != null) {
          baseArgs.logprobs = void 0;
          warnings.push({
            type: "other",
            message: "logprobs is not supported for reasoning models"
          });
        }
      }
      if (baseArgs.frequency_penalty != null) {
        baseArgs.frequency_penalty = void 0;
        warnings.push({
          type: "unsupported-setting",
          setting: "frequencyPenalty",
          details: "frequencyPenalty is not supported for reasoning models"
        });
      }
      if (baseArgs.presence_penalty != null) {
        baseArgs.presence_penalty = void 0;
        warnings.push({
          type: "unsupported-setting",
          setting: "presencePenalty",
          details: "presencePenalty is not supported for reasoning models"
        });
      }
      if (baseArgs.logit_bias != null) {
        baseArgs.logit_bias = void 0;
        warnings.push({
          type: "other",
          message: "logitBias is not supported for reasoning models"
        });
      }
      if (baseArgs.top_logprobs != null) {
        baseArgs.top_logprobs = void 0;
        warnings.push({
          type: "other",
          message: "topLogprobs is not supported for reasoning models"
        });
      }
      if (baseArgs.max_tokens != null) {
        if (baseArgs.max_completion_tokens == null) {
          baseArgs.max_completion_tokens = baseArgs.max_tokens;
        }
        baseArgs.max_tokens = void 0;
      }
    } else if (this.modelId.startsWith("gpt-4o-search-preview") || this.modelId.startsWith("gpt-4o-mini-search-preview")) {
      if (baseArgs.temperature != null) {
        baseArgs.temperature = void 0;
        warnings.push({
          type: "unsupported-setting",
          setting: "temperature",
          details: "temperature is not supported for the search preview models and has been removed."
        });
      }
    }
    if (openaiOptions.serviceTier === "flex" && !modelCapabilities.supportsFlexProcessing) {
      warnings.push({
        type: "unsupported-setting",
        setting: "serviceTier",
        details: "flex processing is only available for o3, o4-mini, and gpt-5 models"
      });
      baseArgs.service_tier = void 0;
    }
    if (openaiOptions.serviceTier === "priority" && !modelCapabilities.supportsPriorityProcessing) {
      warnings.push({
        type: "unsupported-setting",
        setting: "serviceTier",
        details: "priority processing is only available for supported models (gpt-4, gpt-5, gpt-5-mini, o3, o4-mini) and requires Enterprise access. gpt-5-nano is not supported"
      });
      baseArgs.service_tier = void 0;
    }
    const {
      tools: openaiTools,
      toolChoice: openaiToolChoice,
      toolWarnings
    } = prepareChatTools({
      tools,
      toolChoice,
      structuredOutputs,
      strictJsonSchema
    });
    return {
      args: {
        ...baseArgs,
        tools: openaiTools,
        tool_choice: openaiToolChoice
      },
      warnings: [...warnings, ...toolWarnings]
    };
  }
  async doGenerate(options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n;
    const { args: body, warnings } = await this.getArgs(options);
    const {
      responseHeaders,
      value: response,
      rawValue: rawResponse
    } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/chat/completions",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createJsonResponseHandler(
        openaiChatResponseSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const choice = response.choices[0];
    const content = [];
    const text = choice.message.content;
    if (text != null && text.length > 0) {
      content.push({ type: "text", text });
    }
    for (const toolCall of (_a = choice.message.tool_calls) != null ? _a : []) {
      content.push({
        type: "tool-call",
        toolCallId: (_b = toolCall.id) != null ? _b : chunkA45A2T7R_cjs.generateId(),
        toolName: toolCall.function.name,
        input: toolCall.function.arguments
      });
    }
    for (const annotation of (_c = choice.message.annotations) != null ? _c : []) {
      content.push({
        type: "source",
        sourceType: "url",
        id: chunkA45A2T7R_cjs.generateId(),
        url: annotation.url_citation.url,
        title: annotation.url_citation.title
      });
    }
    const completionTokenDetails = (_d = response.usage) == null ? void 0 : _d.completion_tokens_details;
    const promptTokenDetails = (_e = response.usage) == null ? void 0 : _e.prompt_tokens_details;
    const providerMetadata = { openai: {} };
    if ((completionTokenDetails == null ? void 0 : completionTokenDetails.accepted_prediction_tokens) != null) {
      providerMetadata.openai.acceptedPredictionTokens = completionTokenDetails == null ? void 0 : completionTokenDetails.accepted_prediction_tokens;
    }
    if ((completionTokenDetails == null ? void 0 : completionTokenDetails.rejected_prediction_tokens) != null) {
      providerMetadata.openai.rejectedPredictionTokens = completionTokenDetails == null ? void 0 : completionTokenDetails.rejected_prediction_tokens;
    }
    if (((_f = choice.logprobs) == null ? void 0 : _f.content) != null) {
      providerMetadata.openai.logprobs = choice.logprobs.content;
    }
    return {
      content,
      finishReason: mapOpenAIFinishReason(choice.finish_reason),
      usage: {
        inputTokens: (_h = (_g = response.usage) == null ? void 0 : _g.prompt_tokens) != null ? _h : void 0,
        outputTokens: (_j = (_i = response.usage) == null ? void 0 : _i.completion_tokens) != null ? _j : void 0,
        totalTokens: (_l = (_k = response.usage) == null ? void 0 : _k.total_tokens) != null ? _l : void 0,
        reasoningTokens: (_m = completionTokenDetails == null ? void 0 : completionTokenDetails.reasoning_tokens) != null ? _m : void 0,
        cachedInputTokens: (_n = promptTokenDetails == null ? void 0 : promptTokenDetails.cached_tokens) != null ? _n : void 0
      },
      request: { body },
      response: {
        ...getResponseMetadata(response),
        headers: responseHeaders,
        body: rawResponse
      },
      warnings,
      providerMetadata
    };
  }
  async doStream(options) {
    const { args, warnings } = await this.getArgs(options);
    const body = {
      ...args,
      stream: true,
      stream_options: {
        include_usage: true
      }
    };
    const { responseHeaders, value: response } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/chat/completions",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createEventSourceResponseHandler(
        openaiChatChunkSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const toolCalls = [];
    let finishReason = "unknown";
    const usage = {
      inputTokens: void 0,
      outputTokens: void 0,
      totalTokens: void 0
    };
    let metadataExtracted = false;
    let isActiveText = false;
    const providerMetadata = { openai: {} };
    return {
      stream: response.pipeThrough(
        new TransformStream({
          start(controller) {
            controller.enqueue({ type: "stream-start", warnings });
          },
          transform(chunk, controller) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x;
            if (options.includeRawChunks) {
              controller.enqueue({ type: "raw", rawValue: chunk.rawValue });
            }
            if (!chunk.success) {
              finishReason = "error";
              controller.enqueue({ type: "error", error: chunk.error });
              return;
            }
            const value = chunk.value;
            if ("error" in value) {
              finishReason = "error";
              controller.enqueue({ type: "error", error: value.error });
              return;
            }
            if (!metadataExtracted) {
              const metadata = getResponseMetadata(value);
              if (Object.values(metadata).some(Boolean)) {
                metadataExtracted = true;
                controller.enqueue({
                  type: "response-metadata",
                  ...getResponseMetadata(value)
                });
              }
            }
            if (value.usage != null) {
              usage.inputTokens = (_a = value.usage.prompt_tokens) != null ? _a : void 0;
              usage.outputTokens = (_b = value.usage.completion_tokens) != null ? _b : void 0;
              usage.totalTokens = (_c = value.usage.total_tokens) != null ? _c : void 0;
              usage.reasoningTokens = (_e = (_d = value.usage.completion_tokens_details) == null ? void 0 : _d.reasoning_tokens) != null ? _e : void 0;
              usage.cachedInputTokens = (_g = (_f = value.usage.prompt_tokens_details) == null ? void 0 : _f.cached_tokens) != null ? _g : void 0;
              if (((_h = value.usage.completion_tokens_details) == null ? void 0 : _h.accepted_prediction_tokens) != null) {
                providerMetadata.openai.acceptedPredictionTokens = (_i = value.usage.completion_tokens_details) == null ? void 0 : _i.accepted_prediction_tokens;
              }
              if (((_j = value.usage.completion_tokens_details) == null ? void 0 : _j.rejected_prediction_tokens) != null) {
                providerMetadata.openai.rejectedPredictionTokens = (_k = value.usage.completion_tokens_details) == null ? void 0 : _k.rejected_prediction_tokens;
              }
            }
            const choice = value.choices[0];
            if ((choice == null ? void 0 : choice.finish_reason) != null) {
              finishReason = mapOpenAIFinishReason(choice.finish_reason);
            }
            if (((_l = choice == null ? void 0 : choice.logprobs) == null ? void 0 : _l.content) != null) {
              providerMetadata.openai.logprobs = choice.logprobs.content;
            }
            if ((choice == null ? void 0 : choice.delta) == null) {
              return;
            }
            const delta = choice.delta;
            if (delta.content != null) {
              if (!isActiveText) {
                controller.enqueue({ type: "text-start", id: "0" });
                isActiveText = true;
              }
              controller.enqueue({
                type: "text-delta",
                id: "0",
                delta: delta.content
              });
            }
            if (delta.tool_calls != null) {
              for (const toolCallDelta of delta.tool_calls) {
                const index = toolCallDelta.index;
                if (toolCalls[index] == null) {
                  if (toolCallDelta.type != null && toolCallDelta.type !== "function") {
                    throw new chunkA45A2T7R_cjs.InvalidResponseDataError({
                      data: toolCallDelta,
                      message: `Expected 'function' type.`
                    });
                  }
                  if (toolCallDelta.id == null) {
                    throw new chunkA45A2T7R_cjs.InvalidResponseDataError({
                      data: toolCallDelta,
                      message: `Expected 'id' to be a string.`
                    });
                  }
                  if (((_m = toolCallDelta.function) == null ? void 0 : _m.name) == null) {
                    throw new chunkA45A2T7R_cjs.InvalidResponseDataError({
                      data: toolCallDelta,
                      message: `Expected 'function.name' to be a string.`
                    });
                  }
                  controller.enqueue({
                    type: "tool-input-start",
                    id: toolCallDelta.id,
                    toolName: toolCallDelta.function.name
                  });
                  toolCalls[index] = {
                    id: toolCallDelta.id,
                    type: "function",
                    function: {
                      name: toolCallDelta.function.name,
                      arguments: (_n = toolCallDelta.function.arguments) != null ? _n : ""
                    },
                    hasFinished: false
                  };
                  const toolCall2 = toolCalls[index];
                  if (((_o = toolCall2.function) == null ? void 0 : _o.name) != null && ((_p = toolCall2.function) == null ? void 0 : _p.arguments) != null) {
                    if (toolCall2.function.arguments.length > 0) {
                      controller.enqueue({
                        type: "tool-input-delta",
                        id: toolCall2.id,
                        delta: toolCall2.function.arguments
                      });
                    }
                    if (chunkA45A2T7R_cjs.isParsableJson(toolCall2.function.arguments)) {
                      controller.enqueue({
                        type: "tool-input-end",
                        id: toolCall2.id
                      });
                      controller.enqueue({
                        type: "tool-call",
                        toolCallId: (_q = toolCall2.id) != null ? _q : chunkA45A2T7R_cjs.generateId(),
                        toolName: toolCall2.function.name,
                        input: toolCall2.function.arguments
                      });
                      toolCall2.hasFinished = true;
                    }
                  }
                  continue;
                }
                const toolCall = toolCalls[index];
                if (toolCall.hasFinished) {
                  continue;
                }
                if (((_r = toolCallDelta.function) == null ? void 0 : _r.arguments) != null) {
                  toolCall.function.arguments += (_t = (_s = toolCallDelta.function) == null ? void 0 : _s.arguments) != null ? _t : "";
                }
                controller.enqueue({
                  type: "tool-input-delta",
                  id: toolCall.id,
                  delta: (_u = toolCallDelta.function.arguments) != null ? _u : ""
                });
                if (((_v = toolCall.function) == null ? void 0 : _v.name) != null && ((_w = toolCall.function) == null ? void 0 : _w.arguments) != null && chunkA45A2T7R_cjs.isParsableJson(toolCall.function.arguments)) {
                  controller.enqueue({
                    type: "tool-input-end",
                    id: toolCall.id
                  });
                  controller.enqueue({
                    type: "tool-call",
                    toolCallId: (_x = toolCall.id) != null ? _x : chunkA45A2T7R_cjs.generateId(),
                    toolName: toolCall.function.name,
                    input: toolCall.function.arguments
                  });
                  toolCall.hasFinished = true;
                }
              }
            }
            if (delta.annotations != null) {
              for (const annotation of delta.annotations) {
                controller.enqueue({
                  type: "source",
                  sourceType: "url",
                  id: chunkA45A2T7R_cjs.generateId(),
                  url: annotation.url_citation.url,
                  title: annotation.url_citation.title
                });
              }
            }
          },
          flush(controller) {
            if (isActiveText) {
              controller.enqueue({ type: "text-end", id: "0" });
            }
            controller.enqueue({
              type: "finish",
              finishReason,
              usage,
              ...providerMetadata != null ? { providerMetadata } : {}
            });
          }
        })
      ),
      request: { body },
      response: { headers: responseHeaders }
    };
  }
};
function convertToOpenAICompletionPrompt({
  prompt,
  user = "user",
  assistant = "assistant"
}) {
  let text = "";
  if (prompt[0].role === "system") {
    text += `${prompt[0].content}

`;
    prompt = prompt.slice(1);
  }
  for (const { role, content } of prompt) {
    switch (role) {
      case "system": {
        throw new chunkA45A2T7R_cjs.InvalidPromptError({
          message: "Unexpected system message in prompt: ${content}",
          prompt
        });
      }
      case "user": {
        const userMessage = content.map((part) => {
          switch (part.type) {
            case "text": {
              return part.text;
            }
          }
        }).filter(Boolean).join("");
        text += `${user}:
${userMessage}

`;
        break;
      }
      case "assistant": {
        const assistantMessage = content.map((part) => {
          switch (part.type) {
            case "text": {
              return part.text;
            }
            case "tool-call": {
              throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
                functionality: "tool-call messages"
              });
            }
          }
        }).join("");
        text += `${assistant}:
${assistantMessage}

`;
        break;
      }
      case "tool": {
        throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
          functionality: "tool messages"
        });
      }
      default: {
        const _exhaustiveCheck = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }
  text += `${assistant}:
`;
  return {
    prompt: text,
    stopSequences: [`
${user}:`]
  };
}
function getResponseMetadata2({
  id,
  model,
  created
}) {
  return {
    id: id != null ? id : void 0,
    modelId: model != null ? model : void 0,
    timestamp: created != null ? new Date(created * 1e3) : void 0
  };
}
function mapOpenAIFinishReason2(finishReason) {
  switch (finishReason) {
    case "stop":
      return "stop";
    case "length":
      return "length";
    case "content_filter":
      return "content-filter";
    case "function_call":
    case "tool_calls":
      return "tool-calls";
    default:
      return "unknown";
  }
}
var openaiCompletionResponseSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      id: v4.z.string().nullish(),
      created: v4.z.number().nullish(),
      model: v4.z.string().nullish(),
      choices: v4.z.array(
        v4.z.object({
          text: v4.z.string(),
          finish_reason: v4.z.string(),
          logprobs: v4.z.object({
            tokens: v4.z.array(v4.z.string()),
            token_logprobs: v4.z.array(v4.z.number()),
            top_logprobs: v4.z.array(v4.z.record(v4.z.string(), v4.z.number())).nullish()
          }).nullish()
        })
      ),
      usage: v4.z.object({
        prompt_tokens: v4.z.number(),
        completion_tokens: v4.z.number(),
        total_tokens: v4.z.number()
      }).nullish()
    })
  )
);
var openaiCompletionChunkSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.union([
      v4.z.object({
        id: v4.z.string().nullish(),
        created: v4.z.number().nullish(),
        model: v4.z.string().nullish(),
        choices: v4.z.array(
          v4.z.object({
            text: v4.z.string(),
            finish_reason: v4.z.string().nullish(),
            index: v4.z.number(),
            logprobs: v4.z.object({
              tokens: v4.z.array(v4.z.string()),
              token_logprobs: v4.z.array(v4.z.number()),
              top_logprobs: v4.z.array(v4.z.record(v4.z.string(), v4.z.number())).nullish()
            }).nullish()
          })
        ),
        usage: v4.z.object({
          prompt_tokens: v4.z.number(),
          completion_tokens: v4.z.number(),
          total_tokens: v4.z.number()
        }).nullish()
      }),
      openaiErrorDataSchema
    ])
  )
);
var openaiCompletionProviderOptions = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      /**
      Echo back the prompt in addition to the completion.
         */
      echo: v4.z.boolean().optional(),
      /**
      Modify the likelihood of specified tokens appearing in the completion.
      
      Accepts a JSON object that maps tokens (specified by their token ID in
      the GPT tokenizer) to an associated bias value from -100 to 100. You
      can use this tokenizer tool to convert text to token IDs. Mathematically,
      the bias is added to the logits generated by the model prior to sampling.
      The exact effect will vary per model, but values between -1 and 1 should
      decrease or increase likelihood of selection; values like -100 or 100
      should result in a ban or exclusive selection of the relevant token.
      
      As an example, you can pass {"50256": -100} to prevent the <|endoftext|>
      token from being generated.
       */
      logitBias: v4.z.record(v4.z.string(), v4.z.number()).optional(),
      /**
      The suffix that comes after a completion of inserted text.
       */
      suffix: v4.z.string().optional(),
      /**
      A unique identifier representing your end-user, which can help OpenAI to
      monitor and detect abuse. Learn more.
       */
      user: v4.z.string().optional(),
      /**
      Return the log probabilities of the tokens. Including logprobs will increase
      the response size and can slow down response times. However, it can
      be useful to better understand how the model is behaving.
      Setting to true will return the log probabilities of the tokens that
      were generated.
      Setting to a number will return the log probabilities of the top n
      tokens that were generated.
         */
      logprobs: v4.z.union([v4.z.boolean(), v4.z.number()]).optional()
    })
  )
);
var OpenAICompletionLanguageModel = class {
  constructor(modelId, config) {
    this.specificationVersion = "v2";
    this.supportedUrls = {
      // No URLs are supported for completion models.
    };
    this.modelId = modelId;
    this.config = config;
  }
  get providerOptionsName() {
    return this.config.provider.split(".")[0].trim();
  }
  get provider() {
    return this.config.provider;
  }
  async getArgs({
    prompt,
    maxOutputTokens,
    temperature,
    topP,
    topK,
    frequencyPenalty,
    presencePenalty,
    stopSequences: userStopSequences,
    responseFormat,
    tools,
    toolChoice,
    seed,
    providerOptions
  }) {
    const warnings = [];
    const openaiOptions = {
      ...await chunkA45A2T7R_cjs.parseProviderOptions({
        provider: "openai",
        providerOptions,
        schema: openaiCompletionProviderOptions
      }),
      ...await chunkA45A2T7R_cjs.parseProviderOptions({
        provider: this.providerOptionsName,
        providerOptions,
        schema: openaiCompletionProviderOptions
      })
    };
    if (topK != null) {
      warnings.push({ type: "unsupported-setting", setting: "topK" });
    }
    if (tools == null ? void 0 : tools.length) {
      warnings.push({ type: "unsupported-setting", setting: "tools" });
    }
    if (toolChoice != null) {
      warnings.push({ type: "unsupported-setting", setting: "toolChoice" });
    }
    if (responseFormat != null && responseFormat.type !== "text") {
      warnings.push({
        type: "unsupported-setting",
        setting: "responseFormat",
        details: "JSON response format is not supported."
      });
    }
    const { prompt: completionPrompt, stopSequences } = convertToOpenAICompletionPrompt({ prompt });
    const stop = [...stopSequences != null ? stopSequences : [], ...userStopSequences != null ? userStopSequences : []];
    return {
      args: {
        // model id:
        model: this.modelId,
        // model specific settings:
        echo: openaiOptions.echo,
        logit_bias: openaiOptions.logitBias,
        logprobs: (openaiOptions == null ? void 0 : openaiOptions.logprobs) === true ? 0 : (openaiOptions == null ? void 0 : openaiOptions.logprobs) === false ? void 0 : openaiOptions == null ? void 0 : openaiOptions.logprobs,
        suffix: openaiOptions.suffix,
        user: openaiOptions.user,
        // standardized settings:
        max_tokens: maxOutputTokens,
        temperature,
        top_p: topP,
        frequency_penalty: frequencyPenalty,
        presence_penalty: presencePenalty,
        seed,
        // prompt:
        prompt: completionPrompt,
        // stop sequences:
        stop: stop.length > 0 ? stop : void 0
      },
      warnings
    };
  }
  async doGenerate(options) {
    var _a, _b, _c;
    const { args, warnings } = await this.getArgs(options);
    const {
      responseHeaders,
      value: response,
      rawValue: rawResponse
    } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/completions",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body: args,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createJsonResponseHandler(
        openaiCompletionResponseSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const choice = response.choices[0];
    const providerMetadata = { openai: {} };
    if (choice.logprobs != null) {
      providerMetadata.openai.logprobs = choice.logprobs;
    }
    return {
      content: [{ type: "text", text: choice.text }],
      usage: {
        inputTokens: (_a = response.usage) == null ? void 0 : _a.prompt_tokens,
        outputTokens: (_b = response.usage) == null ? void 0 : _b.completion_tokens,
        totalTokens: (_c = response.usage) == null ? void 0 : _c.total_tokens
      },
      finishReason: mapOpenAIFinishReason2(choice.finish_reason),
      request: { body: args },
      response: {
        ...getResponseMetadata2(response),
        headers: responseHeaders,
        body: rawResponse
      },
      providerMetadata,
      warnings
    };
  }
  async doStream(options) {
    const { args, warnings } = await this.getArgs(options);
    const body = {
      ...args,
      stream: true,
      stream_options: {
        include_usage: true
      }
    };
    const { responseHeaders, value: response } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/completions",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createEventSourceResponseHandler(
        openaiCompletionChunkSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    let finishReason = "unknown";
    const providerMetadata = { openai: {} };
    const usage = {
      inputTokens: void 0,
      outputTokens: void 0,
      totalTokens: void 0
    };
    let isFirstChunk = true;
    return {
      stream: response.pipeThrough(
        new TransformStream({
          start(controller) {
            controller.enqueue({ type: "stream-start", warnings });
          },
          transform(chunk, controller) {
            if (options.includeRawChunks) {
              controller.enqueue({ type: "raw", rawValue: chunk.rawValue });
            }
            if (!chunk.success) {
              finishReason = "error";
              controller.enqueue({ type: "error", error: chunk.error });
              return;
            }
            const value = chunk.value;
            if ("error" in value) {
              finishReason = "error";
              controller.enqueue({ type: "error", error: value.error });
              return;
            }
            if (isFirstChunk) {
              isFirstChunk = false;
              controller.enqueue({
                type: "response-metadata",
                ...getResponseMetadata2(value)
              });
              controller.enqueue({ type: "text-start", id: "0" });
            }
            if (value.usage != null) {
              usage.inputTokens = value.usage.prompt_tokens;
              usage.outputTokens = value.usage.completion_tokens;
              usage.totalTokens = value.usage.total_tokens;
            }
            const choice = value.choices[0];
            if ((choice == null ? void 0 : choice.finish_reason) != null) {
              finishReason = mapOpenAIFinishReason2(choice.finish_reason);
            }
            if ((choice == null ? void 0 : choice.logprobs) != null) {
              providerMetadata.openai.logprobs = choice.logprobs;
            }
            if ((choice == null ? void 0 : choice.text) != null && choice.text.length > 0) {
              controller.enqueue({
                type: "text-delta",
                id: "0",
                delta: choice.text
              });
            }
          },
          flush(controller) {
            if (!isFirstChunk) {
              controller.enqueue({ type: "text-end", id: "0" });
            }
            controller.enqueue({
              type: "finish",
              finishReason,
              providerMetadata,
              usage
            });
          }
        })
      ),
      request: { body },
      response: { headers: responseHeaders }
    };
  }
};
var openaiEmbeddingProviderOptions = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      /**
      The number of dimensions the resulting output embeddings should have.
      Only supported in text-embedding-3 and later models.
         */
      dimensions: v4.z.number().optional(),
      /**
      A unique identifier representing your end-user, which can help OpenAI to
      monitor and detect abuse. Learn more.
      */
      user: v4.z.string().optional()
    })
  )
);
var openaiTextEmbeddingResponseSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      data: v4.z.array(v4.z.object({ embedding: v4.z.array(v4.z.number()) })),
      usage: v4.z.object({ prompt_tokens: v4.z.number() }).nullish()
    })
  )
);
var OpenAIEmbeddingModel = class {
  constructor(modelId, config) {
    this.specificationVersion = "v2";
    this.maxEmbeddingsPerCall = 2048;
    this.supportsParallelCalls = true;
    this.modelId = modelId;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  async doEmbed({
    values,
    headers,
    abortSignal,
    providerOptions
  }) {
    var _a;
    if (values.length > this.maxEmbeddingsPerCall) {
      throw new chunkA45A2T7R_cjs.TooManyEmbeddingValuesForCallError({
        provider: this.provider,
        modelId: this.modelId,
        maxEmbeddingsPerCall: this.maxEmbeddingsPerCall,
        values
      });
    }
    const openaiOptions = (_a = await chunkA45A2T7R_cjs.parseProviderOptions({
      provider: "openai",
      providerOptions,
      schema: openaiEmbeddingProviderOptions
    })) != null ? _a : {};
    const {
      responseHeaders,
      value: response,
      rawValue
    } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/embeddings",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), headers),
      body: {
        model: this.modelId,
        input: values,
        encoding_format: "float",
        dimensions: openaiOptions.dimensions,
        user: openaiOptions.user
      },
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createJsonResponseHandler(
        openaiTextEmbeddingResponseSchema
      ),
      abortSignal,
      fetch: this.config.fetch
    });
    return {
      embeddings: response.data.map((item) => item.embedding),
      usage: response.usage ? { tokens: response.usage.prompt_tokens } : void 0,
      response: { headers: responseHeaders, body: rawValue }
    };
  }
};
var openaiImageResponseSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      created: v4.z.number().nullish(),
      data: v4.z.array(
        v4.z.object({
          b64_json: v4.z.string(),
          revised_prompt: v4.z.string().nullish()
        })
      ),
      background: v4.z.string().nullish(),
      output_format: v4.z.string().nullish(),
      size: v4.z.string().nullish(),
      quality: v4.z.string().nullish(),
      usage: v4.z.object({
        input_tokens: v4.z.number().nullish(),
        output_tokens: v4.z.number().nullish(),
        total_tokens: v4.z.number().nullish(),
        input_tokens_details: v4.z.object({
          image_tokens: v4.z.number().nullish(),
          text_tokens: v4.z.number().nullish()
        }).nullish()
      }).nullish()
    })
  )
);
var modelMaxImagesPerCall = {
  "dall-e-3": 1,
  "dall-e-2": 10,
  "gpt-image-1": 10,
  "gpt-image-1-mini": 10,
  "gpt-image-1.5": 10,
  "gpt-image-2": 10
};
var hasDefaultResponseFormat = /* @__PURE__ */ new Set([
  "gpt-image-1",
  "gpt-image-1-mini",
  "gpt-image-1.5",
  "gpt-image-2"
]);
var baseImageModelOptionsObject = v4.z.object({
  /**
   * Quality of the generated image(s).
   *
   * Valid values: `standard`, `hd`, `low`, `medium`, `high`, `auto`.
   */
  quality: v4.z.enum(["standard", "hd", "low", "medium", "high", "auto"]).optional(),
  /**
   * Background behavior for the generated image(s).
   *
   * If `transparent`, the output format must support transparency
   * (i.e. `png` or `webp`).
   */
  background: v4.z.enum(["transparent", "opaque", "auto"]).optional(),
  /**
   * Format in which the generated image(s) are returned.
   */
  outputFormat: v4.z.enum(["png", "jpeg", "webp"]).optional(),
  /**
   * Compression level (0-100) for the generated image(s). Applies to the
   * `jpeg` and `webp` output formats.
   */
  outputCompression: v4.z.number().int().min(0).max(100).optional(),
  /**
   * A unique identifier representing your end-user, which can help OpenAI
   * to monitor and detect abuse.
   */
  user: v4.z.string().optional()
});
chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(baseImageModelOptionsObject)
);
var openaiImageModelGenerationOptions = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    baseImageModelOptionsObject.extend({
      /**
       * Style of the generated image. `vivid` produces hyper-real and
       * dramatic images; `natural` produces more subdued, less hyper-real
       * looking images.
       */
      style: v4.z.enum(["vivid", "natural"]).optional(),
      /**
       * Content moderation level for the generated image(s). `low` applies
       * less restrictive filtering.
       */
      moderation: v4.z.enum(["auto", "low"]).optional()
    })
  )
);
var OpenAIImageModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v2";
  }
  get maxImagesPerCall() {
    var _a;
    return (_a = modelMaxImagesPerCall[this.modelId]) != null ? _a : 1;
  }
  get provider() {
    return this.config.provider;
  }
  async doGenerate({
    prompt,
    n,
    size,
    aspectRatio,
    seed,
    providerOptions,
    headers,
    abortSignal
  }) {
    var _a, _b, _c, _d;
    const warnings = [];
    if (aspectRatio != null) {
      warnings.push({
        type: "unsupported-setting",
        setting: "aspectRatio",
        details: "This model does not support aspect ratio. Use `size` instead."
      });
    }
    if (seed != null) {
      warnings.push({ type: "unsupported-setting", setting: "seed" });
    }
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const openaiOptions = (_d = await chunkA45A2T7R_cjs.parseProviderOptions({
      provider: "openai",
      providerOptions,
      schema: openaiImageModelGenerationOptions
    })) != null ? _d : {};
    const { value: response, responseHeaders } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/images/generations",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), headers),
      body: {
        model: this.modelId,
        prompt,
        n,
        size,
        quality: openaiOptions.quality,
        style: openaiOptions.style,
        background: openaiOptions.background,
        moderation: openaiOptions.moderation,
        output_format: openaiOptions.outputFormat,
        output_compression: openaiOptions.outputCompression,
        user: openaiOptions.user,
        ...!hasDefaultResponseFormat.has(this.modelId) ? { response_format: "b64_json" } : {}
      },
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createJsonResponseHandler(
        openaiImageResponseSchema
      ),
      abortSignal,
      fetch: this.config.fetch
    });
    return {
      images: response.data.map((item) => item.b64_json),
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders
      },
      providerMetadata: {
        openai: {
          images: response.data.map((item, index) => ({
            ...item.revised_prompt ? { revisedPrompt: item.revised_prompt } : {},
            ...response.created != null ? { created: response.created } : {},
            ...response.size != null ? { size: response.size } : {},
            ...response.quality != null ? { quality: response.quality } : {},
            ...response.background != null ? { background: response.background } : {},
            ...response.output_format != null ? { outputFormat: response.output_format } : {},
            ...distributeTokenDetails(
              response.usage,
              index,
              response.data.length
            )
          }))
        }
      }
    };
  }
};
function distributeTokenDetails(usage, index, total) {
  if (usage == null) {
    return {};
  }
  const result = {};
  const details = usage.input_tokens_details;
  if ((details == null ? void 0 : details.image_tokens) != null) {
    const base = Math.floor(details.image_tokens / total);
    const remainder = details.image_tokens - base * (total - 1);
    result.imageTokens = index === total - 1 ? remainder : base;
  }
  if ((details == null ? void 0 : details.text_tokens) != null) {
    const base = Math.floor(details.text_tokens / total);
    const remainder = details.text_tokens - base * (total - 1);
    result.textTokens = index === total - 1 ? remainder : base;
  }
  if (usage.input_tokens != null) {
    const base = Math.floor(usage.input_tokens / total);
    const remainder = usage.input_tokens - base * (total - 1);
    result.inputTokens = index === total - 1 ? remainder : base;
  }
  if (usage.output_tokens != null) {
    const base = Math.floor(usage.output_tokens / total);
    const remainder = usage.output_tokens - base * (total - 1);
    result.outputTokens = index === total - 1 ? remainder : base;
  }
  return result;
}
var openaiTranscriptionResponseSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      text: v4.z.string(),
      language: v4.z.string().nullish(),
      duration: v4.z.number().nullish(),
      words: v4.z.array(
        v4.z.object({
          word: v4.z.string(),
          start: v4.z.number(),
          end: v4.z.number()
        })
      ).nullish(),
      segments: v4.z.array(
        v4.z.object({
          id: v4.z.number(),
          seek: v4.z.number(),
          start: v4.z.number(),
          end: v4.z.number(),
          text: v4.z.string(),
          tokens: v4.z.array(v4.z.number()),
          temperature: v4.z.number(),
          avg_logprob: v4.z.number(),
          compression_ratio: v4.z.number(),
          no_speech_prob: v4.z.number()
        })
      ).nullish()
    })
  )
);
var openAITranscriptionProviderOptions = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      /**
       * Additional information to include in the transcription response.
       */
      include: v4.z.array(v4.z.string()).optional(),
      /**
       * The language of the input audio in ISO-639-1 format.
       */
      language: v4.z.string().optional(),
      /**
       * An optional text to guide the model's style or continue a previous audio segment.
       */
      prompt: v4.z.string().optional(),
      /**
       * The sampling temperature, between 0 and 1.
       * @default 0
       */
      temperature: v4.z.number().min(0).max(1).default(0).optional(),
      /**
       * The timestamp granularities to populate for this transcription.
       * @default ['segment']
       */
      timestampGranularities: v4.z.array(v4.z.enum(["word", "segment"])).default(["segment"]).optional()
    })
  )
);
var languageMap = {
  afrikaans: "af",
  arabic: "ar",
  armenian: "hy",
  azerbaijani: "az",
  belarusian: "be",
  bosnian: "bs",
  bulgarian: "bg",
  catalan: "ca",
  chinese: "zh",
  croatian: "hr",
  czech: "cs",
  danish: "da",
  dutch: "nl",
  english: "en",
  estonian: "et",
  finnish: "fi",
  french: "fr",
  galician: "gl",
  german: "de",
  greek: "el",
  hebrew: "he",
  hindi: "hi",
  hungarian: "hu",
  icelandic: "is",
  indonesian: "id",
  italian: "it",
  japanese: "ja",
  kannada: "kn",
  kazakh: "kk",
  korean: "ko",
  latvian: "lv",
  lithuanian: "lt",
  macedonian: "mk",
  malay: "ms",
  marathi: "mr",
  maori: "mi",
  nepali: "ne",
  norwegian: "no",
  persian: "fa",
  polish: "pl",
  portuguese: "pt",
  romanian: "ro",
  russian: "ru",
  serbian: "sr",
  slovak: "sk",
  slovenian: "sl",
  spanish: "es",
  swahili: "sw",
  swedish: "sv",
  tagalog: "tl",
  tamil: "ta",
  thai: "th",
  turkish: "tr",
  ukrainian: "uk",
  urdu: "ur",
  vietnamese: "vi",
  welsh: "cy"
};
var OpenAITranscriptionModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v2";
  }
  get provider() {
    return this.config.provider;
  }
  async getArgs({
    audio,
    mediaType,
    providerOptions
  }) {
    const warnings = [];
    const openAIOptions = await chunkA45A2T7R_cjs.parseProviderOptions({
      provider: "openai",
      providerOptions,
      schema: openAITranscriptionProviderOptions
    });
    const formData = new FormData();
    const blob = audio instanceof Uint8Array ? new Blob([audio]) : new Blob([chunkA45A2T7R_cjs.convertBase64ToUint8Array(audio)]);
    formData.append("model", this.modelId);
    const fileExtension = chunkA45A2T7R_cjs.mediaTypeToExtension(mediaType);
    formData.append(
      "file",
      new File([blob], "audio", { type: mediaType }),
      `audio.${fileExtension}`
    );
    if (openAIOptions) {
      const transcriptionModelOptions = {
        include: openAIOptions.include,
        language: openAIOptions.language,
        prompt: openAIOptions.prompt,
        // https://platform.openai.com/docs/api-reference/audio/createTranscription#audio_createtranscription-response_format
        // prefer verbose_json to get segments for models that support it
        response_format: [
          "gpt-4o-transcribe",
          "gpt-4o-mini-transcribe"
        ].includes(this.modelId) ? "json" : "verbose_json",
        temperature: openAIOptions.temperature,
        timestamp_granularities: openAIOptions.timestampGranularities
      };
      for (const [key, value] of Object.entries(transcriptionModelOptions)) {
        if (value != null) {
          if (Array.isArray(value)) {
            for (const item of value) {
              formData.append(`${key}[]`, String(item));
            }
          } else {
            formData.append(key, String(value));
          }
        }
      }
    }
    return {
      formData,
      warnings
    };
  }
  async doGenerate(options) {
    var _a, _b, _c, _d, _e, _f, _g, _h;
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const { formData, warnings } = await this.getArgs(options);
    const {
      value: response,
      responseHeaders,
      rawValue: rawResponse
    } = await chunkA45A2T7R_cjs.postFormDataToApi({
      url: this.config.url({
        path: "/audio/transcriptions",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      formData,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createJsonResponseHandler(
        openaiTranscriptionResponseSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const language = response.language != null && response.language in languageMap ? languageMap[response.language] : void 0;
    return {
      text: response.text,
      segments: (_g = (_f = (_d = response.segments) == null ? void 0 : _d.map((segment) => ({
        text: segment.text,
        startSecond: segment.start,
        endSecond: segment.end
      }))) != null ? _f : (_e = response.words) == null ? void 0 : _e.map((word) => ({
        text: word.word,
        startSecond: word.start,
        endSecond: word.end
      }))) != null ? _g : [],
      language,
      durationInSeconds: (_h = response.duration) != null ? _h : void 0,
      warnings,
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders,
        body: rawResponse
      }
    };
  }
};
var openaiSpeechProviderOptionsSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      instructions: v4.z.string().nullish(),
      speed: v4.z.number().min(0.25).max(4).default(1).nullish()
    })
  )
);
var OpenAISpeechModel = class {
  constructor(modelId, config) {
    this.modelId = modelId;
    this.config = config;
    this.specificationVersion = "v2";
  }
  get provider() {
    return this.config.provider;
  }
  async getArgs({
    text,
    voice = "alloy",
    outputFormat = "mp3",
    speed,
    instructions,
    language,
    providerOptions
  }) {
    const warnings = [];
    const openAIOptions = await chunkA45A2T7R_cjs.parseProviderOptions({
      provider: "openai",
      providerOptions,
      schema: openaiSpeechProviderOptionsSchema
    });
    const requestBody = {
      model: this.modelId,
      input: text,
      voice,
      response_format: "mp3",
      speed,
      instructions
    };
    if (outputFormat) {
      if (["mp3", "opus", "aac", "flac", "wav", "pcm"].includes(outputFormat)) {
        requestBody.response_format = outputFormat;
      } else {
        warnings.push({
          type: "unsupported-setting",
          setting: "outputFormat",
          details: `Unsupported output format: ${outputFormat}. Using mp3 instead.`
        });
      }
    }
    if (openAIOptions) {
      const speechModelOptions = {};
      for (const key in speechModelOptions) {
        const value = speechModelOptions[key];
        if (value !== void 0) {
          requestBody[key] = value;
        }
      }
    }
    if (language) {
      warnings.push({
        type: "unsupported-setting",
        setting: "language",
        details: `OpenAI speech models do not support language selection. Language parameter "${language}" was ignored.`
      });
    }
    return {
      requestBody,
      warnings
    };
  }
  async doGenerate(options) {
    var _a, _b, _c;
    const currentDate = (_c = (_b = (_a = this.config._internal) == null ? void 0 : _a.currentDate) == null ? void 0 : _b.call(_a)) != null ? _c : /* @__PURE__ */ new Date();
    const { requestBody, warnings } = await this.getArgs(options);
    const {
      value: audio,
      responseHeaders,
      rawValue: rawResponse
    } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/audio/speech",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body: requestBody,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createBinaryResponseHandler(),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    return {
      audio,
      warnings,
      request: {
        body: JSON.stringify(requestBody)
      },
      response: {
        timestamp: currentDate,
        modelId: this.modelId,
        headers: responseHeaders,
        body: rawResponse
      }
    };
  }
};
var localShellInputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      action: v4.z.object({
        type: v4.z.literal("exec"),
        command: v4.z.array(v4.z.string()),
        timeoutMs: v4.z.number().optional(),
        user: v4.z.string().optional(),
        workingDirectory: v4.z.string().optional(),
        env: v4.z.record(v4.z.string(), v4.z.string()).optional()
      })
    })
  )
);
var localShellOutputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(v4.z.object({ output: v4.z.string() }))
);
chunkA45A2T7R_cjs.createProviderDefinedToolFactoryWithOutputSchema({
  id: "openai.local_shell",
  name: "local_shell",
  inputSchema: localShellInputSchema,
  outputSchema: localShellOutputSchema
});
function isFileId(data, prefixes) {
  if (!prefixes) return false;
  return prefixes.some((prefix) => data.startsWith(prefix));
}
async function convertToOpenAIResponsesInput({
  prompt,
  systemMessageMode,
  fileIdPrefixes,
  store,
  hasLocalShellTool = false
}) {
  var _a, _b, _c, _d, _e, _f;
  let input = [];
  const warnings = [];
  for (const { role, content } of prompt) {
    switch (role) {
      case "system": {
        switch (systemMessageMode) {
          case "system": {
            input.push({ role: "system", content });
            break;
          }
          case "developer": {
            input.push({ role: "developer", content });
            break;
          }
          case "remove": {
            warnings.push({
              type: "other",
              message: "system messages are removed for this model"
            });
            break;
          }
          default: {
            const _exhaustiveCheck = systemMessageMode;
            throw new Error(
              `Unsupported system message mode: ${_exhaustiveCheck}`
            );
          }
        }
        break;
      }
      case "user": {
        input.push({
          role: "user",
          content: content.map((part, index) => {
            var _a2, _b2, _c2;
            switch (part.type) {
              case "text": {
                return { type: "input_text", text: part.text };
              }
              case "file": {
                if (part.mediaType.startsWith("image/")) {
                  const mediaType = part.mediaType === "image/*" ? "image/jpeg" : part.mediaType;
                  return {
                    type: "input_image",
                    ...part.data instanceof URL ? { image_url: part.data.toString() } : typeof part.data === "string" && isFileId(part.data, fileIdPrefixes) ? { file_id: part.data } : {
                      image_url: `data:${mediaType};base64,${chunkA45A2T7R_cjs.convertToBase64(part.data)}`
                    },
                    detail: (_b2 = (_a2 = part.providerOptions) == null ? void 0 : _a2.openai) == null ? void 0 : _b2.imageDetail
                  };
                } else if (part.mediaType === "application/pdf") {
                  if (part.data instanceof URL) {
                    return {
                      type: "input_file",
                      file_url: part.data.toString()
                    };
                  }
                  return {
                    type: "input_file",
                    ...typeof part.data === "string" && isFileId(part.data, fileIdPrefixes) ? { file_id: part.data } : {
                      filename: (_c2 = part.filename) != null ? _c2 : `part-${index}.pdf`,
                      file_data: `data:application/pdf;base64,${chunkA45A2T7R_cjs.convertToBase64(part.data)}`
                    }
                  };
                } else {
                  throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
                    functionality: `file part media type ${part.mediaType}`
                  });
                }
              }
            }
          })
        });
        break;
      }
      case "assistant": {
        const reasoningMessages = {};
        for (const part of content) {
          switch (part.type) {
            case "text": {
              const id = (_b = (_a = part.providerOptions) == null ? void 0 : _a.openai) == null ? void 0 : _b.itemId;
              const phase = (_d = (_c = part.providerOptions) == null ? void 0 : _c.openai) == null ? void 0 : _d.phase;
              if (store && id != null) {
                input.push({ type: "item_reference", id });
                break;
              }
              input.push({
                role: "assistant",
                content: [{ type: "output_text", text: part.text }],
                id,
                ...phase != null && { phase }
              });
              break;
            }
            case "tool-call": {
              if (part.providerExecuted) {
                break;
              }
              const id = (_f = (_e = part.providerOptions) == null ? void 0 : _e.openai) == null ? void 0 : _f.itemId;
              if (store && id != null) {
                input.push({ type: "item_reference", id });
                break;
              }
              if (hasLocalShellTool && part.toolName === "local_shell") {
                const parsedInput = await chunkA45A2T7R_cjs.validateTypes({
                  value: part.input,
                  schema: localShellInputSchema
                });
                input.push({
                  type: "local_shell_call",
                  call_id: part.toolCallId,
                  id,
                  action: {
                    type: "exec",
                    command: parsedInput.action.command,
                    timeout_ms: parsedInput.action.timeoutMs,
                    user: parsedInput.action.user,
                    working_directory: parsedInput.action.workingDirectory,
                    env: parsedInput.action.env
                  }
                });
                break;
              }
              input.push({
                type: "function_call",
                call_id: part.toolCallId,
                name: part.toolName,
                arguments: JSON.stringify(part.input),
                id
              });
              break;
            }
            // assistant tool result parts are from provider-executed tools:
            case "tool-result": {
              if (store) {
                input.push({ type: "item_reference", id: part.toolCallId });
              } else {
                warnings.push({
                  type: "other",
                  message: `Results for OpenAI tool ${part.toolName} are not sent to the API when store is false`
                });
              }
              break;
            }
            case "reasoning": {
              const providerOptions = await chunkA45A2T7R_cjs.parseProviderOptions({
                provider: "openai",
                providerOptions: part.providerOptions,
                schema: openaiResponsesReasoningProviderOptionsSchema
              });
              const reasoningId = providerOptions == null ? void 0 : providerOptions.itemId;
              if (reasoningId != null) {
                const reasoningMessage = reasoningMessages[reasoningId];
                if (store) {
                  if (reasoningMessage === void 0) {
                    input.push({ type: "item_reference", id: reasoningId });
                    reasoningMessages[reasoningId] = {
                      type: "reasoning",
                      id: reasoningId,
                      summary: []
                    };
                  }
                } else {
                  const summaryParts = [];
                  if (part.text.length > 0) {
                    summaryParts.push({
                      type: "summary_text",
                      text: part.text
                    });
                  } else if (reasoningMessage !== void 0) {
                    warnings.push({
                      type: "other",
                      message: `Cannot append empty reasoning part to existing reasoning sequence. Skipping reasoning part: ${JSON.stringify(part)}.`
                    });
                  }
                  if (reasoningMessage === void 0) {
                    reasoningMessages[reasoningId] = {
                      type: "reasoning",
                      id: reasoningId,
                      encrypted_content: providerOptions == null ? void 0 : providerOptions.reasoningEncryptedContent,
                      summary: summaryParts
                    };
                    input.push(reasoningMessages[reasoningId]);
                  } else {
                    reasoningMessage.summary.push(...summaryParts);
                    if ((providerOptions == null ? void 0 : providerOptions.reasoningEncryptedContent) != null) {
                      reasoningMessage.encrypted_content = providerOptions.reasoningEncryptedContent;
                    }
                  }
                }
              } else {
                warnings.push({
                  type: "other",
                  message: `Non-OpenAI reasoning parts are not supported. Skipping reasoning part: ${JSON.stringify(part)}.`
                });
              }
              break;
            }
          }
        }
        break;
      }
      case "tool": {
        for (const part of content) {
          const output = part.output;
          if (hasLocalShellTool && part.toolName === "local_shell" && output.type === "json") {
            const parsedOutput = await chunkA45A2T7R_cjs.validateTypes({
              value: output.value,
              schema: localShellOutputSchema
            });
            input.push({
              type: "local_shell_call_output",
              call_id: part.toolCallId,
              output: parsedOutput.output
            });
            break;
          }
          let contentValue;
          switch (output.type) {
            case "text":
            case "error-text":
              contentValue = output.value;
              break;
            case "json":
            case "error-json":
              contentValue = JSON.stringify(output.value);
              break;
            case "content":
              contentValue = output.value.map((item) => {
                switch (item.type) {
                  case "text": {
                    return { type: "input_text", text: item.text };
                  }
                  case "media": {
                    return item.mediaType.startsWith("image/") ? {
                      type: "input_image",
                      image_url: `data:${item.mediaType};base64,${item.data}`
                    } : {
                      type: "input_file",
                      filename: "data",
                      file_data: `data:${item.mediaType};base64,${item.data}`
                    };
                  }
                }
              });
              break;
          }
          input.push({
            type: "function_call_output",
            call_id: part.toolCallId,
            output: contentValue
          });
        }
        break;
      }
      default: {
        const _exhaustiveCheck = role;
        throw new Error(`Unsupported role: ${_exhaustiveCheck}`);
      }
    }
  }
  if (!store && input.some(
    (item) => "type" in item && item.type === "reasoning" && item.encrypted_content == null
  )) {
    warnings.push({
      type: "other",
      message: "Reasoning parts without encrypted content are not supported when store is false. Skipping reasoning parts."
    });
    input = input.filter(
      (item) => !("type" in item) || item.type !== "reasoning" || item.encrypted_content != null
    );
  }
  return { input, warnings };
}
var openaiResponsesReasoningProviderOptionsSchema = v4.z.object({
  itemId: v4.z.string().nullish(),
  reasoningEncryptedContent: v4.z.string().nullish()
});
function mapOpenAIResponseFinishReason({
  finishReason,
  hasFunctionCall
}) {
  switch (finishReason) {
    case void 0:
    case null:
      return hasFunctionCall ? "tool-calls" : "stop";
    case "max_output_tokens":
      return "length";
    case "content_filter":
      return "content-filter";
    default:
      return hasFunctionCall ? "tool-calls" : "unknown";
  }
}
var openaiResponsesChunkSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.union([
      v4.z.object({
        type: v4.z.literal("response.output_text.delta"),
        item_id: v4.z.string(),
        delta: v4.z.string(),
        logprobs: v4.z.array(
          v4.z.object({
            token: v4.z.string(),
            logprob: v4.z.number(),
            top_logprobs: v4.z.array(
              v4.z.object({
                token: v4.z.string(),
                logprob: v4.z.number()
              })
            )
          })
        ).nullish()
      }),
      v4.z.object({
        type: v4.z.enum(["response.completed", "response.incomplete"]),
        response: v4.z.object({
          incomplete_details: v4.z.object({ reason: v4.z.string() }).nullish(),
          usage: v4.z.object({
            input_tokens: v4.z.number(),
            input_tokens_details: v4.z.object({ cached_tokens: v4.z.number().nullish() }).nullish(),
            output_tokens: v4.z.number(),
            output_tokens_details: v4.z.object({ reasoning_tokens: v4.z.number().nullish() }).nullish()
          }),
          service_tier: v4.z.string().nullish()
        })
      }),
      v4.z.object({
        type: v4.z.literal("response.created"),
        response: v4.z.object({
          id: v4.z.string(),
          created_at: v4.z.number(),
          model: v4.z.string(),
          service_tier: v4.z.string().nullish()
        })
      }),
      v4.z.object({
        type: v4.z.literal("response.output_item.added"),
        output_index: v4.z.number(),
        item: v4.z.discriminatedUnion("type", [
          v4.z.object({
            type: v4.z.literal("message"),
            id: v4.z.string(),
            phase: v4.z.enum(["commentary", "final_answer"]).nullish()
          }),
          v4.z.object({
            type: v4.z.literal("reasoning"),
            id: v4.z.string(),
            encrypted_content: v4.z.string().nullish()
          }),
          v4.z.object({
            type: v4.z.literal("function_call"),
            id: v4.z.string(),
            call_id: v4.z.string(),
            name: v4.z.string(),
            arguments: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("web_search_call"),
            id: v4.z.string(),
            status: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("computer_call"),
            id: v4.z.string(),
            status: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("file_search_call"),
            id: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("image_generation_call"),
            id: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("code_interpreter_call"),
            id: v4.z.string(),
            container_id: v4.z.string(),
            code: v4.z.string().nullable(),
            outputs: v4.z.array(
              v4.z.discriminatedUnion("type", [
                v4.z.object({ type: v4.z.literal("logs"), logs: v4.z.string() }),
                v4.z.object({ type: v4.z.literal("image"), url: v4.z.string() })
              ])
            ).nullable(),
            status: v4.z.string()
          })
        ])
      }),
      v4.z.object({
        type: v4.z.literal("response.output_item.done"),
        output_index: v4.z.number(),
        item: v4.z.discriminatedUnion("type", [
          v4.z.object({
            type: v4.z.literal("message"),
            id: v4.z.string(),
            phase: v4.z.enum(["commentary", "final_answer"]).nullish()
          }),
          v4.z.object({
            type: v4.z.literal("reasoning"),
            id: v4.z.string(),
            encrypted_content: v4.z.string().nullish()
          }),
          v4.z.object({
            type: v4.z.literal("function_call"),
            id: v4.z.string(),
            call_id: v4.z.string(),
            name: v4.z.string(),
            arguments: v4.z.string(),
            status: v4.z.literal("completed")
          }),
          v4.z.object({
            type: v4.z.literal("code_interpreter_call"),
            id: v4.z.string(),
            code: v4.z.string().nullable(),
            container_id: v4.z.string(),
            outputs: v4.z.array(
              v4.z.discriminatedUnion("type", [
                v4.z.object({ type: v4.z.literal("logs"), logs: v4.z.string() }),
                v4.z.object({ type: v4.z.literal("image"), url: v4.z.string() })
              ])
            ).nullable()
          }),
          v4.z.object({
            type: v4.z.literal("image_generation_call"),
            id: v4.z.string(),
            result: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("web_search_call"),
            id: v4.z.string(),
            status: v4.z.string(),
            action: v4.z.discriminatedUnion("type", [
              v4.z.object({
                type: v4.z.literal("search"),
                query: v4.z.string().nullish(),
                sources: v4.z.array(
                  v4.z.discriminatedUnion("type", [
                    v4.z.object({ type: v4.z.literal("url"), url: v4.z.string() }),
                    v4.z.object({ type: v4.z.literal("api"), name: v4.z.string() })
                  ])
                ).nullish()
              }),
              v4.z.object({
                type: v4.z.literal("open_page"),
                url: v4.z.string().nullish()
              }),
              v4.z.object({
                type: v4.z.literal("find_in_page"),
                url: v4.z.string().nullish(),
                pattern: v4.z.string().nullish()
              })
            ]).nullish()
          }),
          v4.z.object({
            type: v4.z.literal("file_search_call"),
            id: v4.z.string(),
            queries: v4.z.array(v4.z.string()),
            results: v4.z.array(
              v4.z.object({
                attributes: v4.z.record(v4.z.string(), v4.z.unknown()),
                file_id: v4.z.string(),
                filename: v4.z.string(),
                score: v4.z.number(),
                text: v4.z.string()
              })
            ).nullish()
          }),
          v4.z.object({
            type: v4.z.literal("local_shell_call"),
            id: v4.z.string(),
            call_id: v4.z.string(),
            action: v4.z.object({
              type: v4.z.literal("exec"),
              command: v4.z.array(v4.z.string()),
              timeout_ms: v4.z.number().optional(),
              user: v4.z.string().optional(),
              working_directory: v4.z.string().optional(),
              env: v4.z.record(v4.z.string(), v4.z.string()).optional()
            })
          }),
          v4.z.object({
            type: v4.z.literal("computer_call"),
            id: v4.z.string(),
            status: v4.z.literal("completed")
          })
        ])
      }),
      v4.z.object({
        type: v4.z.literal("response.function_call_arguments.delta"),
        item_id: v4.z.string(),
        output_index: v4.z.number(),
        delta: v4.z.string()
      }),
      v4.z.object({
        type: v4.z.literal("response.image_generation_call.partial_image"),
        item_id: v4.z.string(),
        output_index: v4.z.number(),
        partial_image_b64: v4.z.string()
      }),
      v4.z.object({
        type: v4.z.literal("response.code_interpreter_call_code.delta"),
        item_id: v4.z.string(),
        output_index: v4.z.number(),
        delta: v4.z.string()
      }),
      v4.z.object({
        type: v4.z.literal("response.code_interpreter_call_code.done"),
        item_id: v4.z.string(),
        output_index: v4.z.number(),
        code: v4.z.string()
      }),
      v4.z.object({
        type: v4.z.literal("response.output_text.annotation.added"),
        annotation: v4.z.discriminatedUnion("type", [
          v4.z.object({
            type: v4.z.literal("url_citation"),
            start_index: v4.z.number(),
            end_index: v4.z.number(),
            url: v4.z.string(),
            title: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("file_citation"),
            file_id: v4.z.string(),
            filename: v4.z.string().nullish(),
            index: v4.z.number().nullish(),
            start_index: v4.z.number().nullish(),
            end_index: v4.z.number().nullish(),
            quote: v4.z.string().nullish()
          })
        ])
      }),
      v4.z.object({
        type: v4.z.literal("response.reasoning_summary_part.added"),
        item_id: v4.z.string(),
        summary_index: v4.z.number()
      }),
      v4.z.object({
        type: v4.z.literal("response.reasoning_summary_text.delta"),
        item_id: v4.z.string(),
        summary_index: v4.z.number(),
        delta: v4.z.string()
      }),
      v4.z.object({
        type: v4.z.literal("response.reasoning_summary_part.done"),
        item_id: v4.z.string(),
        summary_index: v4.z.number()
      }),
      v4.z.object({
        type: v4.z.literal("error"),
        sequence_number: v4.z.number(),
        error: v4.z.object({
          type: v4.z.string(),
          code: v4.z.string(),
          message: v4.z.string(),
          param: v4.z.string().nullish()
        })
      }),
      v4.z.object({ type: v4.z.string() }).loose().transform((value) => ({
        type: "unknown_chunk",
        message: value.type
      }))
      // fallback for unknown chunks
    ])
  )
);
var openaiResponsesResponseSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      id: v4.z.string().optional(),
      created_at: v4.z.number().optional(),
      error: v4.z.object({
        message: v4.z.string(),
        type: v4.z.string(),
        param: v4.z.string().nullish(),
        code: v4.z.string()
      }).nullish(),
      model: v4.z.string().optional(),
      output: v4.z.array(
        v4.z.discriminatedUnion("type", [
          v4.z.object({
            type: v4.z.literal("message"),
            role: v4.z.literal("assistant"),
            id: v4.z.string(),
            phase: v4.z.enum(["commentary", "final_answer"]).nullish(),
            content: v4.z.array(
              v4.z.object({
                type: v4.z.literal("output_text"),
                text: v4.z.string(),
                logprobs: v4.z.array(
                  v4.z.object({
                    token: v4.z.string(),
                    logprob: v4.z.number(),
                    top_logprobs: v4.z.array(
                      v4.z.object({
                        token: v4.z.string(),
                        logprob: v4.z.number()
                      })
                    )
                  })
                ).nullish(),
                annotations: v4.z.array(
                  v4.z.discriminatedUnion("type", [
                    v4.z.object({
                      type: v4.z.literal("url_citation"),
                      start_index: v4.z.number(),
                      end_index: v4.z.number(),
                      url: v4.z.string(),
                      title: v4.z.string()
                    }),
                    v4.z.object({
                      type: v4.z.literal("file_citation"),
                      file_id: v4.z.string(),
                      filename: v4.z.string().nullish(),
                      index: v4.z.number().nullish(),
                      start_index: v4.z.number().nullish(),
                      end_index: v4.z.number().nullish(),
                      quote: v4.z.string().nullish()
                    }),
                    v4.z.object({
                      type: v4.z.literal("container_file_citation"),
                      container_id: v4.z.string(),
                      file_id: v4.z.string(),
                      filename: v4.z.string().nullish(),
                      start_index: v4.z.number().nullish(),
                      end_index: v4.z.number().nullish(),
                      index: v4.z.number().nullish()
                    }),
                    v4.z.object({
                      type: v4.z.literal("file_path"),
                      file_id: v4.z.string(),
                      index: v4.z.number().nullish()
                    })
                  ])
                )
              })
            )
          }),
          v4.z.object({
            type: v4.z.literal("web_search_call"),
            id: v4.z.string(),
            status: v4.z.string(),
            action: v4.z.discriminatedUnion("type", [
              v4.z.object({
                type: v4.z.literal("search"),
                query: v4.z.string().nullish(),
                sources: v4.z.array(
                  v4.z.discriminatedUnion("type", [
                    v4.z.object({ type: v4.z.literal("url"), url: v4.z.string() }),
                    v4.z.object({
                      type: v4.z.literal("api"),
                      name: v4.z.string()
                    })
                  ])
                ).nullish()
              }),
              v4.z.object({
                type: v4.z.literal("open_page"),
                url: v4.z.string().nullish()
              }),
              v4.z.object({
                type: v4.z.literal("find_in_page"),
                url: v4.z.string().nullish(),
                pattern: v4.z.string().nullish()
              })
            ]).nullish()
          }),
          v4.z.object({
            type: v4.z.literal("file_search_call"),
            id: v4.z.string(),
            queries: v4.z.array(v4.z.string()),
            results: v4.z.array(
              v4.z.object({
                attributes: v4.z.record(
                  v4.z.string(),
                  v4.z.union([v4.z.string(), v4.z.number(), v4.z.boolean()])
                ),
                file_id: v4.z.string(),
                filename: v4.z.string(),
                score: v4.z.number(),
                text: v4.z.string()
              })
            ).nullish()
          }),
          v4.z.object({
            type: v4.z.literal("code_interpreter_call"),
            id: v4.z.string(),
            code: v4.z.string().nullable(),
            container_id: v4.z.string(),
            outputs: v4.z.array(
              v4.z.discriminatedUnion("type", [
                v4.z.object({ type: v4.z.literal("logs"), logs: v4.z.string() }),
                v4.z.object({ type: v4.z.literal("image"), url: v4.z.string() })
              ])
            ).nullable()
          }),
          v4.z.object({
            type: v4.z.literal("image_generation_call"),
            id: v4.z.string(),
            result: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("local_shell_call"),
            id: v4.z.string(),
            call_id: v4.z.string(),
            action: v4.z.object({
              type: v4.z.literal("exec"),
              command: v4.z.array(v4.z.string()),
              timeout_ms: v4.z.number().optional(),
              user: v4.z.string().optional(),
              working_directory: v4.z.string().optional(),
              env: v4.z.record(v4.z.string(), v4.z.string()).optional()
            })
          }),
          v4.z.object({
            type: v4.z.literal("function_call"),
            call_id: v4.z.string(),
            name: v4.z.string(),
            arguments: v4.z.string(),
            id: v4.z.string()
          }),
          v4.z.object({
            type: v4.z.literal("computer_call"),
            id: v4.z.string(),
            status: v4.z.string().optional()
          }),
          v4.z.object({
            type: v4.z.literal("reasoning"),
            id: v4.z.string(),
            encrypted_content: v4.z.string().nullish(),
            summary: v4.z.array(
              v4.z.object({
                type: v4.z.literal("summary_text"),
                text: v4.z.string()
              })
            )
          })
        ])
      ).optional(),
      service_tier: v4.z.string().nullish(),
      incomplete_details: v4.z.object({ reason: v4.z.string() }).nullish(),
      usage: v4.z.object({
        input_tokens: v4.z.number(),
        input_tokens_details: v4.z.object({ cached_tokens: v4.z.number().nullish() }).nullish(),
        output_tokens: v4.z.number(),
        output_tokens_details: v4.z.object({ reasoning_tokens: v4.z.number().nullish() }).nullish()
      }).optional()
    })
  )
);
var TOP_LOGPROBS_MAX = 20;
var openaiResponsesProviderOptionsSchema = chunkA45A2T7R_cjs.lazyValidator(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      conversation: v4.z.string().nullish(),
      include: v4.z.array(
        v4.z.enum([
          "reasoning.encrypted_content",
          // handled internally by default, only needed for unknown reasoning models
          "file_search_call.results",
          "message.output_text.logprobs"
        ])
      ).nullish(),
      instructions: v4.z.string().nullish(),
      /**
       * Return the log probabilities of the tokens.
       *
       * Setting to true will return the log probabilities of the tokens that
       * were generated.
       *
       * Setting to a number will return the log probabilities of the top n
       * tokens that were generated.
       *
       * @see https://platform.openai.com/docs/api-reference/responses/create
       * @see https://cookbook.openai.com/examples/using_logprobs
       */
      logprobs: v4.z.union([v4.z.boolean(), v4.z.number().min(1).max(TOP_LOGPROBS_MAX)]).optional(),
      /**
       * The maximum number of total calls to built-in tools that can be processed in a response.
       * This maximum number applies across all built-in tool calls, not per individual tool.
       * Any further attempts to call a tool by the model will be ignored.
       */
      maxToolCalls: v4.z.number().nullish(),
      metadata: v4.z.any().nullish(),
      parallelToolCalls: v4.z.boolean().nullish(),
      previousResponseId: v4.z.string().nullish(),
      promptCacheKey: v4.z.string().nullish(),
      /**
       * The retention policy for the prompt cache.
       * - 'in_memory': Default. Standard prompt caching behavior.
       * - '24h': Extended prompt caching that keeps cached prefixes active for up to 24 hours.
       *          Currently only available for 5.1 series models.
       *
       * @default 'in_memory'
       */
      promptCacheRetention: v4.z.enum(["in_memory", "24h"]).nullish(),
      /**
       * Reasoning effort for reasoning models. Defaults to `medium`. If you use
       * `providerOptions` to set the `reasoningEffort` option, this model setting will be ignored.
       * Valid values: 'none' | 'minimal' | 'low' | 'medium' | 'high' | 'xhigh'
       *
       * The 'none' type for `reasoningEffort` is only available for OpenAI's GPT-5.1
       * models. Also, the 'xhigh' type for `reasoningEffort` is only available for
       * OpenAI's GPT-5.1-Codex-Max model. Setting `reasoningEffort` to 'none' or 'xhigh' with unsupported models will result in
       * an error.
       */
      reasoningEffort: v4.z.string().nullish(),
      reasoningSummary: v4.z.string().nullish(),
      safetyIdentifier: v4.z.string().nullish(),
      serviceTier: v4.z.enum(["auto", "flex", "priority", "default"]).nullish(),
      store: v4.z.boolean().nullish(),
      strictJsonSchema: v4.z.boolean().nullish(),
      textVerbosity: v4.z.enum(["low", "medium", "high"]).nullish(),
      truncation: v4.z.enum(["auto", "disabled"]).nullish(),
      user: v4.z.string().nullish()
    })
  )
);
var codeInterpreterInputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      code: v4.z.string().nullish(),
      containerId: v4.z.string()
    })
  )
);
var codeInterpreterOutputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      outputs: v4.z.array(
        v4.z.discriminatedUnion("type", [
          v4.z.object({ type: v4.z.literal("logs"), logs: v4.z.string() }),
          v4.z.object({ type: v4.z.literal("image"), url: v4.z.string() })
        ])
      ).nullish()
    })
  )
);
var codeInterpreterArgsSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      container: v4.z.union([
        v4.z.string(),
        v4.z.object({
          fileIds: v4.z.array(v4.z.string()).optional()
        })
      ]).optional()
    })
  )
);
var codeInterpreterToolFactory = chunkA45A2T7R_cjs.createProviderDefinedToolFactoryWithOutputSchema({
  id: "openai.code_interpreter",
  name: "code_interpreter",
  inputSchema: codeInterpreterInputSchema,
  outputSchema: codeInterpreterOutputSchema
});
var codeInterpreter = (args = {}) => {
  return codeInterpreterToolFactory(args);
};
var comparisonFilterSchema = v4.z.object({
  key: v4.z.string(),
  type: v4.z.enum(["eq", "ne", "gt", "gte", "lt", "lte", "in", "nin"]),
  value: v4.z.union([v4.z.string(), v4.z.number(), v4.z.boolean(), v4.z.array(v4.z.string())])
});
var compoundFilterSchema = v4.z.object({
  type: v4.z.enum(["and", "or"]),
  filters: v4.z.array(
    v4.z.union([comparisonFilterSchema, v4.z.lazy(() => compoundFilterSchema)])
  )
});
var fileSearchArgsSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      vectorStoreIds: v4.z.array(v4.z.string()),
      maxNumResults: v4.z.number().optional(),
      ranking: v4.z.object({
        ranker: v4.z.string().optional(),
        scoreThreshold: v4.z.number().optional()
      }).optional(),
      filters: v4.z.union([comparisonFilterSchema, compoundFilterSchema]).optional()
    })
  )
);
var fileSearchOutputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      queries: v4.z.array(v4.z.string()),
      results: v4.z.array(
        v4.z.object({
          attributes: v4.z.record(v4.z.string(), v4.z.unknown()),
          fileId: v4.z.string(),
          filename: v4.z.string(),
          score: v4.z.number(),
          text: v4.z.string()
        })
      ).nullable()
    })
  )
);
var fileSearch = chunkA45A2T7R_cjs.createProviderDefinedToolFactoryWithOutputSchema({
  id: "openai.file_search",
  name: "file_search",
  inputSchema: v4.z.object({}),
  outputSchema: fileSearchOutputSchema
});
var webSearchArgsSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      externalWebAccess: v4.z.boolean().optional(),
      filters: v4.z.object({ allowedDomains: v4.z.array(v4.z.string()).optional() }).optional(),
      searchContextSize: v4.z.enum(["low", "medium", "high"]).optional(),
      userLocation: v4.z.object({
        type: v4.z.literal("approximate"),
        country: v4.z.string().optional(),
        city: v4.z.string().optional(),
        region: v4.z.string().optional(),
        timezone: v4.z.string().optional()
      }).optional()
    })
  )
);
var webSearchInputSchema = chunkA45A2T7R_cjs.lazySchema(() => chunkA45A2T7R_cjs.zodSchema(v4.z.object({})));
var webSearchOutputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      action: v4.z.discriminatedUnion("type", [
        v4.z.object({
          type: v4.z.literal("search"),
          query: v4.z.string().optional()
        }),
        v4.z.object({
          type: v4.z.literal("openPage"),
          url: v4.z.string().nullish()
        }),
        v4.z.object({
          type: v4.z.literal("findInPage"),
          url: v4.z.string().nullish(),
          pattern: v4.z.string().nullish()
        })
      ]).optional(),
      sources: v4.z.array(
        v4.z.discriminatedUnion("type", [
          v4.z.object({ type: v4.z.literal("url"), url: v4.z.string() }),
          v4.z.object({ type: v4.z.literal("api"), name: v4.z.string() })
        ])
      ).optional()
    })
  )
);
chunkA45A2T7R_cjs.createProviderDefinedToolFactoryWithOutputSchema({
  id: "openai.web_search",
  name: "web_search",
  inputSchema: webSearchInputSchema,
  outputSchema: webSearchOutputSchema
});
var webSearchPreviewArgsSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      searchContextSize: v4.z.enum(["low", "medium", "high"]).optional(),
      userLocation: v4.z.object({
        type: v4.z.literal("approximate"),
        country: v4.z.string().optional(),
        city: v4.z.string().optional(),
        region: v4.z.string().optional(),
        timezone: v4.z.string().optional()
      }).optional()
    })
  )
);
var webSearchPreviewInputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(v4.z.object({}))
);
var webSearchPreviewOutputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      action: v4.z.discriminatedUnion("type", [
        v4.z.object({
          type: v4.z.literal("search"),
          query: v4.z.string().optional()
        }),
        v4.z.object({
          type: v4.z.literal("openPage"),
          url: v4.z.string().nullish()
        }),
        v4.z.object({
          type: v4.z.literal("findInPage"),
          url: v4.z.string().nullish(),
          pattern: v4.z.string().nullish()
        })
      ]).optional()
    })
  )
);
var webSearchPreview = chunkA45A2T7R_cjs.createProviderDefinedToolFactoryWithOutputSchema({
  id: "openai.web_search_preview",
  name: "web_search_preview",
  inputSchema: webSearchPreviewInputSchema,
  outputSchema: webSearchPreviewOutputSchema
});
var imageGenerationArgsSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(
    v4.z.object({
      background: v4.z.enum(["auto", "opaque", "transparent"]).optional(),
      inputFidelity: v4.z.enum(["low", "high"]).optional(),
      inputImageMask: v4.z.object({
        fileId: v4.z.string().optional(),
        imageUrl: v4.z.string().optional()
      }).optional(),
      model: v4.z.string().optional(),
      moderation: v4.z.enum(["auto"]).optional(),
      outputCompression: v4.z.number().int().min(0).max(100).optional(),
      outputFormat: v4.z.enum(["png", "jpeg", "webp"]).optional(),
      partialImages: v4.z.number().int().min(0).max(3).optional(),
      quality: v4.z.enum(["auto", "low", "medium", "high"]).optional(),
      size: v4.z.enum(["1024x1024", "1024x1536", "1536x1024", "auto"]).optional()
    }).strict()
  )
);
var imageGenerationInputSchema = chunkA45A2T7R_cjs.lazySchema(() => chunkA45A2T7R_cjs.zodSchema(v4.z.object({})));
var imageGenerationOutputSchema = chunkA45A2T7R_cjs.lazySchema(
  () => chunkA45A2T7R_cjs.zodSchema(v4.z.object({ result: v4.z.string() }))
);
var imageGenerationToolFactory = chunkA45A2T7R_cjs.createProviderDefinedToolFactoryWithOutputSchema({
  id: "openai.image_generation",
  name: "image_generation",
  inputSchema: imageGenerationInputSchema,
  outputSchema: imageGenerationOutputSchema
});
var imageGeneration = (args = {}) => {
  return imageGenerationToolFactory(args);
};
async function prepareResponsesTools({
  tools,
  toolChoice,
  strictJsonSchema
}) {
  tools = (tools == null ? void 0 : tools.length) ? tools : void 0;
  const toolWarnings = [];
  if (tools == null) {
    return { tools: void 0, toolChoice: void 0, toolWarnings };
  }
  const openaiTools = [];
  for (const tool of tools) {
    switch (tool.type) {
      case "function":
        openaiTools.push({
          type: "function",
          name: tool.name,
          description: tool.description,
          parameters: tool.inputSchema,
          strict: strictJsonSchema
        });
        break;
      case "provider-defined": {
        switch (tool.id) {
          case "openai.file_search": {
            const args = await chunkA45A2T7R_cjs.validateTypes({
              value: tool.args,
              schema: fileSearchArgsSchema
            });
            openaiTools.push({
              type: "file_search",
              vector_store_ids: args.vectorStoreIds,
              max_num_results: args.maxNumResults,
              ranking_options: args.ranking ? {
                ranker: args.ranking.ranker,
                score_threshold: args.ranking.scoreThreshold
              } : void 0,
              filters: args.filters
            });
            break;
          }
          case "openai.local_shell": {
            openaiTools.push({
              type: "local_shell"
            });
            break;
          }
          case "openai.web_search_preview": {
            const args = await chunkA45A2T7R_cjs.validateTypes({
              value: tool.args,
              schema: webSearchPreviewArgsSchema
            });
            openaiTools.push({
              type: "web_search_preview",
              search_context_size: args.searchContextSize,
              user_location: args.userLocation
            });
            break;
          }
          case "openai.web_search": {
            const args = await chunkA45A2T7R_cjs.validateTypes({
              value: tool.args,
              schema: webSearchArgsSchema
            });
            openaiTools.push({
              type: "web_search",
              filters: args.filters != null ? { allowed_domains: args.filters.allowedDomains } : void 0,
              external_web_access: args.externalWebAccess,
              search_context_size: args.searchContextSize,
              user_location: args.userLocation
            });
            break;
          }
          case "openai.code_interpreter": {
            const args = await chunkA45A2T7R_cjs.validateTypes({
              value: tool.args,
              schema: codeInterpreterArgsSchema
            });
            openaiTools.push({
              type: "code_interpreter",
              container: args.container == null ? { type: "auto", file_ids: void 0 } : typeof args.container === "string" ? args.container : { type: "auto", file_ids: args.container.fileIds }
            });
            break;
          }
          case "openai.image_generation": {
            const args = await chunkA45A2T7R_cjs.validateTypes({
              value: tool.args,
              schema: imageGenerationArgsSchema
            });
            openaiTools.push({
              type: "image_generation",
              background: args.background,
              input_fidelity: args.inputFidelity,
              input_image_mask: args.inputImageMask ? {
                file_id: args.inputImageMask.fileId,
                image_url: args.inputImageMask.imageUrl
              } : void 0,
              model: args.model,
              size: args.size,
              quality: args.quality,
              moderation: args.moderation,
              output_format: args.outputFormat,
              output_compression: args.outputCompression
            });
            break;
          }
        }
        break;
      }
      default:
        toolWarnings.push({ type: "unsupported-tool", tool });
        break;
    }
  }
  if (toolChoice == null) {
    return { tools: openaiTools, toolChoice: void 0, toolWarnings };
  }
  const type = toolChoice.type;
  switch (type) {
    case "auto":
    case "none":
    case "required":
      return { tools: openaiTools, toolChoice: type, toolWarnings };
    case "tool":
      return {
        tools: openaiTools,
        toolChoice: toolChoice.toolName === "code_interpreter" || toolChoice.toolName === "file_search" || toolChoice.toolName === "image_generation" || toolChoice.toolName === "web_search_preview" || toolChoice.toolName === "web_search" ? { type: toolChoice.toolName } : { type: "function", name: toolChoice.toolName },
        toolWarnings
      };
    default: {
      const _exhaustiveCheck = type;
      throw new chunkA45A2T7R_cjs.UnsupportedFunctionalityError({
        functionality: `tool choice type: ${_exhaustiveCheck}`
      });
    }
  }
}
var OpenAIResponsesLanguageModel = class {
  constructor(modelId, config) {
    this.specificationVersion = "v2";
    this.supportedUrls = {
      "image/*": [/^https?:\/\/.*$/],
      "application/pdf": [/^https?:\/\/.*$/]
    };
    this.modelId = modelId;
    this.config = config;
  }
  get provider() {
    return this.config.provider;
  }
  async getArgs({
    maxOutputTokens,
    temperature,
    stopSequences,
    topP,
    topK,
    presencePenalty,
    frequencyPenalty,
    seed,
    prompt,
    providerOptions,
    tools,
    toolChoice,
    responseFormat
  }) {
    var _a, _b, _c, _d;
    const warnings = [];
    const modelCapabilities = getOpenAILanguageModelCapabilities(this.modelId);
    if (topK != null) {
      warnings.push({ type: "unsupported-setting", setting: "topK" });
    }
    if (seed != null) {
      warnings.push({ type: "unsupported-setting", setting: "seed" });
    }
    if (presencePenalty != null) {
      warnings.push({
        type: "unsupported-setting",
        setting: "presencePenalty"
      });
    }
    if (frequencyPenalty != null) {
      warnings.push({
        type: "unsupported-setting",
        setting: "frequencyPenalty"
      });
    }
    if (stopSequences != null) {
      warnings.push({ type: "unsupported-setting", setting: "stopSequences" });
    }
    const openaiOptions = await chunkA45A2T7R_cjs.parseProviderOptions({
      provider: "openai",
      providerOptions,
      schema: openaiResponsesProviderOptionsSchema
    });
    if ((openaiOptions == null ? void 0 : openaiOptions.conversation) && (openaiOptions == null ? void 0 : openaiOptions.previousResponseId)) {
      warnings.push({
        type: "unsupported-setting",
        setting: "conversation",
        details: "conversation and previousResponseId cannot be used together"
      });
    }
    const { input, warnings: inputWarnings } = await convertToOpenAIResponsesInput({
      prompt,
      systemMessageMode: modelCapabilities.systemMessageMode,
      fileIdPrefixes: this.config.fileIdPrefixes,
      store: (_a = openaiOptions == null ? void 0 : openaiOptions.store) != null ? _a : true,
      hasLocalShellTool: hasOpenAITool("openai.local_shell")
    });
    warnings.push(...inputWarnings);
    const strictJsonSchema = (_b = openaiOptions == null ? void 0 : openaiOptions.strictJsonSchema) != null ? _b : false;
    let include = openaiOptions == null ? void 0 : openaiOptions.include;
    function addInclude(key) {
      if (include == null) {
        include = [key];
      } else if (!include.includes(key)) {
        include = [...include, key];
      }
    }
    function hasOpenAITool(id) {
      return (tools == null ? void 0 : tools.find(
        (tool) => tool.type === "provider-defined" && tool.id === id
      )) != null;
    }
    const topLogprobs = typeof (openaiOptions == null ? void 0 : openaiOptions.logprobs) === "number" ? openaiOptions == null ? void 0 : openaiOptions.logprobs : (openaiOptions == null ? void 0 : openaiOptions.logprobs) === true ? TOP_LOGPROBS_MAX : void 0;
    if (topLogprobs) {
      addInclude("message.output_text.logprobs");
    }
    const webSearchToolName = (_c = tools == null ? void 0 : tools.find(
      (tool) => tool.type === "provider-defined" && (tool.id === "openai.web_search" || tool.id === "openai.web_search_preview")
    )) == null ? void 0 : _c.name;
    if (webSearchToolName) {
      addInclude("web_search_call.action.sources");
    }
    if (hasOpenAITool("openai.code_interpreter")) {
      addInclude("code_interpreter_call.outputs");
    }
    const store = openaiOptions == null ? void 0 : openaiOptions.store;
    if (store === false && modelCapabilities.isReasoningModel) {
      addInclude("reasoning.encrypted_content");
    }
    const baseArgs = {
      model: this.modelId,
      input,
      temperature,
      top_p: topP,
      max_output_tokens: maxOutputTokens,
      ...((responseFormat == null ? void 0 : responseFormat.type) === "json" || (openaiOptions == null ? void 0 : openaiOptions.textVerbosity)) && {
        text: {
          ...(responseFormat == null ? void 0 : responseFormat.type) === "json" && {
            format: responseFormat.schema != null ? {
              type: "json_schema",
              strict: strictJsonSchema,
              name: (_d = responseFormat.name) != null ? _d : "response",
              description: responseFormat.description,
              schema: responseFormat.schema
            } : { type: "json_object" }
          },
          ...(openaiOptions == null ? void 0 : openaiOptions.textVerbosity) && {
            verbosity: openaiOptions.textVerbosity
          }
        }
      },
      // provider options:
      conversation: openaiOptions == null ? void 0 : openaiOptions.conversation,
      max_tool_calls: openaiOptions == null ? void 0 : openaiOptions.maxToolCalls,
      metadata: openaiOptions == null ? void 0 : openaiOptions.metadata,
      parallel_tool_calls: openaiOptions == null ? void 0 : openaiOptions.parallelToolCalls,
      previous_response_id: openaiOptions == null ? void 0 : openaiOptions.previousResponseId,
      store,
      user: openaiOptions == null ? void 0 : openaiOptions.user,
      instructions: openaiOptions == null ? void 0 : openaiOptions.instructions,
      service_tier: openaiOptions == null ? void 0 : openaiOptions.serviceTier,
      include,
      prompt_cache_key: openaiOptions == null ? void 0 : openaiOptions.promptCacheKey,
      prompt_cache_retention: openaiOptions == null ? void 0 : openaiOptions.promptCacheRetention,
      safety_identifier: openaiOptions == null ? void 0 : openaiOptions.safetyIdentifier,
      top_logprobs: topLogprobs,
      truncation: openaiOptions == null ? void 0 : openaiOptions.truncation,
      // model-specific settings:
      ...modelCapabilities.isReasoningModel && ((openaiOptions == null ? void 0 : openaiOptions.reasoningEffort) != null || (openaiOptions == null ? void 0 : openaiOptions.reasoningSummary) != null) && {
        reasoning: {
          ...(openaiOptions == null ? void 0 : openaiOptions.reasoningEffort) != null && {
            effort: openaiOptions.reasoningEffort
          },
          ...(openaiOptions == null ? void 0 : openaiOptions.reasoningSummary) != null && {
            summary: openaiOptions.reasoningSummary
          }
        }
      }
    };
    if (modelCapabilities.isReasoningModel) {
      if (!((openaiOptions == null ? void 0 : openaiOptions.reasoningEffort) === "none" && modelCapabilities.supportsNonReasoningParameters)) {
        if (baseArgs.temperature != null) {
          baseArgs.temperature = void 0;
          warnings.push({
            type: "unsupported-setting",
            setting: "temperature",
            details: "temperature is not supported for reasoning models"
          });
        }
        if (baseArgs.top_p != null) {
          baseArgs.top_p = void 0;
          warnings.push({
            type: "unsupported-setting",
            setting: "topP",
            details: "topP is not supported for reasoning models"
          });
        }
      }
    } else {
      if ((openaiOptions == null ? void 0 : openaiOptions.reasoningEffort) != null) {
        warnings.push({
          type: "unsupported-setting",
          setting: "reasoningEffort",
          details: "reasoningEffort is not supported for non-reasoning models"
        });
      }
      if ((openaiOptions == null ? void 0 : openaiOptions.reasoningSummary) != null) {
        warnings.push({
          type: "unsupported-setting",
          setting: "reasoningSummary",
          details: "reasoningSummary is not supported for non-reasoning models"
        });
      }
    }
    if ((openaiOptions == null ? void 0 : openaiOptions.serviceTier) === "flex" && !modelCapabilities.supportsFlexProcessing) {
      warnings.push({
        type: "unsupported-setting",
        setting: "serviceTier",
        details: "flex processing is only available for o3, o4-mini, and gpt-5 models"
      });
      delete baseArgs.service_tier;
    }
    if ((openaiOptions == null ? void 0 : openaiOptions.serviceTier) === "priority" && !modelCapabilities.supportsPriorityProcessing) {
      warnings.push({
        type: "unsupported-setting",
        setting: "serviceTier",
        details: "priority processing is only available for supported models (gpt-4, gpt-5, gpt-5-mini, o3, o4-mini) and requires Enterprise access. gpt-5-nano is not supported"
      });
      delete baseArgs.service_tier;
    }
    const {
      tools: openaiTools,
      toolChoice: openaiToolChoice,
      toolWarnings
    } = await prepareResponsesTools({
      tools,
      toolChoice,
      strictJsonSchema
    });
    return {
      webSearchToolName,
      args: {
        ...baseArgs,
        tools: openaiTools,
        tool_choice: openaiToolChoice
      },
      warnings: [...warnings, ...toolWarnings],
      store
    };
  }
  async doGenerate(options) {
    var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x, _y, _z, _A, _B;
    const {
      args: body,
      warnings,
      webSearchToolName
    } = await this.getArgs(options);
    const url = this.config.url({
      path: "/responses",
      modelId: this.modelId
    });
    const providerKey = this.config.provider.replace(".responses", "");
    const {
      responseHeaders,
      value: response,
      rawValue: rawResponse
    } = await chunkA45A2T7R_cjs.postJsonToApi({
      url,
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body,
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createJsonResponseHandler(
        openaiResponsesResponseSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    if (response.error) {
      throw new chunkA45A2T7R_cjs.APICallError({
        message: response.error.message,
        url,
        requestBodyValues: body,
        statusCode: 400,
        responseHeaders,
        responseBody: rawResponse,
        isRetryable: false
      });
    }
    const content = [];
    const logprobs = [];
    let hasFunctionCall = false;
    for (const part of response.output) {
      switch (part.type) {
        case "reasoning": {
          if (part.summary.length === 0) {
            part.summary.push({ type: "summary_text", text: "" });
          }
          for (const summary of part.summary) {
            content.push({
              type: "reasoning",
              text: summary.text,
              providerMetadata: {
                [providerKey]: {
                  itemId: part.id,
                  reasoningEncryptedContent: (_a = part.encrypted_content) != null ? _a : null
                }
              }
            });
          }
          break;
        }
        case "image_generation_call": {
          content.push({
            type: "tool-call",
            toolCallId: part.id,
            toolName: "image_generation",
            input: "{}",
            providerExecuted: true
          });
          content.push({
            type: "tool-result",
            toolCallId: part.id,
            toolName: "image_generation",
            result: {
              result: part.result
            },
            providerExecuted: true
          });
          break;
        }
        case "local_shell_call": {
          content.push({
            type: "tool-call",
            toolCallId: part.call_id,
            toolName: "local_shell",
            input: JSON.stringify({
              action: part.action
            }),
            providerMetadata: {
              [providerKey]: {
                itemId: part.id
              }
            }
          });
          break;
        }
        case "message": {
          for (const contentPart of part.content) {
            if (((_c = (_b = options.providerOptions) == null ? void 0 : _b.openai) == null ? void 0 : _c.logprobs) && contentPart.logprobs) {
              logprobs.push(contentPart.logprobs);
            }
            content.push({
              type: "text",
              text: contentPart.text,
              providerMetadata: {
                [providerKey]: {
                  itemId: part.id,
                  ...part.phase != null && { phase: part.phase }
                }
              }
            });
            for (const annotation of contentPart.annotations) {
              if (annotation.type === "url_citation") {
                content.push({
                  type: "source",
                  sourceType: "url",
                  id: (_f = (_e = (_d = this.config).generateId) == null ? void 0 : _e.call(_d)) != null ? _f : chunkA45A2T7R_cjs.generateId(),
                  url: annotation.url,
                  title: annotation.title
                });
              } else if (annotation.type === "file_citation") {
                content.push({
                  type: "source",
                  sourceType: "document",
                  id: (_i = (_h = (_g = this.config).generateId) == null ? void 0 : _h.call(_g)) != null ? _i : chunkA45A2T7R_cjs.generateId(),
                  mediaType: "text/plain",
                  title: (_k = (_j = annotation.quote) != null ? _j : annotation.filename) != null ? _k : "Document",
                  filename: (_l = annotation.filename) != null ? _l : annotation.file_id,
                  ...annotation.file_id ? {
                    providerMetadata: {
                      [providerKey]: {
                        fileId: annotation.file_id
                      }
                    }
                  } : {}
                });
              } else if (annotation.type === "container_file_citation") {
                content.push({
                  type: "source",
                  sourceType: "document",
                  id: (_o = (_n = (_m = this.config).generateId) == null ? void 0 : _n.call(_m)) != null ? _o : chunkA45A2T7R_cjs.generateId(),
                  mediaType: "text/plain",
                  title: (_q = (_p = annotation.filename) != null ? _p : annotation.file_id) != null ? _q : "Document",
                  filename: (_r = annotation.filename) != null ? _r : annotation.file_id,
                  providerMetadata: {
                    [providerKey]: {
                      fileId: annotation.file_id,
                      containerId: annotation.container_id,
                      ...annotation.index != null ? { index: annotation.index } : {}
                    }
                  }
                });
              } else if (annotation.type === "file_path") {
                content.push({
                  type: "source",
                  sourceType: "document",
                  id: (_u = (_t = (_s = this.config).generateId) == null ? void 0 : _t.call(_s)) != null ? _u : chunkA45A2T7R_cjs.generateId(),
                  mediaType: "application/octet-stream",
                  title: annotation.file_id,
                  filename: annotation.file_id,
                  providerMetadata: {
                    [providerKey]: {
                      fileId: annotation.file_id,
                      ...annotation.index != null ? { index: annotation.index } : {}
                    }
                  }
                });
              }
            }
          }
          break;
        }
        case "function_call": {
          hasFunctionCall = true;
          content.push({
            type: "tool-call",
            toolCallId: part.call_id,
            toolName: part.name,
            input: part.arguments,
            providerMetadata: {
              [providerKey]: {
                itemId: part.id
              }
            }
          });
          break;
        }
        case "web_search_call": {
          content.push({
            type: "tool-call",
            toolCallId: part.id,
            toolName: webSearchToolName != null ? webSearchToolName : "web_search",
            input: JSON.stringify({}),
            providerExecuted: true
          });
          content.push({
            type: "tool-result",
            toolCallId: part.id,
            toolName: webSearchToolName != null ? webSearchToolName : "web_search",
            result: mapWebSearchOutput(part.action),
            providerExecuted: true
          });
          break;
        }
        case "computer_call": {
          content.push({
            type: "tool-call",
            toolCallId: part.id,
            toolName: "computer_use",
            input: "",
            providerExecuted: true
          });
          content.push({
            type: "tool-result",
            toolCallId: part.id,
            toolName: "computer_use",
            result: {
              type: "computer_use_tool_result",
              status: part.status || "completed"
            },
            providerExecuted: true
          });
          break;
        }
        case "file_search_call": {
          content.push({
            type: "tool-call",
            toolCallId: part.id,
            toolName: "file_search",
            input: "{}",
            providerExecuted: true
          });
          content.push({
            type: "tool-result",
            toolCallId: part.id,
            toolName: "file_search",
            result: {
              queries: part.queries,
              results: (_w = (_v = part.results) == null ? void 0 : _v.map((result) => ({
                attributes: result.attributes,
                fileId: result.file_id,
                filename: result.filename,
                score: result.score,
                text: result.text
              }))) != null ? _w : null
            },
            providerExecuted: true
          });
          break;
        }
        case "code_interpreter_call": {
          content.push({
            type: "tool-call",
            toolCallId: part.id,
            toolName: "code_interpreter",
            input: JSON.stringify({
              code: part.code,
              containerId: part.container_id
            }),
            providerExecuted: true
          });
          content.push({
            type: "tool-result",
            toolCallId: part.id,
            toolName: "code_interpreter",
            result: {
              outputs: part.outputs
            },
            providerExecuted: true
          });
          break;
        }
      }
    }
    const providerMetadata = {
      [providerKey]: {
        ...response.id != null ? { responseId: response.id } : {}
      }
    };
    if (logprobs.length > 0) {
      providerMetadata[providerKey].logprobs = logprobs;
    }
    if (typeof response.service_tier === "string") {
      providerMetadata[providerKey].serviceTier = response.service_tier;
    }
    const usage = response.usage;
    return {
      content,
      finishReason: mapOpenAIResponseFinishReason({
        finishReason: (_x = response.incomplete_details) == null ? void 0 : _x.reason,
        hasFunctionCall
      }),
      usage: {
        inputTokens: usage.input_tokens,
        outputTokens: usage.output_tokens,
        totalTokens: usage.input_tokens + usage.output_tokens,
        reasoningTokens: (_z = (_y = usage.output_tokens_details) == null ? void 0 : _y.reasoning_tokens) != null ? _z : void 0,
        cachedInputTokens: (_B = (_A = usage.input_tokens_details) == null ? void 0 : _A.cached_tokens) != null ? _B : void 0
      },
      request: { body },
      response: {
        id: response.id,
        timestamp: new Date(response.created_at * 1e3),
        modelId: response.model,
        headers: responseHeaders,
        body: rawResponse
      },
      providerMetadata,
      warnings
    };
  }
  async doStream(options) {
    const {
      args: body,
      warnings,
      webSearchToolName,
      store
    } = await this.getArgs(options);
    const { responseHeaders, value: response } = await chunkA45A2T7R_cjs.postJsonToApi({
      url: this.config.url({
        path: "/responses",
        modelId: this.modelId
      }),
      headers: chunkA45A2T7R_cjs.combineHeaders(this.config.headers(), options.headers),
      body: {
        ...body,
        stream: true
      },
      failedResponseHandler: openaiFailedResponseHandler,
      successfulResponseHandler: chunkA45A2T7R_cjs.createEventSourceResponseHandler(
        openaiResponsesChunkSchema
      ),
      abortSignal: options.abortSignal,
      fetch: this.config.fetch
    });
    const self = this;
    const providerKey = this.config.provider.replace(".responses", "");
    let finishReason = "unknown";
    const usage = {
      inputTokens: void 0,
      outputTokens: void 0,
      totalTokens: void 0
    };
    const logprobs = [];
    let responseId = null;
    const ongoingToolCalls = {};
    const ongoingAnnotations = [];
    let activeMessagePhase;
    let hasFunctionCall = false;
    const activeReasoning = {};
    let serviceTier;
    return {
      stream: response.pipeThrough(
        new TransformStream({
          start(controller) {
            controller.enqueue({ type: "stream-start", warnings });
          },
          transform(chunk, controller) {
            var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l, _m, _n, _o, _p, _q, _r, _s, _t, _u, _v, _w, _x;
            if (options.includeRawChunks) {
              controller.enqueue({ type: "raw", rawValue: chunk.rawValue });
            }
            if (!chunk.success) {
              finishReason = "error";
              controller.enqueue({ type: "error", error: chunk.error });
              return;
            }
            const value = chunk.value;
            if (isResponseOutputItemAddedChunk(value)) {
              if (value.item.type === "function_call") {
                ongoingToolCalls[value.output_index] = {
                  toolName: value.item.name,
                  toolCallId: value.item.call_id
                };
                controller.enqueue({
                  type: "tool-input-start",
                  id: value.item.call_id,
                  toolName: value.item.name
                });
              } else if (value.item.type === "web_search_call") {
                ongoingToolCalls[value.output_index] = {
                  toolName: webSearchToolName != null ? webSearchToolName : "web_search",
                  toolCallId: value.item.id
                };
                controller.enqueue({
                  type: "tool-input-start",
                  id: value.item.id,
                  toolName: webSearchToolName != null ? webSearchToolName : "web_search",
                  providerExecuted: true
                });
                controller.enqueue({
                  type: "tool-input-end",
                  id: value.item.id
                });
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: value.item.id,
                  toolName: webSearchToolName != null ? webSearchToolName : "web_search",
                  input: JSON.stringify({}),
                  providerExecuted: true
                });
              } else if (value.item.type === "computer_call") {
                ongoingToolCalls[value.output_index] = {
                  toolName: "computer_use",
                  toolCallId: value.item.id
                };
                controller.enqueue({
                  type: "tool-input-start",
                  id: value.item.id,
                  toolName: "computer_use",
                  providerExecuted: true
                });
              } else if (value.item.type === "code_interpreter_call") {
                ongoingToolCalls[value.output_index] = {
                  toolName: "code_interpreter",
                  toolCallId: value.item.id,
                  codeInterpreter: {
                    containerId: value.item.container_id
                  }
                };
                controller.enqueue({
                  type: "tool-input-start",
                  id: value.item.id,
                  toolName: "code_interpreter",
                  providerExecuted: true
                });
                controller.enqueue({
                  type: "tool-input-delta",
                  id: value.item.id,
                  delta: `{"containerId":"${value.item.container_id}","code":"`
                });
              } else if (value.item.type === "file_search_call") {
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: value.item.id,
                  toolName: "file_search",
                  input: "{}",
                  providerExecuted: true
                });
              } else if (value.item.type === "image_generation_call") {
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: value.item.id,
                  toolName: "image_generation",
                  input: "{}",
                  providerExecuted: true
                });
              } else if (value.item.type === "message") {
                ongoingAnnotations.splice(0, ongoingAnnotations.length);
                activeMessagePhase = (_a = value.item.phase) != null ? _a : void 0;
                controller.enqueue({
                  type: "text-start",
                  id: value.item.id,
                  providerMetadata: {
                    [providerKey]: {
                      itemId: value.item.id,
                      ...value.item.phase != null && {
                        phase: value.item.phase
                      }
                    }
                  }
                });
              } else if (isResponseOutputItemAddedChunk(value) && value.item.type === "reasoning") {
                activeReasoning[value.item.id] = {
                  encryptedContent: value.item.encrypted_content,
                  summaryParts: { 0: "active" }
                };
                controller.enqueue({
                  type: "reasoning-start",
                  id: `${value.item.id}:0`,
                  providerMetadata: {
                    [providerKey]: {
                      itemId: value.item.id,
                      reasoningEncryptedContent: (_b = value.item.encrypted_content) != null ? _b : null
                    }
                  }
                });
              }
            } else if (isResponseOutputItemDoneChunk(value)) {
              if (value.item.type === "message") {
                const phase = (_c = value.item.phase) != null ? _c : activeMessagePhase;
                activeMessagePhase = void 0;
                controller.enqueue({
                  type: "text-end",
                  id: value.item.id,
                  providerMetadata: {
                    [providerKey]: {
                      itemId: value.item.id,
                      ...phase != null && { phase },
                      ...ongoingAnnotations.length > 0 && {
                        annotations: ongoingAnnotations
                      }
                    }
                  }
                });
              } else if (value.item.type === "function_call") {
                ongoingToolCalls[value.output_index] = void 0;
                hasFunctionCall = true;
                controller.enqueue({
                  type: "tool-input-end",
                  id: value.item.call_id
                });
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: value.item.call_id,
                  toolName: value.item.name,
                  input: value.item.arguments,
                  providerMetadata: {
                    [providerKey]: {
                      itemId: value.item.id
                    }
                  }
                });
              } else if (value.item.type === "web_search_call") {
                ongoingToolCalls[value.output_index] = void 0;
                controller.enqueue({
                  type: "tool-result",
                  toolCallId: value.item.id,
                  toolName: webSearchToolName != null ? webSearchToolName : "web_search",
                  result: mapWebSearchOutput(value.item.action),
                  providerExecuted: true
                });
              } else if (value.item.type === "computer_call") {
                ongoingToolCalls[value.output_index] = void 0;
                controller.enqueue({
                  type: "tool-input-end",
                  id: value.item.id
                });
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: value.item.id,
                  toolName: "computer_use",
                  input: "",
                  providerExecuted: true
                });
                controller.enqueue({
                  type: "tool-result",
                  toolCallId: value.item.id,
                  toolName: "computer_use",
                  result: {
                    type: "computer_use_tool_result",
                    status: value.item.status || "completed"
                  },
                  providerExecuted: true
                });
              } else if (value.item.type === "file_search_call") {
                ongoingToolCalls[value.output_index] = void 0;
                controller.enqueue({
                  type: "tool-result",
                  toolCallId: value.item.id,
                  toolName: "file_search",
                  result: {
                    queries: value.item.queries,
                    results: (_e = (_d = value.item.results) == null ? void 0 : _d.map((result) => ({
                      attributes: result.attributes,
                      fileId: result.file_id,
                      filename: result.filename,
                      score: result.score,
                      text: result.text
                    }))) != null ? _e : null
                  },
                  providerExecuted: true
                });
              } else if (value.item.type === "code_interpreter_call") {
                ongoingToolCalls[value.output_index] = void 0;
                controller.enqueue({
                  type: "tool-result",
                  toolCallId: value.item.id,
                  toolName: "code_interpreter",
                  result: {
                    outputs: value.item.outputs
                  },
                  providerExecuted: true
                });
              } else if (value.item.type === "image_generation_call") {
                controller.enqueue({
                  type: "tool-result",
                  toolCallId: value.item.id,
                  toolName: "image_generation",
                  result: {
                    result: value.item.result
                  },
                  providerExecuted: true
                });
              } else if (value.item.type === "local_shell_call") {
                ongoingToolCalls[value.output_index] = void 0;
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: value.item.call_id,
                  toolName: "local_shell",
                  input: JSON.stringify({
                    action: {
                      type: "exec",
                      command: value.item.action.command,
                      timeoutMs: value.item.action.timeout_ms,
                      user: value.item.action.user,
                      workingDirectory: value.item.action.working_directory,
                      env: value.item.action.env
                    }
                  }),
                  providerMetadata: {
                    [providerKey]: { itemId: value.item.id }
                  }
                });
              } else if (value.item.type === "reasoning") {
                const activeReasoningPart = activeReasoning[value.item.id];
                const summaryPartIndices = Object.entries(
                  activeReasoningPart.summaryParts
                ).filter(
                  ([_, status]) => status === "active" || status === "can-conclude"
                ).map(([summaryIndex]) => summaryIndex);
                for (const summaryIndex of summaryPartIndices) {
                  controller.enqueue({
                    type: "reasoning-end",
                    id: `${value.item.id}:${summaryIndex}`,
                    providerMetadata: {
                      [providerKey]: {
                        itemId: value.item.id,
                        reasoningEncryptedContent: (_f = value.item.encrypted_content) != null ? _f : null
                      }
                    }
                  });
                }
                delete activeReasoning[value.item.id];
              }
            } else if (isResponseFunctionCallArgumentsDeltaChunk(value)) {
              const toolCall = ongoingToolCalls[value.output_index];
              if (toolCall != null) {
                controller.enqueue({
                  type: "tool-input-delta",
                  id: toolCall.toolCallId,
                  delta: value.delta
                });
              }
            } else if (isResponseCodeInterpreterCallCodeDeltaChunk(value)) {
              const toolCall = ongoingToolCalls[value.output_index];
              if (toolCall != null) {
                controller.enqueue({
                  type: "tool-input-delta",
                  id: toolCall.toolCallId,
                  // The delta is code, which is embedding in a JSON string.
                  // To escape it, we use JSON.stringify and slice to remove the outer quotes.
                  delta: JSON.stringify(value.delta).slice(1, -1)
                });
              }
            } else if (isResponseCodeInterpreterCallCodeDoneChunk(value)) {
              const toolCall = ongoingToolCalls[value.output_index];
              if (toolCall != null) {
                controller.enqueue({
                  type: "tool-input-delta",
                  id: toolCall.toolCallId,
                  delta: '"}'
                });
                controller.enqueue({
                  type: "tool-input-end",
                  id: toolCall.toolCallId
                });
                controller.enqueue({
                  type: "tool-call",
                  toolCallId: toolCall.toolCallId,
                  toolName: "code_interpreter",
                  input: JSON.stringify({
                    code: value.code,
                    containerId: toolCall.codeInterpreter.containerId
                  }),
                  providerExecuted: true
                });
              }
            } else if (isResponseCreatedChunk(value)) {
              responseId = value.response.id;
              controller.enqueue({
                type: "response-metadata",
                id: value.response.id,
                timestamp: new Date(value.response.created_at * 1e3),
                modelId: value.response.model
              });
            } else if (isTextDeltaChunk(value)) {
              controller.enqueue({
                type: "text-delta",
                id: value.item_id,
                delta: value.delta
              });
              if (((_h = (_g = options.providerOptions) == null ? void 0 : _g.openai) == null ? void 0 : _h.logprobs) && value.logprobs) {
                logprobs.push(value.logprobs);
              }
            } else if (value.type === "response.reasoning_summary_part.added") {
              if (value.summary_index > 0) {
                const activeReasoningPart = activeReasoning[value.item_id];
                activeReasoningPart.summaryParts[value.summary_index] = "active";
                for (const summaryIndex of Object.keys(
                  activeReasoningPart.summaryParts
                )) {
                  if (activeReasoningPart.summaryParts[summaryIndex] === "can-conclude") {
                    controller.enqueue({
                      type: "reasoning-end",
                      id: `${value.item_id}:${summaryIndex}`,
                      providerMetadata: {
                        [providerKey]: { itemId: value.item_id }
                      }
                    });
                    activeReasoningPart.summaryParts[summaryIndex] = "concluded";
                  }
                }
                controller.enqueue({
                  type: "reasoning-start",
                  id: `${value.item_id}:${value.summary_index}`,
                  providerMetadata: {
                    [providerKey]: {
                      itemId: value.item_id,
                      reasoningEncryptedContent: (_j = (_i = activeReasoning[value.item_id]) == null ? void 0 : _i.encryptedContent) != null ? _j : null
                    }
                  }
                });
              }
            } else if (value.type === "response.reasoning_summary_text.delta") {
              controller.enqueue({
                type: "reasoning-delta",
                id: `${value.item_id}:${value.summary_index}`,
                delta: value.delta,
                providerMetadata: {
                  [providerKey]: {
                    itemId: value.item_id
                  }
                }
              });
            } else if (value.type === "response.reasoning_summary_part.done") {
              if (store) {
                controller.enqueue({
                  type: "reasoning-end",
                  id: `${value.item_id}:${value.summary_index}`,
                  providerMetadata: {
                    [providerKey]: { itemId: value.item_id }
                  }
                });
                activeReasoning[value.item_id].summaryParts[value.summary_index] = "concluded";
              } else {
                activeReasoning[value.item_id].summaryParts[value.summary_index] = "can-conclude";
              }
            } else if (isResponseFinishedChunk(value)) {
              finishReason = mapOpenAIResponseFinishReason({
                finishReason: (_k = value.response.incomplete_details) == null ? void 0 : _k.reason,
                hasFunctionCall
              });
              usage.inputTokens = value.response.usage.input_tokens;
              usage.outputTokens = value.response.usage.output_tokens;
              usage.totalTokens = value.response.usage.input_tokens + value.response.usage.output_tokens;
              usage.reasoningTokens = (_m = (_l = value.response.usage.output_tokens_details) == null ? void 0 : _l.reasoning_tokens) != null ? _m : void 0;
              usage.cachedInputTokens = (_o = (_n = value.response.usage.input_tokens_details) == null ? void 0 : _n.cached_tokens) != null ? _o : void 0;
              if (typeof value.response.service_tier === "string") {
                serviceTier = value.response.service_tier;
              }
            } else if (isResponseAnnotationAddedChunk(value)) {
              ongoingAnnotations.push(value.annotation);
              if (value.annotation.type === "url_citation") {
                controller.enqueue({
                  type: "source",
                  sourceType: "url",
                  id: (_r = (_q = (_p = self.config).generateId) == null ? void 0 : _q.call(_p)) != null ? _r : chunkA45A2T7R_cjs.generateId(),
                  url: value.annotation.url,
                  title: value.annotation.title
                });
              } else if (value.annotation.type === "file_citation") {
                controller.enqueue({
                  type: "source",
                  sourceType: "document",
                  id: (_u = (_t = (_s = self.config).generateId) == null ? void 0 : _t.call(_s)) != null ? _u : chunkA45A2T7R_cjs.generateId(),
                  mediaType: "text/plain",
                  title: (_w = (_v = value.annotation.quote) != null ? _v : value.annotation.filename) != null ? _w : "Document",
                  filename: (_x = value.annotation.filename) != null ? _x : value.annotation.file_id,
                  ...value.annotation.file_id ? {
                    providerMetadata: {
                      [providerKey]: {
                        fileId: value.annotation.file_id
                      }
                    }
                  } : {}
                });
              }
            } else if (isErrorChunk(value)) {
              controller.enqueue({ type: "error", error: value });
            }
          },
          flush(controller) {
            const providerMetadata = {
              [providerKey]: {
                responseId
              }
            };
            if (logprobs.length > 0) {
              providerMetadata[providerKey].logprobs = logprobs;
            }
            if (serviceTier !== void 0) {
              providerMetadata[providerKey].serviceTier = serviceTier;
            }
            controller.enqueue({
              type: "finish",
              finishReason,
              usage,
              providerMetadata
            });
          }
        })
      ),
      request: { body },
      response: { headers: responseHeaders }
    };
  }
};
function isTextDeltaChunk(chunk) {
  return chunk.type === "response.output_text.delta";
}
function isResponseOutputItemDoneChunk(chunk) {
  return chunk.type === "response.output_item.done";
}
function isResponseFinishedChunk(chunk) {
  return chunk.type === "response.completed" || chunk.type === "response.incomplete";
}
function isResponseCreatedChunk(chunk) {
  return chunk.type === "response.created";
}
function isResponseFunctionCallArgumentsDeltaChunk(chunk) {
  return chunk.type === "response.function_call_arguments.delta";
}
function isResponseCodeInterpreterCallCodeDeltaChunk(chunk) {
  return chunk.type === "response.code_interpreter_call_code.delta";
}
function isResponseCodeInterpreterCallCodeDoneChunk(chunk) {
  return chunk.type === "response.code_interpreter_call_code.done";
}
function isResponseOutputItemAddedChunk(chunk) {
  return chunk.type === "response.output_item.added";
}
function isResponseAnnotationAddedChunk(chunk) {
  return chunk.type === "response.output_text.annotation.added";
}
function isErrorChunk(chunk) {
  return chunk.type === "error";
}
function mapWebSearchOutput(action) {
  var _a;
  if (action == null) {
    return {};
  }
  switch (action.type) {
    case "search":
      return {
        action: { type: "search", query: (_a = action.query) != null ? _a : void 0 },
        // include sources when provided by the Responses API (behind include flag)
        ...action.sources != null && { sources: action.sources }
      };
    case "open_page":
      return { action: { type: "openPage", url: action.url } };
    case "find_in_page":
      return {
        action: {
          type: "findInPage",
          url: action.url,
          pattern: action.pattern
        }
      };
  }
}

// ../../../../../setup-pnpm/node_modules/.bin/store/v11/links/@ai-sdk/azure/2.0.108/e337a4e854f5f14549b2c4b694ea7f87c6b9285c131e86f3950198f851676d5f/node_modules/@ai-sdk/azure/dist/index.mjs
var azureOpenaiTools = {
  codeInterpreter,
  fileSearch,
  imageGeneration,
  webSearchPreview
};
var VERSION = "2.0.108" ;
function createAzure(options = {}) {
  var _a;
  const getHeaders = () => {
    const baseHeaders = {
      "api-key": chunkA45A2T7R_cjs.loadApiKey({
        apiKey: options.apiKey,
        environmentVariableName: "AZURE_API_KEY",
        description: "Azure OpenAI"
      }),
      ...options.headers
    };
    return chunkA45A2T7R_cjs.withUserAgentSuffix(baseHeaders, `ai-sdk/azure/${VERSION}`);
  };
  const getResourceName = () => chunkA45A2T7R_cjs.loadSetting({
    settingValue: options.resourceName,
    settingName: "resourceName",
    environmentVariableName: "AZURE_RESOURCE_NAME",
    description: "Azure OpenAI resource name"
  });
  const apiVersion = (_a = options.apiVersion) != null ? _a : "v1";
  const url = ({ path, modelId }) => {
    var _a2;
    const baseUrlPrefix = (_a2 = options.baseURL) != null ? _a2 : `https://${getResourceName()}.openai.azure.com/openai`;
    let fullUrl;
    if (options.useDeploymentBasedUrls) {
      fullUrl = new URL(`${baseUrlPrefix}/deployments/${modelId}${path}`);
    } else {
      fullUrl = new URL(`${baseUrlPrefix}/v1${path}`);
    }
    fullUrl.searchParams.set("api-version", apiVersion);
    return fullUrl.toString();
  };
  const createChatModel = (deploymentName) => new OpenAIChatLanguageModel(deploymentName, {
    provider: "azure.chat",
    url,
    headers: getHeaders,
    fetch: options.fetch
  });
  const createCompletionModel = (modelId) => new OpenAICompletionLanguageModel(modelId, {
    provider: "azure.completion",
    url,
    headers: getHeaders,
    fetch: options.fetch
  });
  const createEmbeddingModel = (modelId) => new OpenAIEmbeddingModel(modelId, {
    provider: "azure.embeddings",
    headers: getHeaders,
    url,
    fetch: options.fetch
  });
  const createResponsesModel = (modelId) => new OpenAIResponsesLanguageModel(modelId, {
    provider: "azure.responses",
    url,
    headers: getHeaders,
    fetch: options.fetch,
    fileIdPrefixes: ["assistant-"]
  });
  const createImageModel = (modelId) => new OpenAIImageModel(modelId, {
    provider: "azure.image",
    url,
    headers: getHeaders,
    fetch: options.fetch
  });
  const createTranscriptionModel = (modelId) => new OpenAITranscriptionModel(modelId, {
    provider: "azure.transcription",
    url,
    headers: getHeaders,
    fetch: options.fetch
  });
  const createSpeechModel = (modelId) => new OpenAISpeechModel(modelId, {
    provider: "azure.speech",
    url,
    headers: getHeaders,
    fetch: options.fetch
  });
  const provider = function(deploymentId) {
    if (new.target) {
      throw new Error(
        "The Azure OpenAI model function cannot be called with the new keyword."
      );
    }
    return createChatModel(deploymentId);
  };
  provider.languageModel = createChatModel;
  provider.chat = createChatModel;
  provider.completion = createCompletionModel;
  provider.embedding = createEmbeddingModel;
  provider.image = createImageModel;
  provider.imageModel = createImageModel;
  provider.textEmbedding = createEmbeddingModel;
  provider.textEmbeddingModel = createEmbeddingModel;
  provider.responses = createResponsesModel;
  provider.transcription = createTranscriptionModel;
  provider.speech = createSpeechModel;
  provider.tools = azureOpenaiTools;
  return provider;
}
createAzure();
var TERMINAL_RESPONSE_EVENTS = /* @__PURE__ */ new Set(["response.completed", "response.failed", "response.incomplete", "error"]);
function createOpenAIWebSocketFetch(options) {
  const wsUrl = options?.url ?? "wss://api.openai.com/v1/responses";
  const betaHeader = options?.betaHeader === void 0 ? "responses_websockets=2026-02-06" : options.betaHeader;
  const apiKeyQueryParam = options?.apiKeyQueryParam ?? false;
  let ws = null;
  let connecting = null;
  let connectionKey = null;
  let nextQueryCredentialId = 0;
  const queryCredentialIds = /* @__PURE__ */ new Map();
  let busy = false;
  function getQueryCredentialCacheId(value) {
    if (!value) return "";
    const existing = queryCredentialIds.get(value);
    if (existing) return existing;
    const id = String(++nextQueryCredentialId);
    queryCredentialIds.set(value, id);
    return id;
  }
  function getConnection(authorization, headers, signal) {
    if (signal?.aborted) {
      return Promise.reject(getAbortError(signal));
    }
    const normalizedHeaders = { ...normalizeHeaders(options?.headers), ...headers };
    const apiKey = normalizedHeaders["api-key"];
    delete normalizedHeaders["authorization"];
    delete normalizedHeaders["openai-beta"];
    if (options?.apiKeyAsBearer || apiKeyQueryParam) {
      delete normalizedHeaders["api-key"];
    }
    const queryCredential = apiKeyQueryParam ? `${apiKeyQueryParam}:${getQueryCredentialCacheId(apiKey)}` : "";
    const nextConnectionKey = buildConnectionKey(authorization, normalizedHeaders, queryCredential);
    if (ws?.readyState === WebSocket__default.default.OPEN && connectionKey === nextConnectionKey) {
      return Promise.resolve(ws);
    }
    if (ws?.readyState === WebSocket__default.default.OPEN && connectionKey !== nextConnectionKey) {
      ws.close();
      ws = null;
      connectionKey = null;
    }
    if (connecting && connectionKey === nextConnectionKey) return connecting;
    connectionKey = nextConnectionKey;
    connecting = new Promise((resolve, reject) => {
      let settled = false;
      const socket = new WebSocket__default.default(getWebSocketUrl(wsUrl, apiKeyQueryParam, apiKey), {
        headers: {
          ...normalizedHeaders,
          ...authorization ? { Authorization: authorization } : {},
          ...betaHeader ? { "OpenAI-Beta": betaHeader } : {}
        }
      });
      function cleanupAbortListener() {
        signal?.removeEventListener("abort", onAbort);
      }
      function rejectConnection(err, closeSocket = true) {
        if (settled) return;
        settled = true;
        connecting = null;
        connectionKey = null;
        cleanupAbortListener();
        if (closeSocket) socket.close();
        reject(err);
      }
      function onAbort() {
        rejectConnection(getAbortError(signal));
      }
      signal?.addEventListener("abort", onAbort, { once: true });
      socket.on("open", () => {
        if (signal?.aborted) {
          rejectConnection(getAbortError(signal));
          return;
        }
        settled = true;
        ws = socket;
        connecting = null;
        cleanupAbortListener();
        resolve(socket);
      });
      socket.on("error", (err) => {
        rejectConnection(err, false);
      });
      socket.on("close", () => {
        if (settled) return;
        rejectConnection(new Error("WebSocket closed before the connection opened"), false);
      });
      socket.on("close", () => {
        if (ws === socket) ws = null;
        if (ws === null) connectionKey = null;
      });
    });
    return connecting;
  }
  async function websocketFetch(input, init) {
    const url = input instanceof URL ? input.toString() : typeof input === "string" ? input : input.url;
    if (init?.method !== "POST" || !isResponsesUrl(url)) {
      return globalThis.fetch(input, init);
    }
    let body;
    try {
      body = JSON.parse(typeof init.body === "string" ? init.body : "");
    } catch {
      return globalThis.fetch(input, init);
    }
    if (!body.stream) {
      return globalThis.fetch(input, init);
    }
    if (busy) {
      if (body.previous_response_id) {
        throw new Error(
          "Cannot start an overlapping WebSocket Responses continuation. Wait for the active stream to finish before sending previous_response_id."
        );
      }
      return globalThis.fetch(input, init);
    }
    const headers = normalizeHeaders(init.headers);
    const authorization = headers["authorization"] ?? (options?.apiKeyAsBearer && headers["api-key"] ? `Bearer ${headers["api-key"]}` : "");
    busy = true;
    let connection;
    try {
      connection = await getConnection(authorization, headers, init?.signal);
    } catch (err) {
      busy = false;
      throw err;
    }
    const { stream: _stream, background: _background, ...requestBody } = body;
    const encoder = new TextEncoder();
    let cleanupActiveStream;
    const responseStream = new ReadableStream({
      start(controller) {
        let cleanedUp = false;
        let abortHandler;
        function cleanup({ closeSocket = false } = {}) {
          if (cleanedUp) return;
          cleanedUp = true;
          connection.off("message", onMessage);
          connection.off("error", onError);
          connection.off("close", onClose);
          if (abortHandler) {
            init?.signal?.removeEventListener("abort", abortHandler);
            abortHandler = void 0;
          }
          if (closeSocket && ws === connection) {
            connection.close();
            ws = null;
            connectionKey = null;
          }
          busy = false;
          cleanupActiveStream = void 0;
        }
        cleanupActiveStream = cleanup;
        function onMessage(data) {
          const text = data.toString();
          controller.enqueue(encoder.encode(formatSSEData(text)));
          try {
            const event = JSON.parse(text);
            if (isTerminalWebSocketEvent(event)) {
              controller.enqueue(encoder.encode("data: [DONE]\n\n"));
              cleanup({ closeSocket: shouldReconnectAfterEvent(event) });
              controller.close();
            }
          } catch {
          }
        }
        function onError(err) {
          cleanup();
          controller.error(err);
        }
        function onClose() {
          cleanup();
          try {
            controller.close();
          } catch {
          }
        }
        connection.on("message", onMessage);
        connection.on("error", onError);
        connection.on("close", onClose);
        const signal = init?.signal;
        if (signal) {
          if (signal.aborted) {
            cleanup({ closeSocket: true });
            controller.error(signal.reason ?? new DOMException("Aborted", "AbortError"));
            return;
          }
          abortHandler = () => {
            cleanup({ closeSocket: true });
            try {
              controller.error(signal.reason ?? new DOMException("Aborted", "AbortError"));
            } catch {
            }
          };
          signal.addEventListener("abort", abortHandler, { once: true });
        }
        connection.send(JSON.stringify({ type: "response.create", ...requestBody }));
      },
      cancel() {
        cleanupActiveStream?.({ closeSocket: true });
      }
    });
    return new Response(responseStream, {
      status: 200,
      headers: { "content-type": "text/event-stream" }
    });
  }
  return Object.assign(websocketFetch, {
    /** Close the underlying WebSocket connection. */
    close() {
      if (ws) {
        ws.close();
        ws = null;
      }
      connectionKey = null;
      connecting = null;
    }
  });
}
function isResponsesUrl(url) {
  try {
    return new URL(url).pathname.endsWith("/responses");
  } catch {
    return url.split("?")[0]?.endsWith("/responses") ?? false;
  }
}
function getWebSocketUrl(url, apiKeyQueryParam, apiKey) {
  if (!apiKeyQueryParam || !apiKey) return url;
  const parsedUrl = new URL(url);
  parsedUrl.searchParams.set(apiKeyQueryParam, apiKey);
  return parsedUrl.toString();
}
function formatSSEData(text) {
  return `${text.split(/\r?\n/).map((line) => `data: ${line}`).join("\n")}

`;
}
function buildConnectionKey(authorization, headers, queryCredential = "") {
  return JSON.stringify({
    authorization,
    queryCredential,
    headers: Object.entries(headers).sort(([a], [b]) => a.localeCompare(b))
  });
}
function isTerminalWebSocketEvent(event) {
  return typeof event === "object" && event !== null && "type" in event && typeof event.type === "string" && TERMINAL_RESPONSE_EVENTS.has(event.type);
}
function shouldReconnectAfterEvent(event) {
  if (typeof event !== "object" || event === null || !("error" in event)) return false;
  const error = event.error;
  return typeof error === "object" && error !== null && "code" in error && error.code === "websocket_connection_limit_reached";
}
function getAbortError(signal) {
  return signal?.reason ?? new DOMException("Aborted", "AbortError");
}
function normalizeHeaders(headers) {
  const result = {};
  if (!headers) return result;
  if (headers instanceof Headers) {
    headers.forEach((value, key) => {
      result[key.toLowerCase()] = value;
    });
  } else if (Array.isArray(headers)) {
    for (const [key, value] of headers) {
      result[key.toLowerCase()] = value;
    }
  } else {
    for (const [key, value] of Object.entries(headers)) {
      if (value != null) result[key.toLowerCase()] = value;
    }
  }
  return result;
}

// src/llm/model/gateways/azure.ts
function isRecord(value) {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
function mirrorAzureProviderOptionsForOpenAI(providerOptions) {
  if (!isRecord(providerOptions) || !isRecord(providerOptions.azure)) {
    return providerOptions;
  }
  const openai = isRecord(providerOptions.openai) ? providerOptions.openai : {};
  return {
    ...providerOptions,
    openai: {
      ...openai,
      ...providerOptions.azure
    }
  };
}
function mirrorAzureResponseProviderOptions(prompt) {
  let promptModified = false;
  const mirroredPrompt = prompt.map((message) => {
    const messageWithProviderOptions = message;
    const providerOptions = mirrorAzureProviderOptionsForOpenAI(messageWithProviderOptions.providerOptions);
    const providerOptionsModified = providerOptions !== messageWithProviderOptions.providerOptions;
    if (!Array.isArray(message.content)) {
      if (providerOptionsModified) {
        promptModified = true;
        return { ...message, providerOptions };
      }
      return message;
    }
    let contentModified = false;
    const content = message.content.map((part) => {
      if (!("providerOptions" in part)) {
        return part;
      }
      const providerOptions2 = mirrorAzureProviderOptionsForOpenAI(part.providerOptions);
      if (providerOptions2 === part.providerOptions) {
        return part;
      }
      contentModified = true;
      return { ...part, providerOptions: providerOptions2 };
    });
    if (!contentModified) {
      if (providerOptionsModified) {
        promptModified = true;
        return { ...message, providerOptions };
      }
      return message;
    }
    promptModified = true;
    return { ...message, ...providerOptionsModified ? { providerOptions } : {}, content };
  });
  return promptModified ? mirroredPrompt : prompt;
}
function withAzureResponsesInputCompatibility(model) {
  return new Proxy(model, {
    get(target, property, receiver) {
      if (property === "doGenerate") {
        return (options) => target.doGenerate({
          ...options,
          providerOptions: mirrorAzureProviderOptionsForOpenAI(options.providerOptions),
          prompt: mirrorAzureResponseProviderOptions(options.prompt)
        });
      }
      if (property === "doStream") {
        return (options) => target.doStream({
          ...options,
          providerOptions: mirrorAzureProviderOptionsForOpenAI(options.providerOptions),
          prompt: mirrorAzureResponseProviderOptions(options.prompt)
        });
      }
      return Reflect.get(target, property, receiver);
    }
  });
}
var AzureOpenAIGateway = class extends chunkA45A2T7R_cjs.MastraModelGateway {
  constructor(config) {
    super();
    this.config = config;
    this.validateConfig();
  }
  config;
  id = "azure-openai";
  name = "azure-openai";
  tokenCache = new chunk6QR4RQID_cjs.InMemoryServerCache();
  entraIdTokenRequests = /* @__PURE__ */ new Map();
  validateConfig() {
    if (!this.config.resourceName) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_GATEWAY_INVALID_CONFIG",
        domain: "LLM",
        category: "UNKNOWN",
        text: "resourceName is required for Azure OpenAI gateway"
      });
    }
    if (!this.config.apiKey && this.config.authentication?.type !== "entraId") {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_GATEWAY_INVALID_CONFIG",
        domain: "LLM",
        category: "UNKNOWN",
        text: "apiKey or Entra ID authentication is required for Azure OpenAI gateway"
      });
    }
    if (this.config.authentication?.type === "entraId" && !this.config.authentication.credential) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_GATEWAY_INVALID_CONFIG",
        domain: "LLM",
        category: "UNKNOWN",
        text: "credential is required for Azure OpenAI Entra ID authentication"
      });
    }
    if (this.config.apiKey && this.config.authentication?.type === "entraId") {
      console.warn(
        "[AzureOpenAIGateway] Both apiKey and Entra ID authentication provided. Using Entra ID authentication and ignoring apiKey."
      );
    }
    if (this.config.useResponsesAPI && this.config.useDeploymentBasedUrls === true) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_GATEWAY_INVALID_CONFIG",
        domain: "LLM",
        category: "UNKNOWN",
        text: "useResponsesAPI: true cannot be combined with useDeploymentBasedUrls: true. Omit useDeploymentBasedUrls or set it to false."
      });
    }
    if (this.config.useResponsesAPI && this.config.apiVersion && !["v1", "preview"].includes(this.config.apiVersion)) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_GATEWAY_INVALID_CONFIG",
        domain: "LLM",
        category: "UNKNOWN",
        text: 'useResponsesAPI: true requires apiVersion: "v1" or apiVersion: "preview". Omit apiVersion to use "v1".'
      });
    }
    if (this.config.useDeploymentBasedUrls === false && this.config.apiVersion && !["v1", "preview"].includes(this.config.apiVersion)) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_GATEWAY_INVALID_CONFIG",
        domain: "LLM",
        category: "UNKNOWN",
        text: 'useDeploymentBasedUrls: false requires apiVersion: "v1" or apiVersion: "preview". Omit apiVersion to use "v1".'
      });
    }
    const hasDeployments = this.config.deployments && this.config.deployments.length > 0;
    const hasManagement = this.config.management !== void 0;
    if (hasDeployments && hasManagement) {
      console.warn(
        "[AzureOpenAIGateway] Both deployments and management credentials provided. Using static deployments list and ignoring management API."
      );
    }
    if (hasManagement) {
      this.getManagementCredentials(this.config.management);
    }
  }
  async fetchProviders() {
    if (this.config.deployments && this.config.deployments.length > 0) {
      return {
        "azure-openai": {
          apiKeyEnvVar: [],
          apiKeyHeader: "api-key",
          name: "Azure OpenAI",
          models: this.config.deployments,
          docUrl: "https://learn.microsoft.com/en-us/azure/ai-services/openai/",
          gateway: "azure-openai"
        }
      };
    }
    if (!this.config.management) {
      return {
        "azure-openai": {
          apiKeyEnvVar: [],
          apiKeyHeader: "api-key",
          name: "Azure OpenAI",
          models: [],
          docUrl: "https://learn.microsoft.com/en-us/azure/ai-services/openai/",
          gateway: "azure-openai"
        }
      };
    }
    try {
      const credentials = this.getManagementCredentials(this.config.management);
      const token = await this.getAzureADToken({
        tenantId: credentials.tenantId,
        clientId: credentials.clientId,
        clientSecret: credentials.clientSecret
      });
      const deployments = await this.fetchDeployments(token, {
        subscriptionId: credentials.subscriptionId,
        resourceGroup: credentials.resourceGroup,
        resourceName: this.config.resourceName
      });
      return {
        "azure-openai": {
          apiKeyEnvVar: [],
          apiKeyHeader: "api-key",
          name: "Azure OpenAI",
          models: deployments.map((d) => d.name),
          docUrl: "https://learn.microsoft.com/en-us/azure/ai-services/openai/",
          gateway: "azure-openai"
        }
      };
    } catch (error) {
      const errorMsg = error instanceof Error ? error.message : String(error);
      console.warn(
        `[AzureOpenAIGateway] Deployment discovery failed: ${errorMsg}`,
        "\nReturning fallback configuration. Azure OpenAI can still be used by manually specifying deployment names."
      );
      return {
        "azure-openai": {
          apiKeyEnvVar: [],
          apiKeyHeader: "api-key",
          name: "Azure OpenAI",
          models: [],
          docUrl: "https://learn.microsoft.com/en-us/azure/ai-services/openai/",
          gateway: "azure-openai"
        }
      };
    }
  }
  getManagementCredentials(management) {
    const { tenantId, clientId, clientSecret, subscriptionId, resourceGroup } = management;
    const missing = [];
    if (!tenantId) missing.push("tenantId");
    if (!clientId) missing.push("clientId");
    if (!clientSecret) missing.push("clientSecret");
    if (!subscriptionId) missing.push("subscriptionId");
    if (!resourceGroup) missing.push("resourceGroup");
    if (missing.length > 0) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_MANAGEMENT_CREDENTIALS_MISSING",
        domain: "LLM",
        category: "UNKNOWN",
        text: `Management credentials incomplete. Missing: ${missing.join(", ")}. Required fields: tenantId, clientId, clientSecret, subscriptionId, resourceGroup.`
      });
    }
    return {
      tenantId,
      clientId,
      clientSecret,
      subscriptionId,
      resourceGroup
    };
  }
  async getAzureADToken(credentials) {
    const { tenantId, clientId, clientSecret } = credentials;
    const cacheKey = `azure-mgmt-token:${tenantId}:${clientId}`;
    const cached = await this.tokenCache.get(cacheKey);
    if (cached && cached.expiresAt > Date.now() / 1e3 + 60) {
      return cached.token;
    }
    const tokenEndpoint = `https://login.microsoftonline.com/${tenantId}/oauth2/v2.0/token`;
    const body = new URLSearchParams({
      grant_type: "client_credentials",
      client_id: clientId,
      client_secret: clientSecret,
      scope: "https://management.azure.com/.default"
    });
    const response = await fetch(tokenEndpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded"
      },
      body: body.toString()
    });
    if (!response.ok) {
      const error = await response.text();
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_AD_TOKEN_ERROR",
        domain: "LLM",
        category: "UNKNOWN",
        text: `Failed to get Azure AD token: ${response.status} ${error}`
      });
    }
    const tokenResponse = await response.json();
    const expiresAt = Math.floor(Date.now() / 1e3) + tokenResponse.expires_in;
    await this.tokenCache.set(cacheKey, {
      token: tokenResponse.access_token,
      expiresAt
    });
    return tokenResponse.access_token;
  }
  async fetchDeployments(token, credentials) {
    const { subscriptionId, resourceGroup, resourceName } = credentials;
    let url = `https://management.azure.com/subscriptions/${subscriptionId}/resourceGroups/${resourceGroup}/providers/Microsoft.CognitiveServices/accounts/${resourceName}/deployments?api-version=2024-10-01`;
    const allDeployments = [];
    while (url) {
      const response = await fetch(url, {
        headers: {
          Authorization: `Bearer ${token}`,
          "Content-Type": "application/json"
        }
      });
      if (!response.ok) {
        const error = await response.text();
        throw new chunkXSOONORA_cjs.MastraError({
          id: "AZURE_DEPLOYMENTS_FETCH_ERROR",
          domain: "LLM",
          category: "UNKNOWN",
          text: `Failed to fetch Azure deployments: ${response.status} ${error}`
        });
      }
      const data = await response.json();
      allDeployments.push(...data.value);
      url = data.nextLink;
    }
    const successfulDeployments = allDeployments.filter((d) => d.properties.provisioningState === "Succeeded");
    return successfulDeployments;
  }
  buildUrl(_routerId, _envVars) {
    return void 0;
  }
  async getApiKey(_modelId) {
    return this.config.authentication?.type === "entraId" ? "" : this.config.apiKey ?? "";
  }
  async getEntraIdToken() {
    if (this.config.authentication?.type !== "entraId") {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_ENTRA_ID_AUTH_NOT_CONFIGURED",
        domain: "LLM",
        category: "UNKNOWN",
        text: "Entra ID authentication is not configured for Azure OpenAI gateway"
      });
    }
    const scope = this.config.authentication.scope ?? "https://cognitiveservices.azure.com/.default";
    const cacheKey = `azure-openai-token:${scope}`;
    const cached = await this.tokenCache.get(cacheKey);
    if (cached && cached.expiresAt > Date.now() / 1e3 + 60) {
      return cached.token;
    }
    let tokenRequest = this.entraIdTokenRequests.get(cacheKey);
    if (!tokenRequest) {
      tokenRequest = this.fetchEntraIdToken(scope, cacheKey);
      this.entraIdTokenRequests.set(cacheKey, tokenRequest);
    }
    try {
      const token = await tokenRequest;
      return token.token;
    } finally {
      this.entraIdTokenRequests.delete(cacheKey);
    }
  }
  async fetchEntraIdToken(scope, cacheKey) {
    if (this.config.authentication?.type !== "entraId") {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_ENTRA_ID_AUTH_NOT_CONFIGURED",
        domain: "LLM",
        category: "UNKNOWN",
        text: "Entra ID authentication is not configured for Azure OpenAI gateway"
      });
    }
    const accessToken = await this.config.authentication.credential.getToken(scope);
    if (!accessToken?.token) {
      throw new chunkXSOONORA_cjs.MastraError({
        id: "AZURE_ENTRA_ID_TOKEN_ERROR",
        domain: "LLM",
        category: "UNKNOWN",
        text: "Failed to get Entra ID token for Azure OpenAI gateway"
      });
    }
    const token = {
      token: accessToken.token,
      expiresAt: accessToken.expiresOnTimestamp ? Math.floor(accessToken.expiresOnTimestamp / 1e3) : Math.floor(Date.now() / 1e3) + 300
    };
    await this.tokenCache.set(cacheKey, token);
    return token;
  }
  createEntraIdFetch(innerFetch = fetch) {
    return async (input, init) => {
      const token = await this.getEntraIdToken();
      const headers = new Headers(init?.headers);
      headers.delete("api-key");
      headers.set("Authorization", `Bearer ${token}`);
      return innerFetch(input, {
        ...init,
        headers
      });
    };
  }
  createAzureResponsesWebSocketFetch({
    useEntraId,
    responsesWebSocket
  }) {
    const websocketFetch = createOpenAIWebSocketFetch({
      url: responsesWebSocket?.url ?? `wss://${this.config.resourceName}.openai.azure.com/openai/v1/responses`,
      headers: responsesWebSocket?.headers,
      apiKeyQueryParam: useEntraId ? false : "api-key",
      betaHeader: false
    });
    return useEntraId ? Object.assign(this.createEntraIdFetch(websocketFetch), { close: websocketFetch.close }) : websocketFetch;
  }
  async resolveLanguageModel({
    modelId,
    apiKey,
    headers,
    transport,
    responsesWebSocket
  }) {
    const useResponsesAPI = this.config.useResponsesAPI ?? false;
    const apiVersion = this.config.apiVersion || (useResponsesAPI || this.config.useDeploymentBasedUrls === false ? "v1" : "2024-04-01-preview");
    const useDeploymentBasedUrls = this.config.useDeploymentBasedUrls ?? (useResponsesAPI ? false : true);
    const useEntraId = this.config.authentication?.type === "entraId";
    const useWebSocket = useResponsesAPI && transport === "websocket";
    const websocketFetch = useWebSocket ? this.createAzureResponsesWebSocketFetch({
      useEntraId,
      responsesWebSocket
    }) : void 0;
    const azureConfig = {
      resourceName: this.config.resourceName,
      apiKey: useEntraId ? "" : apiKey,
      apiVersion,
      // Mastra's Azure gateway has historically used deployment-based URLs.
      // Keep that default for compatibility; set false with apiVersion: 'v1'
      // to use the newer Azure OpenAI v1 route.
      useDeploymentBasedUrls,
      headers: { "User-Agent": chunkA45A2T7R_cjs.MASTRA_USER_AGENT, ...headers },
      ...websocketFetch && !useEntraId ? { fetch: websocketFetch } : {}
    };
    const azureProvider = createAzure(
      useEntraId ? {
        ...azureConfig,
        fetch: websocketFetch ?? this.createEntraIdFetch()
      } : azureConfig
    );
    if (useResponsesAPI) {
      const model = withAzureResponsesInputCompatibility(azureProvider.responses(modelId));
      if (websocketFetch) {
        Object.defineProperty(model, chunkA45A2T7R_cjs.MASTRA_GATEWAY_STREAM_TRANSPORT, {
          configurable: true,
          value: {
            type: "openai-websocket",
            close: websocketFetch.close
          }
        });
      }
      return model;
    }
    return azureProvider(modelId);
  }
};

// src/llm/model/gateways/gateway-manager.ts
var MISSING_AUTH_ERROR_IDS = /* @__PURE__ */ new Set([
  "MODEL_ROUTER_NO_GATEWAY_FOUND",
  "MASTRA_GATEWAY_NO_API_KEY",
  "NETLIFY_GATEWAY_NO_TOKEN",
  "NETLIFY_GATEWAY_NO_SITE_ID",
  "AZURE_ENTRA_ID_AUTH_NOT_CONFIGURED"
]);
function isExpectedMissingAuthError(error) {
  if (error instanceof chunkXSOONORA_cjs.MastraError) {
    return MISSING_AUTH_ERROR_IDS.has(error.id);
  }
  if (error instanceof Error) {
    const msg = error.message;
    return /Missing [^ ]+ environment variable/i.test(msg) || /Could not find API key/i.test(msg) || /no api key/i.test(msg) || /Could not find config for provider/i.test(msg) || /Could not identify provider/i.test(msg);
  }
  return false;
}
var GatewayManager = class _GatewayManager {
  gateways;
  constructor(gateways = []) {
    const seen = /* @__PURE__ */ new Set();
    this.gateways = gateways.filter((gateway) => {
      if (!chunk5FCWY4Z3_cjs.shouldEnableGateway(gateway)) return false;
      const id = chunk5FCWY4Z3_cjs.getGatewayId(gateway);
      if (seen.has(id)) return false;
      seen.add(id);
      return true;
    });
  }
  /**
   * Returns the gateway prefix used for model-id parsing.
   * `models.dev` is a provider registry and doesn't use a prefix.
   */
  static getPrefix(gatewayId) {
    return gatewayId === "models.dev" ? void 0 : gatewayId;
  }
  /**
   * Returns the prefix to use when parsing `routerId` for the given gateway.
   * A gateway may claim an unprefixed id (e.g. `handlesModel` matching a bare
   * `anthropic/...` id); in that case the gateway's prefix does not appear in
   * the id, so parse it as an unprefixed `provider/model` id.
   */
  static getPrefixForId(gatewayId, routerId) {
    const prefix = _GatewayManager.getPrefix(gatewayId);
    if (prefix && !routerId.startsWith(`${prefix}/`)) {
      return void 0;
    }
    return prefix;
  }
  /** Find the gateway that handles a given model/router id. */
  findGatewayForModel(routerId) {
    return chunk5FCWY4Z3_cjs.findGatewayForModel(routerId, this.gateways);
  }
  /**
   * Resolve the gateway and parsed provider/model components for a router id
   * in a single pass. Centralises gateway selection + id parsing so callers
   * (router constructor, doGenerate/doStream, supportedUrls) don't each
   * re-derive the gateway and prefix separately.
   */
  resolveModelId(routerId) {
    const gateway = this.findGatewayForModel(routerId);
    const gatewayId = chunk5FCWY4Z3_cjs.getGatewayId(gateway);
    const { providerId, modelId } = chunkWFEGLWHX_cjs.parseModelRouterId(routerId, _GatewayManager.getPrefixForId(gatewayId, routerId));
    return { gateway, gatewayId, providerId, modelId };
  }
  /** Parse a router id into its provider/model/gateway components. */
  parseModelId(routerId) {
    const { gatewayId, providerId, modelId } = this.resolveModelId(routerId);
    return { providerId, modelId, gatewayId };
  }
  /**
   * Resolve auth through the gateway chain: select the gateway via
   * {@link findGatewayForModel}, try its `resolveAuth()` (OAuth / stored
   * credentials), and fall back to `getApiKey()` (which, for registry /
   * openai-compatible providers, reads the API-key env var).
   *
   * The caller is responsible for merging any explicit headers/credentials
   * from per-instance config on top of the result returned here.
   */
  async resolveAuth(routerId) {
    const { gateway, gatewayId, providerId, modelId } = this.resolveModelId(routerId);
    const request = { gatewayId, providerId, modelId, routerId };
    const rawGatewayAuth = await gateway.resolveAuth?.(request);
    const gatewayAuth = rawGatewayAuth?.bearerToken ? {
      ...rawGatewayAuth,
      headers: { ...rawGatewayAuth.headers, Authorization: `Bearer ${rawGatewayAuth.bearerToken}` }
    } : rawGatewayAuth;
    if (gatewayAuth?.apiKey || gatewayAuth?.headers || gatewayAuth?.bearerToken) {
      return {
        ...gatewayAuth,
        source: gatewayAuth.source ?? "gateway"
      };
    }
    return {
      apiKey: await gateway.getApiKey(routerId),
      source: "legacy"
    };
  }
  /**
   * Convenience: whether auth is available for a model.
   * Returns `false` for expected missing-auth states (no matching gateway,
   * missing credentials/env vars) but re-throws unexpected errors (token
   * exchange failures, network bugs, malformed auth hooks) so they surface
   * instead of being silently hidden.
   */
  async hasAuth(routerId) {
    try {
      const auth = await this.resolveAuth(routerId);
      return Boolean(auth.apiKey || auth.bearerToken || auth.headers);
    } catch (error) {
      if (isExpectedMissingAuthError(error)) {
        return false;
      }
      throw error;
    }
  }
  /**
   * Fetch and flatten providers from all gateways, deduped by provider key
   * (configured / earlier gateway wins). Each gateway's `fetchProviders()`
   * is a network call for gateways like models.dev / Netlify.
   */
  async listProviders() {
    const registry = {};
    for (const gateway of this.gateways) {
      try {
        const gatewayProviders = await gateway.fetchProviders();
        for (const [providerId, config] of Object.entries(gatewayProviders)) {
          const key = this.getProviderKey(gateway, providerId);
          if (key in registry) continue;
          registry[key] = { ...config, gateway: chunk5FCWY4Z3_cjs.getGatewayId(gateway) };
        }
      } catch (error) {
        console.warn(`Failed to load providers from gateway ${chunk5FCWY4Z3_cjs.getGatewayId(gateway)}:`, error);
      }
    }
    return registry;
  }
  /**
   * Build the model catalog from gateway providers, resolving auth per
   * provider (using the first model). Returns models without use-count.
   */
  async listAvailableModels() {
    const registry = await this.listProviders();
    const models = [];
    for (const [provider, providerConfig] of Object.entries(registry)) {
      const apiKeyEnvVar = Array.isArray(providerConfig.apiKeyEnvVar) ? providerConfig.apiKeyEnvVar[0] : providerConfig.apiKeyEnvVar;
      const modelNames = providerConfig.models;
      if (!Array.isArray(modelNames)) continue;
      const hasApiKey = modelNames[0] ? await this.hasAuth(`${provider}/${modelNames[0]}`) : false;
      for (const modelName of modelNames) {
        models.push({
          id: `${provider}/${modelName}`,
          provider,
          modelName,
          hasApiKey,
          apiKeyEnvVar: apiKeyEnvVar || void 0
        });
      }
    }
    return models;
  }
  /** Provider key used for catalog ids: prefix with the gateway id unless it's the prefix-less models.dev registry. */
  getProviderKey(gateway, providerId) {
    const gatewayId = chunk5FCWY4Z3_cjs.getGatewayId(gateway);
    if (gatewayId === "models.dev") return providerId;
    return providerId === gatewayId ? gatewayId : `${gatewayId}/${providerId}`;
  }
};

// src/llm/model/gateways/defaults.ts
function getStaticProvidersByGateway(name) {
  return Object.fromEntries(Object.entries(chunk5FCWY4Z3_cjs.PROVIDER_REGISTRY).filter(([_provider, config]) => config.gateway === name));
}
var defaultGateways = [
  new chunkVIRV756L_cjs.NetlifyGateway(),
  new chunk5FCWY4Z3_cjs.MastraGateway(),
  new chunkWFEGLWHX_cjs.ModelsDevGateway(getStaticProvidersByGateway(`models.dev`))
];

exports.AzureOpenAIGateway = AzureOpenAIGateway;
exports.GatewayManager = GatewayManager;
exports.createOpenAIWebSocketFetch = createOpenAIWebSocketFetch;
exports.defaultGateways = defaultGateways;
