// @mastra/core/dist/fga-check-BIMTM2YF.cjs
'use strict';

var chunkW7RLUWK5_cjs = require('./chunk-W7RLUWK5.cjs');



Object.defineProperty(exports, "ACTIONS", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.ACTIONS; }
});
Object.defineProperty(exports, "DEFAULT_ROLES", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.DEFAULT_ROLES; }
});
Object.defineProperty(exports, "FGADeniedError", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.FGADeniedError; }
});
Object.defineProperty(exports, "MastraFGAPermissions", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.MastraFGAPermissions; }
});
Object.defineProperty(exports, "PERMISSIONS", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.PERMISSIONS; }
});
Object.defineProperty(exports, "PERMISSION_PATTERNS", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.PERMISSION_PATTERNS; }
});
Object.defineProperty(exports, "RESOURCES", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.RESOURCES; }
});
Object.defineProperty(exports, "StaticRBACProvider", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.StaticRBACProvider; }
});
Object.defineProperty(exports, "buildCapabilities", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.buildCapabilities; }
});
Object.defineProperty(exports, "checkFGA", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.checkFGA; }
});
Object.defineProperty(exports, "clearLicenseCache", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.clearLicenseCache; }
});
Object.defineProperty(exports, "getAgentFGAResourceId", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getAgentFGAResourceId; }
});
Object.defineProperty(exports, "getAgentToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getAgentToolFGAResourceId; }
});
Object.defineProperty(exports, "getDefaultRole", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getDefaultRole; }
});
Object.defineProperty(exports, "getMCPToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getMCPToolFGAResourceId; }
});
Object.defineProperty(exports, "getSafeLicenseSummary", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getSafeLicenseSummary; }
});
Object.defineProperty(exports, "getStandaloneToolFGAResourceId", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getStandaloneToolFGAResourceId; }
});
Object.defineProperty(exports, "getWorkflowFGAResourceId", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.getWorkflowFGAResourceId; }
});
Object.defineProperty(exports, "hasPermission", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.hasPermission; }
});
Object.defineProperty(exports, "isAuthenticated", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isAuthenticated; }
});
Object.defineProperty(exports, "isDevEnvironment", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isDevEnvironment; }
});
Object.defineProperty(exports, "isEEEnabled", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isEEEnabled; }
});
Object.defineProperty(exports, "isEELicenseValid", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isEELicenseValid; }
});
Object.defineProperty(exports, "isFeatureEnabled", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isFeatureEnabled; }
});
Object.defineProperty(exports, "isLicenseValid", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isLicenseValid; }
});
Object.defineProperty(exports, "isValidPermissionPattern", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.isValidPermissionPattern; }
});
Object.defineProperty(exports, "matchesPermission", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.matchesPermission; }
});
Object.defineProperty(exports, "requireFGA", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.requireFGA; }
});
Object.defineProperty(exports, "resolvePermissions", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.resolvePermissions; }
});
Object.defineProperty(exports, "resolvePermissionsFromMapping", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.resolvePermissionsFromMapping; }
});
Object.defineProperty(exports, "startLicenseValidation", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.startLicenseValidation; }
});
Object.defineProperty(exports, "validateLicense", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.validateLicense; }
});
Object.defineProperty(exports, "validatePermissions", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.validatePermissions; }
});
Object.defineProperty(exports, "warnIfDevEENeedsLicense", {
  enumerable: true,
  get: function () { return chunkW7RLUWK5_cjs.warnIfDevEENeedsLicense; }
});
