// @mastra/core/dist/provider-registry-5U43NFLJ.cjs
'use strict';

var chunk5FCWY4Z3_cjs = require('./chunk-5FCWY4Z3.cjs');



Object.defineProperty(exports, "GatewayRegistry", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.GatewayRegistry; }
});
Object.defineProperty(exports, "PROVIDER_MODELS", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.PROVIDER_MODELS; }
});
Object.defineProperty(exports, "PROVIDER_REGISTRY", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.PROVIDER_REGISTRY; }
});
Object.defineProperty(exports, "getProviderConfig", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.getProviderConfig; }
});
Object.defineProperty(exports, "getRegisteredProviders", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.getRegisteredProviders; }
});
Object.defineProperty(exports, "isOfflineMode", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.isOfflineMode; }
});
Object.defineProperty(exports, "isProviderRegistered", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.isProviderRegistered; }
});
Object.defineProperty(exports, "isValidModelId", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.isValidModelId; }
});
Object.defineProperty(exports, "modelSupportsAttachments", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.modelSupportsAttachments; }
});
Object.defineProperty(exports, "parseModelString", {
  enumerable: true,
  get: function () { return chunk5FCWY4Z3_cjs.parseModelString; }
});
