// @mastra/core/dist/workspace/index.cjs
'use strict';

var chunkQJ3FDYCK_cjs = require('../chunk-QJ3FDYCK.cjs');
var chunkZNDOO55R_cjs = require('../chunk-ZNDOO55R.cjs');



Object.defineProperty(exports, "CompositeFilesystem", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.CompositeFilesystem; }
});
Object.defineProperty(exports, "CompositeVersionedSkillSource", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.CompositeVersionedSkillSource; }
});
Object.defineProperty(exports, "FilesystemNotMountableError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.FilesystemNotMountableError; }
});
Object.defineProperty(exports, "IsolationUnavailableError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.IsolationUnavailableError; }
});
Object.defineProperty(exports, "LocalFilesystem", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.LocalFilesystem; }
});
Object.defineProperty(exports, "LocalSandbox", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.LocalSandbox; }
});
Object.defineProperty(exports, "MastraFilesystem", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.MastraFilesystem; }
});
Object.defineProperty(exports, "MastraSandbox", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.MastraSandbox; }
});
Object.defineProperty(exports, "MountError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.MountError; }
});
Object.defineProperty(exports, "MountManager", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.MountManager; }
});
Object.defineProperty(exports, "MountNotSupportedError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.MountNotSupportedError; }
});
Object.defineProperty(exports, "ProcessHandle", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.ProcessHandle; }
});
Object.defineProperty(exports, "SandboxError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.SandboxError; }
});
Object.defineProperty(exports, "SandboxExecutionError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.SandboxExecutionError; }
});
Object.defineProperty(exports, "SandboxNotReadyError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.SandboxNotReadyError; }
});
Object.defineProperty(exports, "SandboxProcessManager", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.SandboxProcessManager; }
});
Object.defineProperty(exports, "SandboxTimeoutError", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.SandboxTimeoutError; }
});
Object.defineProperty(exports, "VersionedSkillSource", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.VersionedSkillSource; }
});
Object.defineProperty(exports, "WORKSPACE_TOOLS", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.WORKSPACE_TOOLS; }
});
Object.defineProperty(exports, "WORKSPACE_TOOLS_PREFIX", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.WORKSPACE_TOOLS_PREFIX; }
});
Object.defineProperty(exports, "Workspace", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.Workspace; }
});
Object.defineProperty(exports, "callLifecycle", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.callLifecycle; }
});
Object.defineProperty(exports, "collectSkillForPublish", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.collectSkillForPublish; }
});
Object.defineProperty(exports, "createSkillTools", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.createSkillTools; }
});
Object.defineProperty(exports, "createWorkspaceTools", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.createWorkspaceTools; }
});
Object.defineProperty(exports, "deleteFileTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.deleteFileTool; }
});
Object.defineProperty(exports, "detectIsolation", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.detectIsolation; }
});
Object.defineProperty(exports, "editFileTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.editFileTool; }
});
Object.defineProperty(exports, "executeCommandTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.executeCommandTool; }
});
Object.defineProperty(exports, "fileStatTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.fileStatTool; }
});
Object.defineProperty(exports, "formatSkillActivation", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.formatSkillActivation; }
});
Object.defineProperty(exports, "getRecommendedIsolation", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.getRecommendedIsolation; }
});
Object.defineProperty(exports, "indexContentTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.indexContentTool; }
});
Object.defineProperty(exports, "isIsolationAvailable", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.isIsolationAvailable; }
});
Object.defineProperty(exports, "listFilesTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.listFilesTool; }
});
Object.defineProperty(exports, "mkdirTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.mkdirTool; }
});
Object.defineProperty(exports, "publishSkillFromSource", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.publishSkillFromSource; }
});
Object.defineProperty(exports, "readFileTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.readFileTool; }
});
Object.defineProperty(exports, "requireFilesystem", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.requireFilesystem; }
});
Object.defineProperty(exports, "requireSandbox", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.requireSandbox; }
});
Object.defineProperty(exports, "requireWorkspace", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.requireWorkspace; }
});
Object.defineProperty(exports, "resolveToolConfig", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.resolveToolConfig; }
});
Object.defineProperty(exports, "searchTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.searchTool; }
});
Object.defineProperty(exports, "writeFileTool", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.writeFileTool; }
});
Object.defineProperty(exports, "DirectoryNotEmptyError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.DirectoryNotEmptyError; }
});
Object.defineProperty(exports, "DirectoryNotFoundError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.DirectoryNotFoundError; }
});
Object.defineProperty(exports, "FileExistsError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.FileExistsError; }
});
Object.defineProperty(exports, "FileNotFoundError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.FileNotFoundError; }
});
Object.defineProperty(exports, "FileReadRequiredError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.FileReadRequiredError; }
});
Object.defineProperty(exports, "FilesystemError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.FilesystemError; }
});
Object.defineProperty(exports, "FilesystemNotAvailableError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.FilesystemNotAvailableError; }
});
Object.defineProperty(exports, "FilesystemNotReadyError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.FilesystemNotReadyError; }
});
Object.defineProperty(exports, "IsDirectoryError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.IsDirectoryError; }
});
Object.defineProperty(exports, "LocalSkillSource", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.LocalSkillSource; }
});
Object.defineProperty(exports, "NotDirectoryError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.NotDirectoryError; }
});
Object.defineProperty(exports, "PermissionError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.PermissionError; }
});
Object.defineProperty(exports, "SandboxFeatureNotSupportedError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.SandboxFeatureNotSupportedError; }
});
Object.defineProperty(exports, "SandboxNotAvailableError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.SandboxNotAvailableError; }
});
Object.defineProperty(exports, "SearchNotAvailableError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.SearchNotAvailableError; }
});
Object.defineProperty(exports, "StaleFileError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.StaleFileError; }
});
Object.defineProperty(exports, "WorkspaceError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.WorkspaceError; }
});
Object.defineProperty(exports, "WorkspaceNotAvailableError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.WorkspaceNotAvailableError; }
});
Object.defineProperty(exports, "WorkspaceNotReadyError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.WorkspaceNotReadyError; }
});
Object.defineProperty(exports, "WorkspaceReadOnlyError", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.WorkspaceReadOnlyError; }
});
Object.defineProperty(exports, "createGlobMatcher", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.createGlobMatcher; }
});
Object.defineProperty(exports, "extractGlobBase", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.extractGlobBase; }
});
Object.defineProperty(exports, "isGlobPattern", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.isGlobPattern; }
});
Object.defineProperty(exports, "matchGlob", {
  enumerable: true,
  get: function () { return chunkZNDOO55R_cjs.matchGlob; }
});
