// @mastra/core/dist/signals/index.cjs
'use strict';

var chunk7QFVYQT7_cjs = require('../chunk-7QFVYQT7.cjs');
var chunkQJ3FDYCK_cjs = require('../chunk-QJ3FDYCK.cjs');



Object.defineProperty(exports, "TaskSignalProvider", {
  enumerable: true,
  get: function () { return chunk7QFVYQT7_cjs.TaskSignalProvider; }
});
Object.defineProperty(exports, "GoalSignalProvider", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.GoalSignalProvider; }
});
Object.defineProperty(exports, "SignalProvider", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.SignalProvider; }
});
Object.defineProperty(exports, "WebhookSignalProvider", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.WebhookSignalProvider; }
});
Object.defineProperty(exports, "assignTaskIds", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.assignTaskIds; }
});
Object.defineProperty(exports, "isSignalProvider", {
  enumerable: true,
  get: function () { return chunkQJ3FDYCK_cjs.isSignalProvider; }
});
